<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Fred2.Core &mdash; FRED 2.01 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.01',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="FRED 2.01 documentation" href="index.html" />
    <link rel="next" title="Fred2.IO" href="Fred2.IO.html" />
    <link rel="prev" title="Basis modules" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Fred2.IO.html" title="Fred2.IO"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Basis modules"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">FRED 2.01 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="fred2-core">
<h1>Fred2.Core<a class="headerlink" href="#fred2-core" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-Fred2.Core.Allele">
<span id="fred2-core-allele"></span><h2>Fred2.Core.Allele<a class="headerlink" href="#module-Fred2.Core.Allele" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Allele"></span><dl class="class">
<dt id="Fred2.Core.Allele.Allele">
<em class="property">class </em><code class="descclassname">Fred2.Core.Allele.</code><code class="descname">Allele</code><span class="sig-paren">(</span><em>_name</em>, <em>prob=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Allele.html#Allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Allele.Allele" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a></p>
<p>This class represents an Allele and stores additional informations as a 
dictionary</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>name</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the name of the MHC allele (new nomenclature A*01:01)</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Base">
<span id="fred2-core-base"></span><h2>Fred2.Core.Base<a class="headerlink" href="#module-Fred2.Core.Base" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Base"></span><dl class="class">
<dt id="Fred2.Core.Base.AExternal">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">AExternal</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>Base class for external tools</p>
<dl class="attribute">
<dt id="Fred2.Core.Base.AExternal.command">
<code class="descname">command</code><a class="headerlink" href="#Fred2.Core.Base.AExternal.command" title="Permalink to this definition">¶</a></dt>
<dd><p>defines the external execution path ?
:return:</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.AExternal.parse_external_result">
<code class="descname">parse_external_result</code><span class="sig-paren">(</span><em>_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal.parse_external_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal.parse_external_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses external NetMHC results and returns a Result object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>_file</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; The file path or the external prediction results</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Result - Returns a Result object</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.APluginRegister">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">APluginRegister</code><span class="sig-paren">(</span><em>name</em>, <em>bases</em>, <em>nmspc</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#APluginRegister"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.APluginRegister" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/abc.html#abc.ABCMeta" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">abc.ABCMeta</span></code></a></p>
<p>This class allows automatic registration of new plugins.</p>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.ASVM">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">ASVM</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#ASVM"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ASVM" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>Base class for SVM prediction tools</p>
<dl class="method">
<dt id="Fred2.Core.Base.ASVM.encode">
<code class="descname">encode</code><span class="sig-paren">(</span><em>peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#ASVM.encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ASVM.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the feature encoding for peptides</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>peptides</strong> (<em>List(Peptide)/Peptide</em>) &#8211; List oder Peptide object</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list(Object) &#8211; Feature encoding of the Peptide objects</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.MetadataLogger">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">MetadataLogger</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#MetadataLogger"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.MetadataLogger" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><code class="xref py py-class docutils literal"><span class="pre">object</span></code></a></p>
<p>This class provides a simple interface for assigning additional metadata to
any object in our data model. Examples: storing ANNOVAR columns like depth,
base count, dbSNP id, quality information for variants, prediction scores 
for peptides etc. Normally used by custom toolbox functions and importers.</p>
<p>The saved values are accessed via <code class="xref py py-meth docutils literal"><span class="pre">log_metadata()</span></code> and
<code class="xref py py-meth docutils literal"><span class="pre">get_metadata()</span></code></p>
<dl class="method">
<dt id="Fred2.Core.Base.MetadataLogger.get_metadata">
<code class="descname">get_metadata</code><span class="sig-paren">(</span><em>label</em>, <em>only_first=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#MetadataLogger.get_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.MetadataLogger.get_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Getter for the saved metadata with the key <code class="xref py py-attr docutils literal"><span class="pre">label</span></code></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>label</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; key for the metadata that is inferred</li>
<li><strong>only_first</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) &#8211; true if only the the first element of the 
matadata list is to be returned</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.MetadataLogger.log_metadata">
<code class="descname">log_metadata</code><span class="sig-paren">(</span><em>label</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#MetadataLogger.log_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.MetadataLogger.log_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Inserts a new metadata</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>label</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; key for the metadata that will be added</li>
<li><strong>value</strong> (<em>list(object)</em>) &#8211; any kindy of additional value that should be
kept</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Generator">
<span id="fred2-core-generator"></span><h2>Fred2.Core.Generator<a class="headerlink" href="#module-Fred2.Core.Generator" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Generator"></span><dl class="function">
<dt id="Fred2.Core.Generator.generate_peptides_from_protein">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_peptides_from_protein</code><span class="sig-paren">(</span><em>proteins</em>, <em>window_size</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_peptides_from_protein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_peptides_from_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates all peptides for a given window size, from a given protein. The
result is a generator.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>protein</strong> (<a class="reference internal" href="#Fred2.Core.Protein.Protein" title="Fred2.Core.Protein.Protein"><em>Protein</em></a>) &#8211; (list of) protein(s) from which a list of unique
peptides should be generated</li>
<li><strong>window_size</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; size of peptide fragments</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_peptides_from_variants">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_peptides_from_variants</code><span class="sig-paren">(</span><em>vars</em>, <em>length</em>, <em>dbadapter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_peptides_from_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_peptides_from_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>generates polymorphic peptides based on given variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>vars</strong> (<em>list(Variant)</em>) &#8211; list of Variants</li>
<li><strong>length</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; length of peptides</li>
<li><strong>dbadapter</strong> &#8211; DBAdapter to fetch the transcript sequences</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">List(Peptide) &#8211; A list of polymorphic peptides</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_transcripts_from_tumor_variants">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_transcripts_from_tumor_variants</code><span class="sig-paren">(</span><em>normal</em>, <em>tumor</em>, <em>dbadapter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_transcripts_from_tumor_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_transcripts_from_tumor_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>generates all possible transcript variations of the given variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>normal</strong> (<em>list(Variant)</em>) &#8211; A list of variants of the normal tissue</li>
<li><strong>tumor</strong> (<em>list(Variant)</em>) &#8211; A list of variant of the cancer tissue for which transcript should be generated</li>
<li><strong>dbadapter</strong> (<a class="reference internal" href="Fred2.IO.html#module-IO.ADBAdapter" title="IO.ADBAdapter: Base class for specific DB-Adapters"><em>ADBAdapter</em></a>) &#8211; a DBAdapter to fetch the transcript sequences</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">(Generator(Transcripts)) &#8211; a generator of transcripts with all
possible variations determined by the given
variant list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_transcripts_from_variants">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_transcripts_from_variants</code><span class="sig-paren">(</span><em>vars</em>, <em>dbadapter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_transcripts_from_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_transcripts_from_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>generates all possible transcript variations of the given variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>vars</strong> (<em>list(Variant)</em>) &#8211; A list of variants for which transcripts should 
be build</li>
<li><strong>dbadapter</strong> (<a class="reference internal" href="Fred2.IO.html#module-IO.ADBAdapter" title="IO.ADBAdapter: Base class for specific DB-Adapters"><em>ADBAdapter</em></a>) &#8211; a DBAdapter to fetch the transcript sequences</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">(Generator(Transcripts)) &#8211; a generator of transcripts with all 
possible variations determined by the given
variant list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Peptide">
<span id="fred2-core-peptide"></span><h2>Fred2.Core.Peptide<a class="headerlink" href="#module-Fred2.Core.Peptide" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Peptide"></span><dl class="class">
<dt id="Fred2.Core.Peptide.Peptide">
<em class="property">class </em><code class="descclassname">Fred2.Core.Peptide.</code><code class="descname">Peptide</code><span class="sig-paren">(</span><em>_seq</em>, <em>proteins=None</em>, <em>vars=None</em>, <em>transcripts=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Peptide.html#Peptide"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Peptide.Peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>Peptides, belonging to one or several proteins.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For accessing and manipulating the sequence see also <code class="xref py py-mod docutils literal"><span class="pre">Bio.Seq.Seq</span></code>
(from Biopython)</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>proteins</strong> (<em>dict(str,Protein)</em>) &#8211; dict of transcript_IDs to protein
instances that could generate that 
peptide</li>
<li><strong>vars</strong> (<em>dict(str,list(Variant))</em>) &#8211; dict of transcript_IDs to a list of
variants that affected the peptide,
(including frame shifts that started not
directly within the peptide)</li>
<li><strong>transcripts</strong> (<em>dict(str,Transcript)</em>) &#8211; dict of transcript_IDs to 
transcript instances that could 
have generated the peptide</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Protein">
<span id="fred2-core-protein"></span><h2>Fred2.Core.Protein<a class="headerlink" href="#module-Fred2.Core.Protein" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Protein"></span><dl class="class">
<dt id="Fred2.Core.Protein.Protein">
<em class="property">class </em><code class="descclassname">Fred2.Core.Protein.</code><code class="descname">Protein</code><span class="sig-paren">(</span><em>_seq</em>, <em>_gene_id</em>, <em>_transcript_id</em>, <em>_orig_transcript=None</em>, <em>_vars=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Protein.html#Protein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Protein.Protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Protein corresponding to exactly one transcript.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For accessing and manipulating the sequence see also <code class="xref py py-mod docutils literal"><span class="pre">Bio.Seq.Seq</span></code>
(from Biopython)</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gene_id</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; ID of the genome</li>
<li><strong>transcript_id</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; ID of the corresponding transcript</li>
<li><strong>orig_transcript</strong> (<a class="reference internal" href="#Fred2.Core.Transcript.Transcript" title="Fred2.Core.Transcript.Transcript"><em>Transcript</em></a>) &#8211; Reference to the originating transcript</li>
<li><strong>_vars</strong> (<em>dict(int,list(Variant))</em>) &#8211; Nonsynonymous variants that are
assoziated with the protein. 
key=position within protein, 
value=list of variants at that pos</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Result">
<span id="fred2-core-result"></span><h2>Fred2.Core.Result<a class="headerlink" href="#module-Fred2.Core.Result" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Result"></span><dl class="class">
<dt id="Fred2.Core.Result.CleavageFragmentPredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">CleavageFragmentPredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageFragmentPredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageFragmentPredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Epitope prediction result</p>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.CleavageSitePredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">CleavageSitePredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageSitePredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageSitePredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Epitope prediction result</p>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.EpitopePredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">EpitopePredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#EpitopePredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.EpitopePredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>A Result object is a DataFrame with with multi-indexing, where column Id are the prediction model (i.e HLA allele
for epitope prediction), row ID the target of the prediction (i.e. peptide) and the second row ID the predictor
(i.e BIMAS)</p>
<p>Epitope prediction result</p>
<dl class="staticmethod">
<dt id="Fred2.Core.Result.EpitopePredictionResult.filter_result">
<em class="property">static </em><code class="descname">filter_result</code><span class="sig-paren">(</span><em>_filter_criteria</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#EpitopePredictionResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.EpitopePredictionResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list triples (Prediction_Method,Threshold,Sign) where Prediction_Method is the prediction methods
name to filter for, Threshold the value an epitope as to exceed/</p>
</dd></dl>

<dl class="staticmethod">
<dt id="Fred2.Core.Result.EpitopePredictionResult.merge_result">
<em class="property">static </em><code class="descname">merge_result</code><span class="sig-paren">(</span><em>_epi_pred_result</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#EpitopePredictionResult.merge_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.EpitopePredictionResult.merge_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a epitope result object and merges the two and returns the merge
result object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Param:</th><td class="field-body">(EpitopePredictionResult) _epi_pred_result: the result object to be merged</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">EpitopePredictionResult - the merged EpitopePredictionResult object</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raise:</th><td class="field-body">ValueError - if _epi_pred_result is not of type EpitopePredictionResult</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.Result">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">Result</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#Result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.Result" title="Permalink to this definition">¶</a></dt>
<dd><p>A Result object is a DataFrame with with multi-indexing.</p>
<p>This class is used as interface and cann be extended with custom short-cuts for the sometimes often tedious
calls in pandas</p>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.TAPPredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">TAPPredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#TAPPredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.TAPPredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Epitope prediction result</p>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Transcript">
<span id="fred2-core-transcript"></span><h2>Fred2.Core.Transcript<a class="headerlink" href="#module-Fred2.Core.Transcript" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Transcript"></span><dl class="class">
<dt id="Fred2.Core.Transcript.Transcript">
<em class="property">class </em><code class="descclassname">Fred2.Core.Transcript.</code><code class="descname">Transcript</code><span class="sig-paren">(</span><em>_gene_id</em>, <em>_transcript_id</em>, <em>_seq</em>, <em>_vars=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Transcript.html#Transcript"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Transcript.Transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>A Transcript is the mRNA sequence containing at no or several variations.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For accessing and manipulating the sequence see also <code class="xref py py-mod docutils literal"><span class="pre">Bio.Seq.Seq</span></code>
(from Biopython)</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gene_id</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Genome ID</li>
<li><strong>transcript_id</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Transcript RefSeqID</li>
<li><strong>vars</strong> (<em>dict(int,Variant)</em>) &#8211; Dict of Variants for specific positions in
the transcript. key=position, value=Variant</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Fred2.Core.Transcript.Transcript.translate">
<code class="descname">translate</code><span class="sig-paren">(</span><em>table='Standard'</em>, <em>stop_symbol='*'</em>, <em>to_stop=False</em>, <em>cds=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Transcript.html#Transcript.translate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Transcript.Transcript.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Overrides <code class="xref py py-meth docutils literal"><span class="pre">Bio.Seq.Seq.translate()</span></code> (from Biopython) and enables 
the translation from a transcript to a protein instance</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>returns</strong> &#8211; (Protein) &#8211; the protein that corresponds to the 
transcript</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Variant">
<span id="fred2-core-variant"></span><h2>Fred2.Core.Variant<a class="headerlink" href="#module-Fred2.Core.Variant" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Variant"></span><dl class="class">
<dt id="Fred2.Core.Variant.MutationSyntax">
<em class="property">class </em><code class="descclassname">Fred2.Core.Variant.</code><code class="descname">MutationSyntax</code><span class="sig-paren">(</span><em>transID</em>, <em>transPos</em>, <em>protPos</em>, <em>cds</em>, <em>aas</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#MutationSyntax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.MutationSyntax" title="Permalink to this definition">¶</a></dt>
<dd><p>This class represents the mutation syntax of a variant and stores its 
transcript and protein position</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>transID</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the transcript id</li>
<li><strong>transPos</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; the position of the variant within the transcript</li>
<li><strong>protPos</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; the protein position of the variant within the 
transcript</li>
<li><strong>cds</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the complete cds_mutation_syntax string</li>
<li><strong>aas</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; the complete protein_mutation_syntax string</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Variant.Variant">
<em class="property">class </em><code class="descclassname">Fred2.Core.Variant.</code><code class="descname">Variant</code><span class="sig-paren">(</span><em>id</em>, <em>type</em>, <em>chrom</em>, <em>genomePos</em>, <em>ref</em>, <em>obs</em>, <em>coding</em>, <em>isHomozygous</em>, <em>isSynonymous</em>, <em>experimentalDesign=None</em>, <em>metadata=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a></p>
<p>A Variant contains information about a single genetic modification of
the reference genome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>id</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Variant id</li>
<li><strong>type</strong> (<em>Enum.VariationType</em>) &#8211; An Enum type of the variant either SNP, 
DEL, or INS</li>
<li><strong>chrom</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; The chromosome on which the variant lies</li>
<li><strong>genomePos</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; The genomic position of the variant</li>
<li><strong>ref</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; The reference seq at the genomic position</li>
<li><strong>obs</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; The observed variation at the genomic position</li>
<li><strong>coding</strong> (<em>dict(str,MutationSyntax)</em>) &#8211; A dictionary of associated 
transcripts. Key=transcript_id, 
value=MutationSyntax</li>
<li><strong>isHomozygous</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) &#8211; defines if variant is homozygous or not</li>
<li><strong>isSynonymous</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#bool" title="(in Python v2.7)"><em>bool</em></a>) &#8211; defines if variant is a synonymous mutation 
or not</li>
<li><strong>offsets</strong> (<em>dict(str,int)</em>) &#8211; the position offset this variant has in a 
specific transcript-variant
key=transcript-variant-id (xxx:FRED2_nn)
value=offset</li>
<li><strong>metadata</strong> (<em>defaultdict(list)</em>) &#8211; meta information (not relevant for core
functionality of Fred2)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Fred2.Core.Variant.Variant.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>id</em>, <em>type</em>, <em>chrom</em>, <em>genomePos</em>, <em>ref</em>, <em>obs</em>, <em>coding</em>, <em>isHomozygous</em>, <em>isSynonymous</em>, <em>experimentalDesign=None</em>, <em>metadata=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructor for a variant, see init-types in class parameters</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Variant.Variant.get_protein_position">
<code class="descname">get_protein_position</code><span class="sig-paren">(</span><em>trans_variant_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.get_protein_position"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.get_protein_position" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">May only be used for transcript variants that were created from
this variant via 
<code class="xref py py-func docutils literal"><span class="pre">Generator.generate_transcripts_from_variants()</span></code>.</p>
</div>
<p>returns the specific protein position of a given transcript_id. If 
variant is not associated with the given transcript id the function 
throws a KeyError</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>transcriptId</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; A transcript_id</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">(int) &#8211; the protein position of the variant</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body">KeyError</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Variant.Variant.get_transcript_position">
<code class="descname">get_transcript_position</code><span class="sig-paren">(</span><em>trans_variant_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.get_transcript_position"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.get_transcript_position" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">May only be used for transcript variants that were created from
this variant via 
<code class="xref py py-func docutils literal"><span class="pre">Generator.generate_transcripts_from_variants()</span></code>.</p>
</div>
<p>returns the specific transcript position of a given transcript_id. 
If variant is not associated with the given transcript id the function 
throws a KeyError</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>transcriptId</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; A transcript_id</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">(int) &#8211; transcript position</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body">KeyError</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Variant.VariationType">
<code class="descclassname">Fred2.Core.Variant.</code><code class="descname">VariationType</code><a class="headerlink" href="#Fred2.Core.Variant.VariationType" title="Permalink to this definition">¶</a></dt>
<dd><p>Enum for variation types:
type.SNP, type.DEL, type.INS, type.FSDEL, type.FSINS, type.UNKNOWN</p>
<p>alias of <code class="xref py py-class docutils literal"><span class="pre">Enum</span></code></p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Fred2.Core</a><ul>
<li><a class="reference internal" href="#module-Fred2.Core.Allele">Fred2.Core.Allele</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Base">Fred2.Core.Base</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Generator">Fred2.Core.Generator</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Peptide">Fred2.Core.Peptide</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Protein">Fred2.Core.Protein</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Result">Fred2.Core.Result</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Transcript">Fred2.Core.Transcript</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Variant">Fred2.Core.Variant</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Basis modules</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Fred2.IO.html"
                        title="next chapter">Fred2.IO</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Fred2.Core.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Fred2.IO.html" title="Fred2.IO"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Basis modules"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">FRED 2.01 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Schbert Walzer Brachvogel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>