<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Fred2.IO package &mdash; Fred2  documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Fred2  documentation" href="index.html" />
    <link rel="up" title="Fred2 package" href="Fred2.html" />
    <link rel="next" title="Fred2.TAPPrediction package" href="Fred2.TAPPrediction.html" />
    <link rel="prev" title="Fred2.EpitopeSelection package" href="Fred2.EpitopeSelection.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Fred2.TAPPrediction.html" title="Fred2.TAPPrediction package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Fred2.EpitopeSelection.html" title="Fred2.EpitopeSelection package"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Fred2  documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="Fred2.html" accesskey="U">Fred2 package</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="fred2-io-package">
<h1>Fred2.IO package<a class="headerlink" href="#fred2-io-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-Fred2.IO.ADBAdapter">
<span id="fred2-io-adbadapter-module"></span><h2>Fred2.IO.ADBAdapter module<a class="headerlink" href="#module-Fred2.IO.ADBAdapter" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-IO.ADBAdapter"></span><dl class="class">
<dt id="Fred2.IO.ADBAdapter.ADBAdapter">
<em class="property">class </em><code class="descclassname">Fred2.IO.ADBAdapter.</code><code class="descname">ADBAdapter</code><a class="reference internal" href="_modules/Fred2/IO/ADBAdapter.html#ADBAdapter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.ADBAdapter.ADBAdapter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="Fred2.IO.ADBAdapter.ADBAdapter.get_product_sequence">
<code class="descname">get_product_sequence</code><span class="sig-paren">(</span><em>product_refseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/ADBAdapter.html#ADBAdapter.get_product_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.ADBAdapter.ADBAdapter.get_product_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>fetches product sequence for the given id</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>product_refseq</strong> &#8211; given refseq id</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of dictionaries of the requested sequence, the respective strand and the associated gene name</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.ADBAdapter.ADBAdapter.get_transcript_information">
<code class="descname">get_transcript_information</code><span class="sig-paren">(</span><em>transcript_refseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/ADBAdapter.html#ADBAdapter.get_transcript_information"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.ADBAdapter.ADBAdapter.get_transcript_information" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.IO.ADBAdapter.ADBAdapter.get_transcript_sequence">
<code class="descname">get_transcript_sequence</code><span class="sig-paren">(</span><em>transcript_refseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/ADBAdapter.html#ADBAdapter.get_transcript_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.ADBAdapter.ADBAdapter.get_transcript_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches transcript sequence for the given id</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>transcript_refseq</strong> &#8211; </td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of dictionary of the requested sequence, the respective strand and the associated gene name</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="Fred2.IO.ADBAdapter.EAdapterFields">
<code class="descclassname">Fred2.IO.ADBAdapter.</code><code class="descname">EAdapterFields</code><a class="headerlink" href="#Fred2.IO.ADBAdapter.EAdapterFields" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Enum</span></code></p>
</dd></dl>

</div>
<div class="section" id="module-Fred2.IO.EnsemblAdapter">
<span id="fred2-io-ensembladapter-module"></span><h2>Fred2.IO.EnsemblAdapter module<a class="headerlink" href="#module-Fred2.IO.EnsemblAdapter" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB">
<em class="property">class </em><code class="descclassname">Fred2.IO.EnsemblAdapter.</code><code class="descname">EnsemblDB</code><span class="sig-paren">(</span><em>name='fdb'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.exists">
<code class="descname">exists</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.exists" title="Permalink to this definition">¶</a></dt>
<dd><p>fast check if given sequence exists (as subsequence) in one of the EnsembleDB objects collection of sequences.
:param seq: the subsequence to be searched for
:return: True, if it is found somewhere, False otherwise</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.get_sequence">
<code class="descname">get_sequence</code><span class="sig-paren">(</span><em>ensp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.get_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.get_sequence" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.map_ensg">
<code class="descname">map_ensg</code><span class="sig-paren">(</span><em>ensg</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.map_ensg"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.map_ensg" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.map_ensp">
<code class="descname">map_ensp</code><span class="sig-paren">(</span><em>ensp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.map_ensp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.map_ensp" title="Permalink to this definition">¶</a></dt>
<dd><p>looks up ensp from the mapping and returns a ensg and a enst
:param ensp: the protein to map
:return: tuple of gene and protein (should not be None)</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.map_enst">
<code class="descname">map_enst</code><span class="sig-paren">(</span><em>enst</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.map_enst"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.map_enst" title="Permalink to this definition">¶</a></dt>
<dd><p>looks up enst from the mapping and returns a ensg and a ensp
:param enst: the transcript to map
:return: tuple of gene and protein (might be None)</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.read_seqs">
<code class="descname">read_seqs</code><span class="sig-paren">(</span><em>sequence_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.read_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.read_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>read sequences from Ensemble protein files (.fasta) or from lists or dicts of BioPython SeqRecords
and make them available for fast search. Appending also with this function.
:param sequence_file: Ensemble files (.dat or .fasta)
:return:</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.search">
<code class="descname">search</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.search" title="Permalink to this definition">¶</a></dt>
<dd><p>search for first occurrence of given sequence(s) in the EnsembleDB objects collection returning (each) the fasta
header front part of the first occurrence.
:param seq: a string interpreted as a single sequence or a list (of str) interpreted as a coll. of sequences
:return: a dictionary of sequences to lists (of ids, &#8216;null&#8217; if n/a)</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.search_all">
<code class="descname">search_all</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.search_all"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.search_all" title="Permalink to this definition">¶</a></dt>
<dd><p>search for all occurrences of given sequence(s) in the EnsembleDB objects collection returning (each) the
fasta header front part of all occurrences.
:param seq: a string interpreted as a single sequence or a list (of str) interpreted as a coll. of sequences
:return: a dictionary of the given sequences to lists (of ids, &#8216;null&#8217; if n/a)</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.EnsemblAdapter.EnsemblDB.write_seqs">
<code class="descname">write_seqs</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/EnsemblAdapter.html#EnsemblDB.write_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.EnsemblAdapter.EnsemblDB.write_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>writes all fasta entries in the current object into one fasta file
:param name: the complete path with file name where the fasta is going to be written</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.IO.FileReader">
<span id="fred2-io-filereader-module"></span><h2>Fred2.IO.FileReader module<a class="headerlink" href="#module-Fred2.IO.FileReader" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Reader"></span><dl class="function">
<dt id="Fred2.IO.FileReader.read_annovar_exonic">
<code class="descclassname">Fred2.IO.FileReader.</code><code class="descname">read_annovar_exonic</code><span class="sig-paren">(</span><em>annovar_file</em>, <em>gene_filter=None</em>, <em>experimentalDesig=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/FileReader.html#read_annovar_exonic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.FileReader.read_annovar_exonic" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads an gene-based ANNOVAR output file and generates Variant objects containing
all annotated transcript ids an outputs a list variants.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>annovar_file</strong> (<em>str</em>) &#8211; The path ot the ANNOVAR file</li>
<li><strong>gene_filter</strong> (<em>list(str)</em>) &#8211; A list of gene names of interest (only variants associated
with these genes are generated)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list(AVariant) &#8211; List of variants fully annotated</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.IO.FileReader.read_fasta">
<code class="descclassname">Fred2.IO.FileReader.</code><code class="descname">read_fasta</code><span class="sig-paren">(</span><em>files</em>, <em>type=&lt;class 'Fred2.Core.Peptide.Peptide'&gt;</em>, <em>id_position=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/FileReader.html#read_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.FileReader.read_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator function:</p>
<p>Read a (couple of) peptide, protein or rna sequence from a FASTA file.
User needs to specify the correct type of the underlying sequences. It can
either be: Peptide, Protein or Transcript (for RNA).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>files</strong> (<em>List(str)/str</em>) &#8211; A (List) of file names to read in</li>
<li><strong>type</strong> (<em>Peptide/Transcript/Protein</em>) &#8211; The type to read in</li>
<li><strong>id_position</strong> (<em>int</em>) &#8211; the position of the id specified counted by |</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">(list(SequenceType)) &#8211; a list of the specified sequence type
derived from the FASTA file sequences.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Raises ValueError:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><p class="first last">if a file is not readable</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.IO.FileReader.read_lines">
<code class="descclassname">Fred2.IO.FileReader.</code><code class="descname">read_lines</code><span class="sig-paren">(</span><em>files</em>, <em>type=&lt;class 'Fred2.Core.Peptide.Peptide'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/FileReader.html#read_lines"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.FileReader.read_lines" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator function:</p>
<p>Read a sequence directly from a line. User needs to manually specify the 
correct type of the underlying data. It can either be:
Peptide, Protein or Transcript, Allele.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>files</strong> (<em>list(str)/str</em>) &#8211; a list of strings of absolute file names of the FASTA files
that are to be read. Give as: 
get_sequence(<a href="#id1"><span class="problematic" id="id2">*</span></a>[&#8220;path/name1&#8221;, &#8220;path/name2&#8221;]).
This field is required!</li>
<li><strong>type</strong> (<em>Peptide/Protein/Transcript/Allele</em>) &#8211; optional. Use get_sequence(<a href="#id3"><span class="problematic" id="id4">*</span></a>[&#8220;path/name1&#8221;, &#8220;path/name2&#8221;,
type=&#8221;Protein&#8221;). Possible types are Peptide, Protein, Transcript and Allele.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">(list(type)) &#8211; a list of the specified objects</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Exception:</th><td class="field-body"><p class="first last">if a file is not readable</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.IO.MartsAdapter">
<span id="fred2-io-martsadapter-module"></span><h2>Fred2.IO.MartsAdapter module<a class="headerlink" href="#module-Fred2.IO.MartsAdapter" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-IO.MartsAdapter"></span><dl class="class">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter">
<em class="property">class </em><code class="descclassname">Fred2.IO.MartsAdapter.</code><code class="descname">MartsAdapter</code><span class="sig-paren">(</span><em>usr=None</em>, <em>host=None</em>, <em>pwd=None</em>, <em>db=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.IO.ADBAdapter.ADBAdapter" title="Fred2.IO.ADBAdapter.ADBAdapter"><code class="xref py py-class docutils literal"><span class="pre">Fred2.IO.ADBAdapter.ADBAdapter</span></code></a></p>
<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_all_variant_gene">
<code class="descname">get_all_variant_gene</code><span class="sig-paren">(</span><em>locations</em>, <em>_db='hsapiens_gene_ensembl'</em>, <em>_dataset='gene_ensembl_config'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_all_variant_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_all_variant_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches the important db ids and names for given chromosomal location
:param chrom: integer value of the chromosome in question
:param start: integer value of the variation start position on given chromosome
:param stop: integer value of the variation stop position on given chromosome
:return: The respective gene name, i.e. the first one reported</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_all_variant_ids">
<code class="descname">get_all_variant_ids</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_all_variant_ids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_all_variant_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches the important db ids and names for given gene _or_ chromosomal location. The former is recommended.
AResult is a list of dicts with either of the tree combinations:</p>
<blockquote>
<div><ul class="simple">
<li>&#8216;Ensembl Gene ID&#8217;, &#8216;Ensembl Transcript ID&#8217;, &#8216;Ensembl Protein ID&#8217;</li>
<li>&#8216;RefSeq Protein ID [e.g. NP_001005353]&#8217;, &#8216;RefSeq mRNA [e.g. NM_001195597]&#8217;, first triplet</li>
<li>&#8216;RefSeq Predicted Protein ID [e.g. XP_001720922]&#8217;, &#8216;RefSeq mRNA predicted [e.g. XM_001125684]&#8217;, first triplet</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>'locations'</strong> &#8211; list of locations as triplets of integer values representing (chrom, start, stop)</li>
<li><strong>'genes'</strong> &#8211; list of genes as string value of the genes of variation</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The list of dicts of entries with transcript and protein ids (either NM+NP or XM+XP)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_product_sequence">
<code class="descname">get_product_sequence</code><span class="sig-paren">(</span><em>product_refseq</em>, <em>_db='hsapiens_gene_ensembl'</em>, <em>_dataset='gene_ensembl_config'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_product_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_product_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>fetches product sequence for the given id
:param product_refseq: given refseq id
:return: list of dictionaries of the requested sequence, the respective strand and the associated gene name</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_protein_sequence_from_protein_id">
<code class="descname">get_protein_sequence_from_protein_id</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_protein_sequence_from_protein_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_protein_sequence_from_protein_id" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the protein sequence for a given protein ID that can either be refeseq, uniprot or ensamble id</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>kwargs</strong> &#8211; </td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_information">
<code class="descname">get_transcript_information</code><span class="sig-paren">(</span><em>transcript_refseq</em>, <em>_db='hsapiens_gene_ensembl'</em>, <em>_dataset='gene_ensembl_config'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_transcript_information"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_information" title="Permalink to this definition">¶</a></dt>
<dd><p>It also already uses the Field-Enum for DBAdapters</p>
<p>Fetches transcript sequence for the given id
:param transcript_refseq:
:return: list of dictionary of the requested sequence, the respective strand and the associated gene name</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_information_from_protein_id">
<code class="descname">get_transcript_information_from_protein_id</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_transcript_information_from_protein_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_information_from_protein_id" title="Permalink to this definition">¶</a></dt>
<dd><p>It also already uses the Field-Enum for DBAdapters</p>
<p>Fetches transcript sequence for the given id
:param transcript_refseq:
:return: list of dictionary of the requested sequence, the respective strand and the associated gene name</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_sequence">
<code class="descname">get_transcript_sequence</code><span class="sig-paren">(</span><em>transcript_refseq</em>, <em>_db='hsapiens_gene_ensembl'</em>, <em>_dataset='gene_ensembl_config'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_transcript_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches transcript sequence for the given id
:param transcript_refseq:
:return: list of dictionary of the requested sequence, the respective strand and the associated gene name</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_variant_gene">
<code class="descname">get_variant_gene</code><span class="sig-paren">(</span><em>chrom</em>, <em>start</em>, <em>stop</em>, <em>_db='hsapiens_gene_ensembl'</em>, <em>_dataset='gene_ensembl_config'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_variant_gene"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_gene" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches the important db ids and names for given chromosomal location
:param chrom: integer value of the chromosome in question
:param start: integer value of the variation start position on given chromosome
:param stop: integer value of the variation stop position on given chromosome
:return: The respective gene name, i.e. the first one reported</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_variant_id_from_gene_id">
<code class="descname">get_variant_id_from_gene_id</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_variant_id_from_gene_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_id_from_gene_id" title="Permalink to this definition">¶</a></dt>
<dd><p>returns all information needed to instantiate a variation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>trans_id</strong> &#8211; A transcript ID (either ENSAMBLE (ENS) or RefSeq (NM, XN)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of dicts &#8211; containing all information needed for a variant initialization</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_variant_id_from_protein_id">
<code class="descname">get_variant_id_from_protein_id</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_variant_id_from_protein_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_id_from_protein_id" title="Permalink to this definition">¶</a></dt>
<dd><p>returns all information needed to instantiate a variation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>trans_id</strong> &#8211; A transcript ID (either ENSAMBLE (ENS) or RefSeq (NM, XN)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list of dicts &#8211; containing all information needed for a variant initialization</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.MartsAdapter.MartsAdapter.get_variant_ids">
<code class="descname">get_variant_ids</code><span class="sig-paren">(</span><em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/MartsAdapter.html#MartsAdapter.get_variant_ids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches the important db ids and names for given gene _or_ chromosomal location. The former is recommended.
AResult is a list of dicts with either of the tree combinations:</p>
<blockquote>
<div><ul class="simple">
<li>&#8216;Ensembl Gene ID&#8217;, &#8216;Ensembl Transcript ID&#8217;, &#8216;Ensembl Protein ID&#8217;</li>
<li>&#8216;RefSeq Protein ID [e.g. NP_001005353]&#8217;, &#8216;RefSeq mRNA [e.g. NM_001195597]&#8217;, first triplet</li>
<li>&#8216;RefSeq Predicted Protein ID [e.g. XP_001720922]&#8217;, &#8216;RefSeq mRNA predicted [e.g. XM_001125684]&#8217;, first triplet</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>'chrom'</strong> &#8211; integer value of the chromosome in question</li>
<li><strong>'start'</strong> &#8211; integer value of the variation start position on given chromosome</li>
<li><strong>'stop'</strong> &#8211; integer value of the variation stop position on given chromosome</li>
<li><strong>'gene'</strong> &#8211; string value of the gene of variation</li>
<li><strong>'transcript_id'</strong> &#8211; string value of the gene of variation</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The list of dicts of entries with transcript and protein ids (either NM+NP or XM+XP)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.IO.RefSeqAdapter">
<span id="fred2-io-refseqadapter-module"></span><h2>Fred2.IO.RefSeqAdapter module<a class="headerlink" href="#module-Fred2.IO.RefSeqAdapter" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-IO.RefSeqAdapter"></span><dl class="class">
<dt id="Fred2.IO.RefSeqAdapter.RefSeqAdapter">
<em class="property">class </em><code class="descclassname">Fred2.IO.RefSeqAdapter.</code><code class="descname">RefSeqAdapter</code><span class="sig-paren">(</span><em>prot_file=None</em>, <em>prot_vers=None</em>, <em>mrna_file=None</em>, <em>mrna_vers=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/RefSeqAdapter.html#RefSeqAdapter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.RefSeqAdapter.RefSeqAdapter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.IO.ADBAdapter.ADBAdapter" title="Fred2.IO.ADBAdapter.ADBAdapter"><code class="xref py py-class docutils literal"><span class="pre">Fred2.IO.ADBAdapter.ADBAdapter</span></code></a></p>
<dl class="method">
<dt id="Fred2.IO.RefSeqAdapter.RefSeqAdapter.get_product_sequence">
<code class="descname">get_product_sequence</code><span class="sig-paren">(</span><em>product_refseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/RefSeqAdapter.html#RefSeqAdapter.get_product_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.RefSeqAdapter.RefSeqAdapter.get_product_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>fetches product sequence for the given id
:param product_refseq: given refseq id
:return: list of dictionaries of the requested sequence, the respective strand and the associated gene name</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.RefSeqAdapter.RefSeqAdapter.get_transcript_information">
<code class="descname">get_transcript_information</code><span class="sig-paren">(</span><em>transcript_refseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/RefSeqAdapter.html#RefSeqAdapter.get_transcript_information"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.RefSeqAdapter.RefSeqAdapter.get_transcript_information" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.IO.RefSeqAdapter.RefSeqAdapter.get_transcript_sequence">
<code class="descname">get_transcript_sequence</code><span class="sig-paren">(</span><em>transcript_refseq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/RefSeqAdapter.html#RefSeqAdapter.get_transcript_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.RefSeqAdapter.RefSeqAdapter.get_transcript_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches transcript sequence for the given id
:param transcript_refseq:
:return: list of dictionary of the requested sequence, the respective strand and the associated gene name</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.RefSeqAdapter.RefSeqAdapter.load">
<code class="descname">load</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/RefSeqAdapter.html#RefSeqAdapter.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.RefSeqAdapter.RefSeqAdapter.load" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.IO.UniProtAdapter">
<span id="fred2-io-uniprotadapter-module"></span><h2>Fred2.IO.UniProtAdapter module<a class="headerlink" href="#module-Fred2.IO.UniProtAdapter" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-IO.niProtAdapter"></span><dl class="class">
<dt id="Fred2.IO.UniProtAdapter.UniProtDB">
<em class="property">class </em><code class="descclassname">Fred2.IO.UniProtAdapter.</code><code class="descname">UniProtDB</code><span class="sig-paren">(</span><em>name='fdb'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/UniProtAdapter.html#UniProtDB"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.UniProtAdapter.UniProtDB" title="Permalink to this definition">¶</a></dt>
<dd><dl class="method">
<dt id="Fred2.IO.UniProtAdapter.UniProtDB.exists">
<code class="descname">exists</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/UniProtAdapter.html#UniProtDB.exists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.UniProtAdapter.UniProtDB.exists" title="Permalink to this definition">¶</a></dt>
<dd><p>fast check if given sequence exists (as subsequence) in one of the UniProtDB objects collection of sequences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> &#8211; the subsequence to be searched for</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">True, if it is found somewhere, False otherwise</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.UniProtAdapter.UniProtDB.read_seqs">
<code class="descname">read_seqs</code><span class="sig-paren">(</span><em>sequence_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/UniProtAdapter.html#UniProtDB.read_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.UniProtAdapter.UniProtDB.read_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>read sequences from uniprot files (.dat or .fasta) or from lists or dicts of BioPython SeqRecords
and make them available for fast search. Appending also with this function.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sequence_file</strong> &#8211; uniprot files (.dat or .fasta)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.UniProtAdapter.UniProtDB.search">
<code class="descname">search</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/UniProtAdapter.html#UniProtDB.search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.UniProtAdapter.UniProtDB.search" title="Permalink to this definition">¶</a></dt>
<dd><p>search for first occurrence of given sequence(s) in the UniProtDB objects collection returning (each) the fasta
header front part of the first occurrence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> &#8211; a string interpreted as a single sequence or a list (of str) interpreted as a coll. of sequences</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a dictionary of sequences to lists (of ids, &#8216;null&#8217; if n/a)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.UniProtAdapter.UniProtDB.search_all">
<code class="descname">search_all</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/UniProtAdapter.html#UniProtDB.search_all"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.UniProtAdapter.UniProtDB.search_all" title="Permalink to this definition">¶</a></dt>
<dd><p>search for all occurrences of given sequence(s) in the UniProtDB objects collection returning (each) the
fasta header front part of all occurrences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> &#8211; a string interpreted as a single sequence or a list (of str) interpreted as a coll. of sequences</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a dictionary of the given sequences to lists (of ids, &#8216;null&#8217; if n/a)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.IO.UniProtAdapter.UniProtDB.write_seqs">
<code class="descname">write_seqs</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/IO/UniProtAdapter.html#UniProtDB.write_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.IO.UniProtAdapter.UniProtDB.write_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>writes all fasta entries in the current object into one fasta file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>name</strong> &#8211; the complete path with file name where the fasta is going to be written</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.IO">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Fred2.IO" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Fred2.IO package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-Fred2.IO.ADBAdapter">Fred2.IO.ADBAdapter module</a></li>
<li><a class="reference internal" href="#module-Fred2.IO.EnsemblAdapter">Fred2.IO.EnsemblAdapter module</a></li>
<li><a class="reference internal" href="#module-Fred2.IO.FileReader">Fred2.IO.FileReader module</a></li>
<li><a class="reference internal" href="#module-Fred2.IO.MartsAdapter">Fred2.IO.MartsAdapter module</a></li>
<li><a class="reference internal" href="#module-Fred2.IO.RefSeqAdapter">Fred2.IO.RefSeqAdapter module</a></li>
<li><a class="reference internal" href="#module-Fred2.IO.UniProtAdapter">Fred2.IO.UniProtAdapter module</a></li>
<li><a class="reference internal" href="#module-Fred2.IO">Module contents</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Fred2.EpitopeSelection.html"
                        title="previous chapter">Fred2.EpitopeSelection package</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Fred2.TAPPrediction.html"
                        title="next chapter">Fred2.TAPPrediction package</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Fred2.IO.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Author.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.2</a>
      
      |
      <a href="_sources/Fred2.IO.txt"
          rel="nofollow">Page source</a></li>
    </div>

    

    
  </body>
</html>