<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Fred2.Core package &mdash; Fred2  documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Fred2  documentation" href="index.html" />
    <link rel="up" title="Fred2 package" href="Fred2.html" />
    <link rel="next" title="Fred2.Distance2Self package" href="Fred2.Distance2Self.html" />
    <link rel="prev" title="Fred2.CleavagePrediction package" href="Fred2.CleavagePrediction.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Fred2.Distance2Self.html" title="Fred2.Distance2Self package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Fred2.CleavagePrediction.html" title="Fred2.CleavagePrediction package"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Fred2  documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="Fred2.html" accesskey="U">Fred2 package</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="fred2-core-package">
<h1>Fred2.Core package<a class="headerlink" href="#fred2-core-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-Fred2.Core.Allele">
<span id="fred2-core-allele-module"></span><h2>Fred2.Core.Allele module<a class="headerlink" href="#module-Fred2.Core.Allele" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Allele"></span><dl class="class">
<dt id="Fred2.Core.Allele.Allele">
<em class="property">class </em><code class="descclassname">Fred2.Core.Allele.</code><code class="descname">Allele</code><span class="sig-paren">(</span><em>_name</em>, <em>prob=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Allele.html#Allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Allele.Allele" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a></p>
<p>This class represents an Allele and stores additional information as a
dictionary</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>name</strong> (<em>str</em>) &#8211; the name of the MHC allele (new nomenclature A*01:01)</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Base">
<span id="fred2-core-base-module"></span><h2>Fred2.Core.Base module<a class="headerlink" href="#module-Fred2.Core.Base" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Base"></span><p><a class="reference external" href="https://docs.python.org/3/library/abc.html">https://docs.python.org/3/library/abc.html</a></p>
<dl class="class">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">ACleavageFragmentPrediction</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#ACleavageFragmentPrediction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction.cleavagePos">
<code class="descname">cleavagePos</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction.cleavagePos" title="Permalink to this definition">¶</a></dt>
<dd><p>parameter specifying the position of aa (within the prediction window) after which the sequence is cleaved</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction.name">
<code class="descname">name</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the name of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>_aa_seq</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#ACleavageFragmentPrediction.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts the probability that the fragment can be produced by the proteasom</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>_aa_seq</strong> (<em>Bio.Seq</em>) &#8211; The sequence to be cleaved (must be an instance of Bio.Seq</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Returns a AResult object for the specified Bio.Seq</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction.registry">
<code class="descname">registry</code><em class="property"> = {'ginodi': {'1.0': &lt;class 'Fred2.CleavagePrediction.PSSM.PSSMGinodi'&gt;}}</em><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction.registry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction.supportedLength">
<code class="descname">supportedLength</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction.supportedLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the supported lengths of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list(int) - Supported peptide length</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageFragmentPrediction.version">
<code class="descname">version</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageFragmentPrediction.version" title="Permalink to this definition">¶</a></dt>
<dd><p>parameter specifying the version of the prediction method</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.ACleavageSitePrediction">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">ACleavageSitePrediction</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#ACleavageSitePrediction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageSitePrediction.cleavagePos">
<code class="descname">cleavagePos</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction.cleavagePos" title="Permalink to this definition">¶</a></dt>
<dd><p>parameter specifying the position of aa (within the prediction window) after which the sequence is cleaved
(starting from 1)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageSitePrediction.name">
<code class="descname">name</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the name of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.ACleavageSitePrediction.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>_aa_seq</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#ACleavageSitePrediction.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts the proteom cleavage site of the given sequences</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>_aa_seq</strong> (<em>Bio.Seq</em>) &#8211; The sequence to be cleaved (must be an instance of Bio.Seq</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Returns a AResult object for the specified Bio.Seq</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageSitePrediction.registry">
<code class="descname">registry</code><em class="property"> = {'proteasmm_i': {'1.0': &lt;class 'Fred2.CleavagePrediction.PSSM.ProteaSMMImmuno'&gt;}, 'netchop': {'3.1': &lt;class 'Fred2.CleavagePrediction.External.NetChop_3_1'&gt;}, 'proteasmm_c': {'1.0': &lt;class 'Fred2.CleavagePrediction.PSSM.ProteaSMMConsecutive'&gt;}, 'pcm': {'1.0': &lt;class 'Fred2.CleavagePrediction.PSSM.PCM'&gt;}}</em><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction.registry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageSitePrediction.supportedLength">
<code class="descname">supportedLength</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction.supportedLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the supported lengths of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list(int) - Supported peptide length</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ACleavageSitePrediction.version">
<code class="descname">version</code><a class="headerlink" href="#Fred2.Core.Base.ACleavageSitePrediction.version" title="Permalink to this definition">¶</a></dt>
<dd><p>parameter specifying the version of the prediction method</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.AEpitopePrediction">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">AEpitopePrediction</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#AEpitopePrediction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="Fred2.Core.Base.AEpitopePrediction.convert_alleles">
<code class="descname">convert_alleles</code><span class="sig-paren">(</span><em>alleles</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AEpitopePrediction.convert_alleles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.convert_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts alleles into the interal allele representation of the predictor
and returns a string representation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>alleles</strong> (<em>list(Allele)</em>) &#8211; The alleles for which the internal predictor
representation is needed</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Returns a string representation of the input alleles</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.AEpitopePrediction.name">
<code class="descname">name</code><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the name of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.AEpitopePrediction.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>peptides</em>, <em>alleles=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AEpitopePrediction.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts the binding affinity for a given peptide or peptide lists for a given list of alleles.
If alleles is not given, predictions for all valid alleles of the predictor is performed. If, however,
a list of alleles is given, predictions for the valid allele subset is performed.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>peptides</strong> (<em>Peptide/list(Peptide)</em>) &#8211; The peptide objects for which predictions should be performed</li>
<li><strong>alleles</strong> (<em>Allele/list(Allele)</em>) &#8211; An Allele or list of Alleles for which prediction models should be used</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Returns a AResult object for the specified Peptides and Alleles</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.AEpitopePrediction.registry">
<code class="descname">registry</code><em class="property"> = {}</em><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.registry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.AEpitopePrediction.supportedAlleles">
<code class="descname">supportedAlleles</code><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.supportedAlleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of valid allele models</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">List of allele names for which the predictor provides models</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.AEpitopePrediction.supportedLength">
<code class="descname">supportedLength</code><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.supportedLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of supported peptide lenghts</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">List of supported peptide lengths</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.AEpitopePrediction.version">
<code class="descname">version</code><a class="headerlink" href="#Fred2.Core.Base.AEpitopePrediction.version" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.AExternal">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">AExternal</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Base class for external tools</p>
<dl class="attribute">
<dt id="Fred2.Core.Base.AExternal.command">
<code class="descname">command</code><a class="headerlink" href="#Fred2.Core.Base.AExternal.command" title="Permalink to this definition">¶</a></dt>
<dd><p>defines the external execution path ?
:return:</p>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.AExternal.get_external_version">
<code class="descname">get_external_version</code><span class="sig-paren">(</span><em>path=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal.get_external_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal.get_external_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the external version of the tool by executing
&gt;{command} &#8211;version</p>
<p>might be dependent on the method and has to be overwritten
therefore it is declared abstract to enforce the user to
overwrite the method. The function in the base class can be called
with super()</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<em>(str)</em>) &#8211; <ul class="simple">
<li>optional specification of executable path if deviant from self.__command</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">str - The external version of the tool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.AExternal.is_in_path">
<code class="descname">is_in_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal.is_in_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal.is_in_path" title="Permalink to this definition">¶</a></dt>
<dd><p>checks whether the specified execution command can be found in PATH</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">bool - Whether or not command could be found in PATH</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.AExternal.parse_external_result">
<code class="descname">parse_external_result</code><span class="sig-paren">(</span><em>_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal.parse_external_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal.parse_external_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses external NetMHC results and returns a AResult object</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>_file</strong> (<em>str</em>) &#8211; The file path or the external prediction results</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">AResult - Returns a AResult object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.AExternal.prepare_peptide_input">
<code class="descname">prepare_peptide_input</code><span class="sig-paren">(</span><em>_peptides</em>, <em>_file</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#AExternal.prepare_peptide_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.AExternal.prepare_peptide_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepares sequence input for external tools
and writes them to _file in the specific format</p>
<p>NO return value!</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Param:</th><td class="field-body">(list(str)) _peptides: the peptide sequences to write into _file</td>
</tr>
<tr class="field-even field"><th class="field-name">Parameters:</th><td class="field-body"><strong>_file</strong> (<em>(File)</em>) &#8211; File handler to input file for external tool</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.APluginRegister">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">APluginRegister</code><span class="sig-paren">(</span><em>name</em>, <em>bases</em>, <em>nmspc</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#APluginRegister"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.APluginRegister" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">abc.ABCMeta</span></code></p>
<p>This class allows automatic registration of new plugins.</p>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.ASVM">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">ASVM</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#ASVM"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ASVM" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Base class for SVM prediction tools</p>
<dl class="method">
<dt id="Fred2.Core.Base.ASVM.encode">
<code class="descname">encode</code><span class="sig-paren">(</span><em>peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#ASVM.encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ASVM.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the feature encoding for peptides</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>peptides</strong> (<em>List(Peptide)/Peptide</em>) &#8211; List oder Peptide object</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">list(Object) &#8211; Feature encoding of the Peptide objects</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.ATAPPrediction">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">ATAPPrediction</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#ATAPPrediction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ATAPPrediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="attribute">
<dt id="Fred2.Core.Base.ATAPPrediction.name">
<code class="descname">name</code><a class="headerlink" href="#Fred2.Core.Base.ATAPPrediction.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the name of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.ATAPPrediction.predict">
<code class="descname">predict</code><span class="sig-paren">(</span><em>peptides</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#ATAPPrediction.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.ATAPPrediction.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predicts the TAP affinity for the given sequences</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>list(Peptide)/Peptide</strong> &#8211; Peptides for which TAP affinity should be predicted</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Returns a TAPResult object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ATAPPrediction.registry">
<code class="descname">registry</code><em class="property"> = {}</em><a class="headerlink" href="#Fred2.Core.Base.ATAPPrediction.registry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ATAPPrediction.supportedLength">
<code class="descname">supportedLength</code><a class="headerlink" href="#Fred2.Core.Base.ATAPPrediction.supportedLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the supported lengths of the predictor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">list(int) - Supported peptide length</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Base.ATAPPrediction.version">
<code class="descname">version</code><a class="headerlink" href="#Fred2.Core.Base.ATAPPrediction.version" title="Permalink to this definition">¶</a></dt>
<dd><p>parameter specifying the version of the prediction method</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Base.MetadataLogger">
<em class="property">class </em><code class="descclassname">Fred2.Core.Base.</code><code class="descname">MetadataLogger</code><a class="reference internal" href="_modules/Fred2/Core/Base.html#MetadataLogger"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.MetadataLogger" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>This class provides a simple interface for assigning additional metadata to
any object in our data model. Examples: storing ANNOVAR columns like depth,
base count, dbSNP id, quality information for variants, prediction scores 
for peptides etc. Normally used by custom toolbox functions and importers.</p>
<p>The saved values are accessed via <code class="xref py py-meth docutils literal"><span class="pre">log_metadata()</span></code> and
<code class="xref py py-meth docutils literal"><span class="pre">get_metadata()</span></code></p>
<dl class="method">
<dt id="Fred2.Core.Base.MetadataLogger.get_metadata">
<code class="descname">get_metadata</code><span class="sig-paren">(</span><em>label</em>, <em>only_first=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#MetadataLogger.get_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.MetadataLogger.get_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Getter for the saved metadata with the key <code class="xref py py-attr docutils literal"><span class="pre">label</span></code></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>label</strong> (<em>str</em>) &#8211; key for the metadata that is inferred</li>
<li><strong>only_first</strong> (<em>bool</em>) &#8211; true if only the the first element of the 
matadata list is to be returned</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Base.MetadataLogger.log_metadata">
<code class="descname">log_metadata</code><span class="sig-paren">(</span><em>label</em>, <em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Base.html#MetadataLogger.log_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Base.MetadataLogger.log_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Inserts a new metadata</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>label</strong> (<em>str</em>) &#8211; key for the metadata that will be added</li>
<li><strong>value</strong> (<em>list(object)</em>) &#8211; any kindy of additional value that should be
kept</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Generator">
<span id="fred2-core-generator-module"></span><h2>Fred2.Core.Generator module<a class="headerlink" href="#module-Fred2.Core.Generator" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Generator"></span><dl class="function">
<dt id="Fred2.Core.Generator.generate_peptides_from_protein">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_peptides_from_protein</code><span class="sig-paren">(</span><em>proteins</em>, <em>window_size</em>, <em>peptides=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_peptides_from_protein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_peptides_from_protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates all peptides for a given window size, from a given protein. The
result is a generator.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>proteins</strong> (<a class="reference internal" href="#Fred2.Core.Protein.Protein" title="Fred2.Core.Protein.Protein"><em>Protein</em></a>) &#8211; (list of) protein(s) from which a list of unique
peptides should be generated</li>
<li><strong>window_size</strong> (<em>int</em>) &#8211; size of peptide fragments</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_peptides_from_variants">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_peptides_from_variants</code><span class="sig-paren">(</span><em>vars</em>, <em>length</em>, <em>dbadapter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_peptides_from_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_peptides_from_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>generates polymorphic peptides based on given variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>vars</strong> (<em>list(Variant)</em>) &#8211; list of Variants</li>
<li><strong>length</strong> (<em>int</em>) &#8211; length of peptides</li>
<li><strong>dbadapter</strong> &#8211; DBAdapter to fetch the transcript sequences</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">List(Peptide) &#8211; A list of polymorphic peptides</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_proteins_from_transcripts">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_proteins_from_transcripts</code><span class="sig-paren">(</span><em>transcripts</em>, <em>table='Standard'</em>, <em>stop_symbol='*'</em>, <em>to_stop=False</em>, <em>cds=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_proteins_from_transcripts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_proteins_from_transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p>Enables the translation from a transcript to a protein instance</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Param:</th><td class="field-body">list(Transcript)/Transcript - a list of transcripts to translate to proteins</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">(Protein) &#8211; the protein that corresponds to the transcript</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_transcripts_from_tumor_variants">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_transcripts_from_tumor_variants</code><span class="sig-paren">(</span><em>normal</em>, <em>tumor</em>, <em>dbadapter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_transcripts_from_tumor_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_transcripts_from_tumor_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>generates all possible transcript variations of the given variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>normal</strong> (<em>list(Variant)</em>) &#8211; A list of variants of the normal tissue</li>
<li><strong>tumor</strong> (<em>list(Variant)</em>) &#8211; A list of variant of the cancer tissue for which transcript should be generated</li>
<li><strong>dbadapter</strong> (<a class="reference internal" href="Fred2.IO.html#module-IO.ADBAdapter" title="IO.ADBAdapter: Base class for specific DB-Adapters"><em>ADBAdapter</em></a>) &#8211; a DBAdapter to fetch the transcript sequences</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">(Generator(Transcripts)) &#8211; a generator of transcripts with all
possible variations determined by the given
variant list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Fred2.Core.Generator.generate_transcripts_from_variants">
<code class="descclassname">Fred2.Core.Generator.</code><code class="descname">generate_transcripts_from_variants</code><span class="sig-paren">(</span><em>vars</em>, <em>dbadapter</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Generator.html#generate_transcripts_from_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Generator.generate_transcripts_from_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>generates all possible transcript variations of the given variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>vars</strong> (<em>list(Variant)</em>) &#8211; A list of variants for which transcripts should 
be build</li>
<li><strong>dbadapter</strong> (<a class="reference internal" href="Fred2.IO.html#module-IO.ADBAdapter" title="IO.ADBAdapter: Base class for specific DB-Adapters"><em>ADBAdapter</em></a>) &#8211; a DBAdapter to fetch the transcript sequences</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">(Generator(Transcripts)) &#8211; a generator of transcripts with all 
possible variations determined by the given
variant list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Peptide">
<span id="fred2-core-peptide-module"></span><h2>Fred2.Core.Peptide module<a class="headerlink" href="#module-Fred2.Core.Peptide" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="Fred2.Core.Peptide.Peptide">
<em class="property">class </em><code class="descclassname">Fred2.Core.Peptide.</code><code class="descname">Peptide</code><span class="sig-paren">(</span><em>_seq</em>, <em>proteins=None</em>, <em>variants=None</em>, <em>transcripts=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Peptide.html#Peptide"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Peptide.Peptide" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a>, <code class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></code></p>
<p>This class encapsulates a Peptide, belonging to one or several Proteins.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For accessing and manipulating the sequence see also <code class="xref py py-mod docutils literal"><span class="pre">Bio.Seq.Seq</span></code> (from Biopython)</p>
</div>
<dl class="method">
<dt id="Fred2.Core.Peptide.Peptide.get_all_proteins">
<code class="descname">get_all_proteins</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Peptide.html#Peptide.get_all_proteins"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Peptide.Peptide.get_all_proteins" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.Core.Peptide.Peptide.get_all_transcripts">
<code class="descname">get_all_transcripts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Peptide.html#Peptide.get_all_transcripts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Peptide.Peptide.get_all_transcripts" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.Core.Peptide.Peptide.get_all_variants">
<code class="descname">get_all_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Peptide.html#Peptide.get_all_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Peptide.Peptide.get_all_variants" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a concatenated list of all contained variants</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Protein">
<span id="fred2-core-protein-module"></span><h2>Fred2.Core.Protein module<a class="headerlink" href="#module-Fred2.Core.Protein" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="Fred2.Core.Protein.Protein">
<em class="property">class </em><code class="descclassname">Fred2.Core.Protein.</code><code class="descname">Protein</code><span class="sig-paren">(</span><em>_seq</em>, <em>_gene_id='unknown'</em>, <em>_transcript_id=None</em>, <em>_orig_transcript=None</em>, <em>_vars=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Protein.html#Protein"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Protein.Protein" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a>, <code class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></code></p>
<p>Protein corresponding to exactly one transcript.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For accessing and manipulating the sequence see also <code class="xref py py-mod docutils literal"><span class="pre">Bio.Seq.Seq</span></code>
(from Biopython)</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gene_id</strong> (<em>str</em>) &#8211; ID of the genome</li>
<li><strong>transcript_id</strong> (<em>str</em>) &#8211; ID of the corresponding transcript</li>
<li><strong>orig_transcript</strong> (<a class="reference internal" href="#Fred2.Core.Transcript.Transcript" title="Fred2.Core.Transcript.Transcript"><em>Transcript</em></a>) &#8211; Reference to the originating transcript</li>
<li><strong>_vars</strong> (<em>dict(int,list(Variant))</em>) &#8211; Nonsynonymous variants that are
assoziated with the protein. 
key=position within protein, 
value=list of variants at that pos</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="Fred2.Core.Protein.Protein.newid">
<code class="descname">newid</code><em class="property"> = &lt;method-wrapper 'next' of itertools.count object&gt;</em><a class="headerlink" href="#Fred2.Core.Protein.Protein.newid" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Result">
<span id="fred2-core-result-module"></span><h2>Fred2.Core.Result module<a class="headerlink" href="#module-Fred2.Core.Result" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.AResult"></span><dl class="class">
<dt id="Fred2.Core.Result.AResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">AResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#AResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.AResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">pandas.core.frame.DataFrame</span></code></p>
<p>A AResult object is a DataFrame with with multi-indexing.</p>
<p>This class is used as interface and can be extended with custom short-cuts for the sometimes often tedious
calls in pandas</p>
<dl class="method">
<dt id="Fred2.Core.Result.AResult.filter_result">
<code class="descname">filter_result</code><span class="sig-paren">(</span><em>expression</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#AResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.AResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd><p>filter result based on a list of expressions</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expression</strong> &#8211; list((str,comparator,float)) expressions: A list of triples consisting of (method_name, comparator, threshold)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">AResult</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.AResult.merge_results">
<code class="descname">merge_results</code><span class="sig-paren">(</span><em>others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#AResult.merge_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.AResult.merge_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges results of the same type and returns a merged result</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>other</strong> (<a class="reference internal" href="#Fred2.Core.Result.AResult" title="Fred2.Core.Result.AResult"><em>AResult</em></a>) &#8211; </td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.CleavageFragmentPredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">CleavageFragmentPredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageFragmentPredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageFragmentPredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Result.AResult" title="Fred2.Core.Result.AResult"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Result.AResult</span></code></a></p>
<p>Epitope prediction result</p>
<dl class="method">
<dt id="Fred2.Core.Result.CleavageFragmentPredictionResult.filter_result">
<code class="descname">filter_result</code><span class="sig-paren">(</span><em>expressions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageFragmentPredictionResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageFragmentPredictionResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd><p>filters a result data frame based on a specified expression consisting
of a list of triple with (method_name, comparator, threshold). The expression is applied to each row. If any of the
columns full fill the criteria the row remains.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expressions</strong> (<em>list((str,comparator,float))</em>) &#8211; A list of triples consisting of (method_name, comparator, threshold)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">filtered result object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.CleavageFragmentPredictionResult.merge_results">
<code class="descname">merge_results</code><span class="sig-paren">(</span><em>others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageFragmentPredictionResult.merge_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageFragmentPredictionResult.merge_results" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.CleavageSitePredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">CleavageSitePredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageSitePredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageSitePredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Result.AResult" title="Fred2.Core.Result.AResult"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Result.AResult</span></code></a></p>
<p>Epitope prediction result</p>
<dl class="method">
<dt id="Fred2.Core.Result.CleavageSitePredictionResult.filter_result">
<code class="descname">filter_result</code><span class="sig-paren">(</span><em>expressions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageSitePredictionResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageSitePredictionResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd><p>filters a result data frame based on a specified expression consisting
of a list of triple with (method_name, comparator, threshold). The expression is applied to each row. If any of the
columns full fill the criteria the row remains.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expressions</strong> (<em>list((str,comparator,float))</em>) &#8211; A list of triples consisting of (method_name, comparator, threshold)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">filtered result object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.CleavageSitePredictionResult.get_peptides">
<code class="descname">get_peptides</code><span class="sig-paren">(</span><em>expression</em>, <em>length=9</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageSitePredictionResult.get_peptides"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageSitePredictionResult.get_peptides" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.CleavageSitePredictionResult.merge_results">
<code class="descname">merge_results</code><span class="sig-paren">(</span><em>others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#CleavageSitePredictionResult.merge_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.CleavageSitePredictionResult.merge_results" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.Distance2SelfResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">Distance2SelfResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#Distance2SelfResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.Distance2SelfResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Result.AResult" title="Fred2.Core.Result.AResult"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Result.AResult</span></code></a></p>
<p>Distance2Self prediction result</p>
<dl class="method">
<dt id="Fred2.Core.Result.Distance2SelfResult.filter_result">
<code class="descname">filter_result</code><span class="sig-paren">(</span><em>expressions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#Distance2SelfResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.Distance2SelfResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.Distance2SelfResult.merge_results">
<code class="descname">merge_results</code><span class="sig-paren">(</span><em>others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#Distance2SelfResult.merge_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.Distance2SelfResult.merge_results" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.EpitopePredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">EpitopePredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#EpitopePredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.EpitopePredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Result.AResult" title="Fred2.Core.Result.AResult"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Result.AResult</span></code></a></p>
<p>A AResult object is a DataFrame with with multi-indexing, where column Id are the prediction model (i.e HLA allele
for epitope prediction), row ID the target of the prediction (i.e. peptide) and the second row ID the predictor
(i.e BIMAS)</p>
<p>Epitope prediction result</p>
<dl class="method">
<dt id="Fred2.Core.Result.EpitopePredictionResult.filter_result">
<code class="descname">filter_result</code><span class="sig-paren">(</span><em>expressions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#EpitopePredictionResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.EpitopePredictionResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd><p>filters a result data frame based on a specified expression consisting
of a list of triple with (method_name, comparator, threshold). The expression is applied to each row. If any of the
columns full fill the criteria the row remains.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expressions</strong> (<em>list((str,comparator,float))</em>) &#8211; A list of triples consisting of (method_name, comparator, threshold)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">filtered result object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.EpitopePredictionResult.merge_results">
<code class="descname">merge_results</code><span class="sig-paren">(</span><em>others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#EpitopePredictionResult.merge_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.EpitopePredictionResult.merge_results" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Result.TAPPredictionResult">
<em class="property">class </em><code class="descclassname">Fred2.Core.Result.</code><code class="descname">TAPPredictionResult</code><span class="sig-paren">(</span><em>data=None</em>, <em>index=None</em>, <em>columns=None</em>, <em>dtype=None</em>, <em>copy=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#TAPPredictionResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.TAPPredictionResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Result.AResult" title="Fred2.Core.Result.AResult"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Result.AResult</span></code></a></p>
<p>Epitope prediction result</p>
<dl class="method">
<dt id="Fred2.Core.Result.TAPPredictionResult.filter_result">
<code class="descname">filter_result</code><span class="sig-paren">(</span><em>expressions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#TAPPredictionResult.filter_result"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.TAPPredictionResult.filter_result" title="Permalink to this definition">¶</a></dt>
<dd><p>filters a result data frame based on a specified expression consisting
of a list of triple with (method_name, comparator, threshold). The expression is applied to each row. If any of the
columns full fill the criteria the row remains.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expressions</strong> (<em>list((str,comparator,float))</em>) &#8211; A list of triples consisting of (method_name, comparator, threshold)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">filtered result object</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Result.TAPPredictionResult.merge_results">
<code class="descname">merge_results</code><span class="sig-paren">(</span><em>others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Result.html#TAPPredictionResult.merge_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Result.TAPPredictionResult.merge_results" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Transcript">
<span id="fred2-core-transcript-module"></span><h2>Fred2.Core.Transcript module<a class="headerlink" href="#module-Fred2.Core.Transcript" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Core.Transcript"></span><dl class="class">
<dt id="Fred2.Core.Transcript.Transcript">
<em class="property">class </em><code class="descclassname">Fred2.Core.Transcript.</code><code class="descname">Transcript</code><span class="sig-paren">(</span><em>_seq</em>, <em>_gene_id='unknown'</em>, <em>_transcript_id=None</em>, <em>_vars=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Transcript.html#Transcript"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Transcript.Transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a>, <code class="xref py py-class docutils literal"><span class="pre">Bio.Seq.Seq</span></code></p>
<p>A Transcript is the mRNA sequence containing at no or several variations.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For accessing and manipulating the sequence see also <code class="xref py py-mod docutils literal"><span class="pre">Bio.Seq.Seq</span></code>
(from Biopython)</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gene_id</strong> (<em>str</em>) &#8211; Genome ID</li>
<li><strong>transcript_id</strong> (<em>str</em>) &#8211; Transcript RefSeqID</li>
<li><strong>vars</strong> (<em>dict(int,Variant)</em>) &#8211; Dict of Variants for specific positions in
the transcript. key=position, value=Variant</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="Fred2.Core.Transcript.Transcript.newid">
<code class="descname">newid</code><em class="property"> = &lt;method-wrapper 'next' of itertools.count object&gt;</em><a class="headerlink" href="#Fred2.Core.Transcript.Transcript.newid" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-Fred2.Core.Variant">
<span id="fred2-core-variant-module"></span><h2>Fred2.Core.Variant module<a class="headerlink" href="#module-Fred2.Core.Variant" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-Variant"></span><dl class="class">
<dt id="Fred2.Core.Variant.MutationSyntax">
<em class="property">class </em><code class="descclassname">Fred2.Core.Variant.</code><code class="descname">MutationSyntax</code><span class="sig-paren">(</span><em>transID</em>, <em>transPos</em>, <em>protPos</em>, <em>cds</em>, <em>aas</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#MutationSyntax"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.MutationSyntax" title="Permalink to this definition">¶</a></dt>
<dd><p>This class represents the mutation syntax of a variant and stores its 
transcript and protein position</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>transID</strong> (<em>str</em>) &#8211; the transcript id</li>
<li><strong>transPos</strong> (<em>int</em>) &#8211; the position of the variant within the transcript</li>
<li><strong>protPos</strong> (<em>int</em>) &#8211; the protein position of the variant within the 
transcript</li>
<li><strong>cds</strong> (<em>str</em>) &#8211; the complete cds_mutation_syntax string</li>
<li><strong>aas</strong> (<em>str</em>) &#8211; the complete protein_mutation_syntax string</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="Fred2.Core.Variant.Variant">
<em class="property">class </em><code class="descclassname">Fred2.Core.Variant.</code><code class="descname">Variant</code><span class="sig-paren">(</span><em>id</em>, <em>type</em>, <em>chrom</em>, <em>genomePos</em>, <em>ref</em>, <em>obs</em>, <em>coding</em>, <em>isHomozygous</em>, <em>isSynonymous</em>, <em>experimentalDesign=None</em>, <em>metadata=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#Fred2.Core.Base.MetadataLogger" title="Fred2.Core.Base.MetadataLogger"><code class="xref py py-class docutils literal"><span class="pre">Fred2.Core.Base.MetadataLogger</span></code></a></p>
<p>A Variant contains information about a single genetic modification of
the reference genome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>id</strong> (<em>str</em>) &#8211; Variant id</li>
<li><strong>type</strong> (<em>Enum.VariationType</em>) &#8211; An Enum type of the variant either SNP, 
DEL, or INS</li>
<li><strong>chrom</strong> (<em>str</em>) &#8211; The chromosome on which the variant lies</li>
<li><strong>genomePos</strong> (<em>int</em>) &#8211; The genomic position of the variant</li>
<li><strong>ref</strong> (<em>str</em>) &#8211; The reference seq at the genomic position</li>
<li><strong>obs</strong> (<em>str</em>) &#8211; The observed variation at the genomic position</li>
<li><strong>coding</strong> (<em>dict(str,MutationSyntax)</em>) &#8211; A dictionary of associated 
transcripts. Key=transcript_id, 
value=MutationSyntax</li>
<li><strong>isHomozygous</strong> (<em>bool</em>) &#8211; defines if variant is homozygous or not</li>
<li><strong>isSynonymous</strong> (<em>bool</em>) &#8211; defines if variant is a synonymous mutation 
or not</li>
<li><strong>offsets</strong> (<em>dict(str,int)</em>) &#8211; the position offset this variant has in a 
specific transcript-variant
key=transcript-variant-id (xxx:FRED2_nn)
value=offset</li>
<li><strong>metadata</strong> (<em>defaultdict(list)</em>) &#8211; meta information (not relevant for core
functionality of Fred2)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="Fred2.Core.Variant.Variant.get_protein_position">
<code class="descname">get_protein_position</code><span class="sig-paren">(</span><em>trans_variant_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.get_protein_position"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.get_protein_position" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">May only be used for transcript variants that were created from this variant via :func:</p>
</div>
<p><cite>Generator.generate_transcripts_from_variants</cite>.
..TODO:: what? that should not be encoded in transcript id but in separate member</p>
<p>returns the specific protein position of a given transcript_id. If 
variant is not associated with the given transcript id the function 
throws a KeyError</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>transcriptId</strong> (<em>str</em>) &#8211; A transcript_id</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">(int) &#8211; the protein position of the variant</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body">KeyError</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="Fred2.Core.Variant.Variant.get_shift">
<code class="descname">get_shift</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.get_shift"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.get_shift" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.Core.Variant.Variant.get_transcript_offset">
<code class="descname">get_transcript_offset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.get_transcript_offset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.get_transcript_offset" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="Fred2.Core.Variant.Variant.get_transcript_position">
<code class="descname">get_transcript_position</code><span class="sig-paren">(</span><em>trans_variant_id</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/Fred2/Core/Variant.html#Variant.get_transcript_position"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#Fred2.Core.Variant.Variant.get_transcript_position" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">May only be used for transcript variants that were created from this variant via :func:</p>
</div>
<p><cite>Generator.generate_transcripts_from_variants</cite>.
..TODO:: what? that should not be encoded in transcript id but in separate member</p>
<p>returns the specific transcript position of a given transcript_id. 
If variant is not associated with the given transcript id the function 
throws a KeyError</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>transcriptId</strong> (<em>str</em>) &#8211; A transcript_id</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">(int) &#8211; transcript position</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body">KeyError</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="attribute">
<dt id="Fred2.Core.Variant.VariationType">
<code class="descclassname">Fred2.Core.Variant.</code><code class="descname">VariationType</code><a class="headerlink" href="#Fred2.Core.Variant.VariationType" title="Permalink to this definition">¶</a></dt>
<dd><p>Enum for variation types:
type.SNP, type.DEL, type.INS, type.FSDEL, type.FSINS, type.UNKNOWN</p>
<p>alias of <code class="xref py py-class docutils literal"><span class="pre">Enum</span></code></p>
</dd></dl>

</div>
<div class="section" id="module-Fred2.Core">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-Fred2.Core" title="Permalink to this headline">¶</a></h2>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Fred2.Core package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Allele">Fred2.Core.Allele module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Base">Fred2.Core.Base module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Generator">Fred2.Core.Generator module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Peptide">Fred2.Core.Peptide module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Protein">Fred2.Core.Protein module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Result">Fred2.Core.Result module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Transcript">Fred2.Core.Transcript module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core.Variant">Fred2.Core.Variant module</a></li>
<li><a class="reference internal" href="#module-Fred2.Core">Module contents</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Fred2.CleavagePrediction.html"
                        title="previous chapter">Fred2.CleavagePrediction package</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Fred2.Distance2Self.html"
                        title="next chapter">Fred2.Distance2Self package</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Fred2.Core.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Author.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.2</a>
      
      |
      <a href="_sources/Fred2.Core.txt"
          rel="nofollow">Page source</a></li>
    </div>

    

    
  </body>
</html>