{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1> Eptiope Prediction </h1>\n",
    "<h3> by Benjamin Schubert </h3>\n",
    "\n",
    "This tutorial illustrates the use of Fred2 to predict HLA-I/II epitopes and how to analyse results. Fred2 offers a long list of epitope prediction methods and was designed in such a way that extending Fred2 with your favourit method is easy.\n",
    "\n",
    "This tutorial will entail:\n",
    "- Simple epitope prediction from a list of peptide sequences and protein sequences\n",
    "- Manipulation of the results\n",
    "- Consensus prediction with multiple prediction methods\n",
    "- Integration of a new prediction method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "<h2> Chapter 1: The basics </h2>\n",
    "<br/>\n",
    "We first start with importing the needed packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "#is is needed as long as Fred2 is not installed properly\n",
    "import sys \n",
    "\n",
    "sys.path.append(\"/Users/schubert/Dropbox/PhD/Porgramming/Fred2\")\n",
    "from Fred2.Core import Allele, Peptide, Protein,generate_peptides_from_protein\n",
    "from Fred2.IO import read_lines, read_fasta\n",
    "from Fred2.EpitopePrediction import EpitopePredictorFactory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets start of with something playn simple: Defining peptides, and HLA alleles. You find all basic classes under `Fred2.Core`. HLA alleles have to be defined in new nomenclature (see: <a href=\"http://hla.alleles.org/\">http://hla.alleles.org/</a>). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peptides = [Peptide(\"SYFPEITHI\"),Peptide(\"FIASNGVKL\"), Peptide(\"LLGATCMFV\")]\n",
    "allele = Allele(\"HLA-A*02:01\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of course we don't have to specify all peptides and alleles by hand. We can also read those from file with `read_lines`. For doing so, the file has to contain the peptide sequences and HLA alleles one per line. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "peptides2 = read_lines(\"./data/peptides.txt\", type=Peptide)\n",
    "alleles = read_lines(\"./data/alleles.txt\", type=Allele)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Proteins are a bit more compilcated. They require a amino acid sequence as well as a transcript and gene ID. If `read_lines` is used generic transcript IDs are generated. Or one can use `read_fasta` to read in protein sequences in FASTA format. But one has to specify at which position of the fasta entrys  the transcript ID is (based on the contained |). For example if the fasta file has the folloing structure\n",
    "\n",
    "\n",
    ">\\>ID_Type|Gene_ID|ID_Type|Trancsript_ID|comment<br/>\n",
    ">AAAAAAAA\n",
    "\n",
    "we would call read_fasta as followes:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[PROTEIN: MKMRFFSSPCGKAAVDPADRCKEVQQIRDQHPSK (aa-seq)\n",
      "\t\t  NP_852610.1 (orig transcript)\n",
      "\t\t VARIANTS:\n",
      ", PROTEIN: IPVIIERYKGEKQLPVLDKTKFLVPDHVNMSELVKI (aa-seq)\n",
      "\t\t  NP_852610_part2 (orig transcript)\n",
      "\t\t VARIANTS:\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "proteins = read_fasta(\"./data/proteins.fasta\", id_position=3, type=Protein)\n",
    "print proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To be able to use proteins for prediction, we first have to generate k-mers of the sequences. For that we use the function `generate_peptides_from_protein`, which generates unique peptide k-mers based on the given protein sequences. These peptides are associated with the proteins from which they origin."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "peptide3 = generate_peptides_from_protein(proteins, 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fred2 has only one entry point to the different perdiction methods, namely `EpitopePredictorFactory`. It handels the initialization of the different methods and also collects newly implemented prediction methods if properly implemented. To see which prediction methods Fred2 supports `EpitopePredictorFactory` can helps here as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "netmhc 3.4\n",
      "smmpmbec 1.0\n",
      "syfpeithi 1.0\n",
      "netmhcpan 2.4\n",
      "netctlpan 1.1\n",
      "netmchiipan 3.0\n",
      "tepitopepan 1.0\n",
      "arb 1.0\n",
      "pickpocket 1.1\n",
      "epidemix 1.0\n",
      "netmhcii 2.2\n",
      "smm 1.0\n",
      "comblibsidney 1.0\n",
      "unitope 1.0\n",
      "hammer 1.0\n",
      "svmhc 1.0\n",
      "bimas 1.0\n"
     ]
    }
   ],
   "source": [
    "for name,version in EpitopePredictorFactory.available_methods().iteritems():\n",
    "    print name, \",\".join(version)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets select one and do predictions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "      <th>B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(((A)), ((N)), ((G)), ((H)), ((D)), ((A)), ((K)), ((A)), ((K)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((A)), ((T)), ((L)), ((I)), ((S)), ((I)), ((N)), ((T)), ((H)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((H)), ((I)), ((S)), ((L)), ((A)), ((N)), ((G)), ((H)), ((D)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)), ((A)), ((N)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((I)), ((S)), ((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              A*02:01  B*15:01\n",
       "Seq                                                Method                     \n",
       "(((A)), ((N)), ((G)), ((H)), ((D)), ((A)), ((K)... syfpeithi        2        2\n",
       "(((A)), ((T)), ((L)), ((I)), ((S)), ((I)), ((N)... syfpeithi       15        2\n",
       "(((H)), ((I)), ((S)), ((L)), ((A)), ((N)), ((G)... syfpeithi        8        6\n",
       "(((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)... syfpeithi        3        3\n",
       "(((I)), ((S)), ((I)), ((N)), ((T)), ((H)), ((I)... syfpeithi       16        2"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor = EpitopePredictorFactory(\"Syfpeithi\")\n",
    "results = predictor.predict(peptides2, alleles=alleles)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To initializea particular version of a prediction method, we can use the flage `version=\"\"` when calling the PredictorFactories. If we do not specify any version, Fred2 will initialize the most recent version that is supported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "      <th>B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(((A)), ((N)), ((G)), ((H)), ((D)), ((A)), ((K)), ((A)), ((K)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((A)), ((T)), ((L)), ((I)), ((S)), ((I)), ((N)), ((T)), ((H)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((H)), ((I)), ((S)), ((L)), ((A)), ((N)), ((G)), ((H)), ((D)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)), ((A)), ((N)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((I)), ((S)), ((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              A*02:01  B*15:01\n",
       "Seq                                                Method                     \n",
       "(((A)), ((N)), ((G)), ((H)), ((D)), ((A)), ((K)... syfpeithi        2        2\n",
       "(((A)), ((T)), ((L)), ((I)), ((S)), ((I)), ((N)... syfpeithi       15        2\n",
       "(((H)), ((I)), ((S)), ((L)), ((A)), ((N)), ((G)... syfpeithi        8        6\n",
       "(((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)... syfpeithi        3        3\n",
       "(((I)), ((S)), ((I)), ((N)), ((T)), ((H)), ((I)... syfpeithi       16        2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor = EpitopePredictorFactory(\"Syfpeithi\", version=\"1.0\")\n",
    "results = predictor.predict(peptides2, alleles=alleles)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "External tools like `NetMHC` offer two additional flags when calling `.predict()`, `command=\"/path/to/binary\"` and `options=\"command options\"`. `command=\"\"` specifies the path to an alternative binary that should be used instead of the one that is globally registered. With `options=\"\"` you can specify additional commands that will directly be passed to the commandline call without any sanity checks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "      <th>B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(((A)), ((N)), ((G)), ((H)), ((D)), ((A)), ((K)), ((A)), ((K)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((A)), ((T)), ((L)), ((I)), ((S)), ((I)), ((N)), ((T)), ((H)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((H)), ((I)), ((S)), ((L)), ((A)), ((N)), ((G)), ((H)), ((D)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)), ((A)), ((N)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((I)), ((S)), ((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)))</th>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              A*02:01  B*15:01\n",
       "Seq                                                Method                     \n",
       "(((A)), ((N)), ((G)), ((H)), ((D)), ((A)), ((K)... syfpeithi        2        2\n",
       "(((A)), ((T)), ((L)), ((I)), ((S)), ((I)), ((N)... syfpeithi       15        2\n",
       "(((H)), ((I)), ((S)), ((L)), ((A)), ((N)), ((G)... syfpeithi        8        6\n",
       "(((I)), ((N)), ((T)), ((H)), ((I)), ((S)), ((L)... syfpeithi        3        3\n",
       "(((I)), ((S)), ((I)), ((N)), ((T)), ((H)), ((I)... syfpeithi       16        2"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictor = EpitopePredictorFactory(\"Syfpeithi\", version=\"1.0\")\n",
    "results = predictor.predict(peptides2, alleles=alleles, options=\"-s -b\") \n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2> Chapter 2: Data manipulation and consensus prediction</h2>\n",
    "<br/>\n",
    "The predictor all return a data table like object (DataFrame) storing the peptide and allele objects, as well as the predicted results. Because it is a inherited pandas DataFrame we can do all sorts of nifty thinks. \n",
    "\n",
    "For example exporting the results to csv:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "results.to_csv('./data/test.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or we could plot the score distribution of a particular HLA allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes.AxesSubplot object at 0x105e1cf10>,\n",
       "        <matplotlib.axes.AxesSubplot object at 0x105fcd6d0>]], dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAEKCAYAAADticXcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztnXu0HdV93z+SLgIk8TALVzwk91ISbPCiCCcFysPeODgB\n1QF3LbeGFS+waRxWXAoY189gm6w0xStxYxbGjmkMFk4dnJY4VDjYGIwGw6IodpBkQIDBIBsRENSA\neCiqDVf9Y8/hzj06jz0zv5mzt+b7WWvuncc+3/nOzD77zHzPPjMghBBCCCGEEEIIIYQQQgghhBBC\nCCGEEEIIIYQQQogYyIBngYV98/858Jm+ebsD1wBbgSeBDxWWHQb8b+Bp4OfAd/J5o/hQrrMVuLrP\nw/nAD4HtwFeDtkSI0WwCtgEv4uv8t4BlheWD6vy/B+4CXgbWDNCcAV7KNV8E/vsYD6rzgcyftIFE\nmQaOwTfEp+fzjgU+CUzl0ycBn8jHLwUOBd4AnAx8FPitfNk+wA34hnwp8Pf4Rn4YvwV8DHg7/s30\nL4A/LCx/Avgj/IeIEBbsAN4J7AUcCGwBvsDoOv9z4M+Az47QPTLX3Av4vRHlVOdF43waWA38AXBj\nYf7pwC3AA8B/AfbM5z8BnFIo94fAdUO098OfxbxuyPK/yrV7nIw/g+nnj9BZi7DhMXyD2mMl8FA+\nPqzO9/hdhp+pHxq4ftX5EuhMvRpnA38N/E/8WcTrC8t25P9ngHn4xvlAYEOhzI+ANw/Rfiu+wj6X\nT59YGAc4YoDWUnb+EJgXsB1ChNKrT4uA9wD/p7Csv86H8n18Xf8b/Bl4D9X5GqhRL8+JwMH4M/WH\ngY3A7+AvRY8Efh/4BnArcCGwOH/d1oLGC/hLzn6WAVcCFxfm3cncyrtkgBYD9HYghA3z8BHhc8Dz\nwG8An2N4nQ/hrfiG/E3AP+Jz+gX5MtX5GkyNLyL6OAf4Lv7LHYD/lc+7HFjL7BnH9/OhVzn3Bv5v\nPr5P4fU9Xp/rfhF/FTCMl3KtHvvk//v1dNYirNgBnAHchq9X7wJuBw4H/pid63wId+b/t+I/CLbi\nG/j7B5RVnS+BztTLsSf+W/234y8bnwQ+DBwF/Mu8zE+Z+yXOc3m5FYV5RwH3FaZfh2/QbwAuG+Ph\n/gFaW5h7uQo6axHNsAP4W+BV/FUr7Fzn+8uPY17f/35U50ugRr0c7wJewZ+hHJUPhwN34HP2YXwN\nuATYNy//u8CqfNnewM34M5dPBnj4GvAfcp3XAZ9i7pdDC4A98FdhC/DdKRcgRD2KDe8Z+Lr3wIjy\n8/H1cLd8fPd8HHxGvgJfL5fge8lsHqGnOi8a49vAnw6Y/+/wueCwD8mF+L61W4GngIsKy85h5z67\nLzDbD/gkdr7M/FCu0+uzu1th2aW5XnH49LgNE2IEjzHbT/0F/BeVZ415zfvYuR72uhyeDDyIr/Nb\ngG8ytyeM6nwLLADWMbf7XpEr8F8abgCObsuUEBPmjfj3RW/YClwwUUdCBHIx8HV8j49+VgI35ePH\nAne3ZUqIiJiP/+5k+aSNiG4TkqkvwzfcX2HwFxmnA9fm42vxufFSE3dCpMMpwE+AxydtRHSbkEb9\n88BH8DnVIA5mbkXezNz7QgjRBc7E//JRiIkyrlF/J/7+JusY3Qe0f5m6FokusRD4bfxvFoSYKON+\nfHQ8Pl5Zie8ytDe+e1Gx+94TzM0Rl+Xz+nkCOKiyUyE8PwF+ZdIm+jgN+Afgmb75qvPCgsbq/NsY\n3Pul+EXpcQz/orSJs/dLS5TdATsChs8Uxk08j/IY6GknP6M0m/AZk2aMV4HfwHdN7cfK66UGGgPq\nd3B9a8rTrqxjqVWqHpW9TUBP/Lz8/1X4Bn0l8Aj+3snvL6lZh2l7yU3WgtPWgh3XjI3F+C9JP9Dg\nOqbtpDZZCU1Lp1WtYMo06rfnA/jGvMj5NnaESI6Xgf0nbUKISdDEZbMrt/6QS8411vHLKI9V45dR\nmk34jEkzxvhlGFZenYHGgPpdK36x8LQr61hqlapHbd7VbEfL6xuw/rLvsXmv/WmIkp4a95MCk65H\nZYjJa4z1X4RRqh6lfkMvZy+ZWQs6a8GOa3YRZyeVWQk56bSqFUzqjboQQogCil9GovglQiZdj8oQ\nk9cY678Io1PxixBCiAKpN+rOXjKzFnTWgh3X7CLOTiqzEnLSaVUrmNQbdSGEEAWUqY9EmXqETLoe\nlSEmrzHWfxGGMnUhhOgqqTfqzl4ysxZ01oId1+wizk4qsxJy0mlVK5jUG3UhhBAFlKmPRJl6hEy6\nHpUhJq8x1n8RhjJ1IYToKqk36s5eMrMWdNaCHdfsIs5OKrMSctJpVSuY1Bt1IYQQBZSpj0SZeoRM\nuh6VISavMdZ/EYYydSGE6CqpN+rOXjKzFnTWgh3X7CLOTiqzEnLSaVUrmJBGfQ9gLbAe2AhcNqCM\nA7YC6/LhEiN/QgghGmBR/n8KuBs4sW+5A1aP0Zj0syVLPg/U7Bmlhp4mvg9jIKV9EJPXGOu/CKPU\ncQiNX7bl/xcCC4BnB5TRFypCCDFhQhv1+fj4ZQuwBh/DFNkBHA9sAG4CjrAyOAZnL5lZCzprwY5r\ndhFnJ5VZCTnptKoVTGijPgOsAJYBb2Vns/cAy4GjgC8ANxj5EyJm9gWuBx7An+gcN1k7QlSLTD4F\n/BPwuRFlHgN+jbkxzQ7gWmBTPv08/uw/y6dd/r+p6R3+IqPMy06G2X3UhL81s3FZFH5inO6NT+fj\n5xBP1HctcDtwDf77psX4DgM91E9dWGBej/bHn5EA7Al8H/iNvjJLCys9htmGu9/YJInxi6LY/KRA\nLPtgH+DRMWVi8Qpx1n8RhvkXpQcCt+HPqtcCNwLfA87LB4B3A/fmZS4HzixjogbOXjKzFnTWgh3X\njIVDgGeAr+Ljx79gtpeYNc5OKrMSctJpVSuYqYAy9wJvGTD/qsL4F/NBiK4whX9fnA/8AH8y83Hg\n033lVlE/cmTM8sDpLF99qdW66usLml7RsP4k/ayo+Pre+DSRM+lLuRgvP2PzkwKx7IMD8N8d9TgR\n+FZfmVi8Qpz1X4TRSD91IcRcngIeBw7Lp08B7p+cHSHap4lPfVdu/SFnJ2usz1RGeax65jRKswmf\nMWnGdPZ4FD562QB8E//laRErr85AY0D9rnWmbuFpV9ax1CpVj0IydSHEYDYA/2rSJoQoovupj0T3\nU4+QSdejMsTkNcb6L8IoVY+UqQshxC5E6o26s5fMrAWdtWDHNbuIs5PKrIScdFrVCib1Rl0IIUQB\nZeojUaYeIZOuR2WIyWuM9V+EoUxdCCG6SuqNurOXzKwFnbVgxzW7iLOTyqyEnHRa1Qom9UZdCCFE\nAWXqI1GmHiGTrkdliMlrjPVfhKFMXQghukrqjbqzl8ysBZ21YMc1u4izk8qshJx0WtUKJvVGXQgh\nRAFl6iNRph4hk65HZYjJa4z1X4ShTF0IIbpK6o26s5fMrAWdtWDHNbuIs5PKrIScdFrVCmZco74H\n/mHT64GNwGVDyl0BPIy/v/TRZu6EEEKY03tC+hRwN/5ZjEVWAjfl48fmZQYx6SfWxPiMxtj8pEBK\n+yAmrzHWfxFGqeMQEr9sy/8vBBYAz/YtPx24Nh9fC+wLLC1jQgghhA0hjfp8fPyyBViDj2GKHIx/\nAG+PzcCyCl5eDxxUYtgbZepd1Owizk4qsxJy0mlVK5iQZ5TOACvwD9W9GW806yvT391m2OXCKmBT\nPv48/sMiAw4FfgyLfwG7zfjFv8w/cAZNz8yD7S/DL/644MXl/4dN56MBxV6bng/MlLj0mb8NZhb3\nCdM33b/iMX6G+RtbsOz0CmM9l2vW1euNTyOEMOdTwH/um/dl4MzC9IMMjl9GNY5vhjdsDc/6Ht8B\ni/tjoHFUzBQbzSCVcZYnpX0Qk1dl6ulimqnvj8/IAfYE3gGs6yuzGjg7Hz8Ofwa+pYwJIYQQNoxr\n1A8EbsPHJGuBG4HvAeflA/ieL48CjwBXAR9sxOlgnL1kZi3orAU7rtlFnJ1UZiXkpNOqVjDjMvV7\ngbcMmH9V3/T5NnaESI5NwAvAq8AvgWMm6kaIFlGmXtuTMk7i2wePAfsNWRaTV2Xq6WKaqQshxqOb\nXoloSL1Rd/aSmbWgsxbsuGZs7ABuBX4IfKChdTg7qcxKyEmnVa1gQvqpCyGGcwLwJP7Hc7fgu/Te\nUVi+isG/zYByffXLlB8yneWrL7VaV319rf2WIVY/VX/70RufJnKUqdf2pIyTuPfBZ4APF6Zj8qpM\nPV2UqQvREouAvfLxxcBv4nuMCTExUm/Unb1kZi3orAU7rhkTS/FRS+93HN8CvtvAepydVGYl5KTT\nqlYwytSFqM5jzOamQnQOZeq1PSnjJK19EJNXZerpokxdCCG6SuqNurOXzKwFnbVgxzW7iLOTyqyE\nnHRa1Qom9UZdCCHEhFCmXtuTMk7S2gcxeVWmni7K1IUQoquk3qg7e8nMWtBZC3Zcs4s4O6nMSshJ\np1WtYFJv1IUQQkwIZeq1PSnjJK19EJNXZerpokxdCCG6SuqNurOXzKwFnbVgxzW7iLOTyqyEnHRa\n1QompFFfDqwB7gfuAy4YUMYBW4F1+XCJkT8hhBDGHMDsTYuWAA8Bh/eVccDqMTrK1Gt7UsZJWvsg\nJq/K1NPFPFN/Cn9rUYCXgAeAgwaU03MahRBiwpTN1KeBo/H3ji6yAzge2ADcBBxR21kYzl4ysxZ0\n1oId1+wizk4qsxJy0mlVK5gy91NfAlwPXIg/Yy9yDz573wacBtwAHDZAYxVDn9e4fYEfdfnifPbQ\n6VenKPc8QcL0GTMd+vrg50vW1Q994bjpqs9THDVt8bzH3vg0QggzdgNuBi4KLP8YsF/fPGXqtT0p\n4yStfRCTV2Xq6WKeqc8DrgY2ApcPKbOU2Uz9mHy8bKMrhBCiJiGN+gnAe4GTme2yeBpwXj4AvBv/\nwN31+Ib/THOng3H2kpm1oLMW7LhmF3F2UpmVkJNOq1rBhGTqdzK+8f9iPgghhOgIytRre1LGSVr7\nICavytTTxTxTF0IIkQipN+rOXjKzFnTWgh3X7CLOTiqzEnLSaVUrmNQbdSEmzQJ854EbJ21EiLZR\npl7bkzJO4tsHFwNfZ/C9j2Lyqkw9XZSpC9ESy4CVwFfQvY9EJKTeqDt7ycxa0FkLdlwzJj4PfASY\naXg9zk4qsxJy0mlVK5gy934RQszyTuBpfJ7uRpRbxdD7HZW6/02Z8kOms3z1pVbrqq+vtfsDxeqn\n6v2UeuPTRI4y9dqelHESzz74r8Dj+PscPQm8DHytr0wsXkGZespEexzUqNf2FO/BbZEY98HbGNz7\nJSavatTTpdRxUKa+E5m1oLMW7LhmrDTZADo7qcxKyEmnVa1glKkLUZ/b80GITqH4pbYnXQ6T1j6I\nyavil3TpVPwihBCiQOqNurOXzKwFnbVgxzW7iLOTyqyEnHRa1Qom9UZdCCHEhFCmXtuTMk7S2gcx\neVWmni7K1IUQoquk3qg7e8nMWtBZC3Zcs4s4O6nMSshJp1WtYEIa9eXAGuB+4D7ggiHlrgAeBjYA\nR5u4E0IIYc4BzN6YZgnwEHB4X5mVwE35+LHA3QN0lKnX9qSMk7T2QUxelamni3mm/hT+1m4ALwEP\nAAf1lTkduDYfXwvsCywtY0QIIUR9ymbq0/hoZW3f/IPxd6zrsRn/AIGmcfaSmbWgsxbsuGYXcXZS\nmZWQk06rWsGUuffLEuB64EL8GXs//U9+GXTJsIqh95bevsCPunxxPnvo9KtTlLv3MWH6jJke9fr5\nwEzFS9YQ/TmMK1h2uuq9n0dNW9ybujc+jRDCjN2Am4GLhiz/MnBmYfpBdo5fOpKpR5XZ74qktA9i\n8qpMPV3MM/V5wNXARuDyIWVWA2fn48fhz8K3lDEihBCiPiGN+gnAe4GT8Y/uWgecBpyXD+B7vjwK\nPAJcBXzQ3OlgnL1kFrkekE7+3YRmF3F2UpmVkJNOq1rBhGTqdxLW+J9f04sQQoiEUKauTN2ClPZB\nTF6VqaeLeaYuhBAiEVJv1J29ZBa5HpBO/t2EZhdxdlKZlZCTTqtawaTeqAshhJgQytSVqVuQ0j6I\nyasy9XRRpi6EEF0l9Ubd2UtmkesB6eTfTWh2EWcnlVkJOem0qhVM6o26EJNiD/yN7dbjf2192WTt\nCNE+ytSVqVsQ0z5YlP+fwj9D4MS+5TF5VaaeLsrUhWiJbfn/hcACoOyJhhDmpN6oO3vJLHI9IJ38\nuwnNmJiPj1+24B/5uLGh9Tg7qcxKyEmnVa1gytxPXQgxlxn8PeP3wd+a2rFzq7mKoc8QKHVP+cL0\n1Gp4Za/ydrN89aVW6wIK1pie923YsccQw0OYehFe2bsZP3WeATD/ZZjpRXKBzN8GM4v79Hrj0+W0\nPP0PtmiSHSPW92Z4w13w073DpDYDb3oOXt6v3PrLRoTzKPeaNsq3esxiZFQ9miSfAv4J+FxhXlNe\nW6jLvdc0vq8rbkuUdaCpbSlVj1KPX4SYFPvjn8ULsCfwDvxtqYWYKKk36s5eMotcD0gn/25CMxYO\nBG7D5xlrgRuB7zW0LmcnlVkJOSshI1xkOjT03h+LMnUhqnEv8JZJmxCiH2XqI1GmHiGxZuqDUKY+\nHmXqytSFEEIMI/VG3dlLZpHrAenk301odhFnJ5VZCTkrISNcZDpMKlMPadSvwf+44t4hyx2wldmH\nUl9i4kwIIUQjnAQczehGfXWAju79onu/WJDSPmjKawt1ubX6FquvKjS1LaW2N+RM/Q7guTFlYvzS\nQgghOodFpr4DOB7YANwEHGGgGYqzl8wi1wPSyb+b0Owizk4qsxJyVkJGuMh0SLmf+j3Acvwd604D\nbgAOG1J2FUPvg7F9gR91+eJ89tDpV6cod58GwvQZM239+qr65ve9WGGs56h1Hw2b+2AIIQYzzfBM\nvZ/HgEH9x5WpK1O3IKV9oEy9sW2JkmQy9XEsZTZTPyYf132lhRBiAoQ06tcBdwFvBB4HzgXOyweA\nd+PP4tcDlwNn2tscirOXzCLXA9LJv5vQ7CLOTiqzEnJWQka4yHSIOVM/a8zyL+aDEEKICaN7v4xE\n936JkFH1KDaa8tpCXe69Rvd+KUFT21KqHqV+mwAhhBAFUm/Unb1kFrkekE7+3YRmF3F2UpmVkLMS\nMsJFpkPM934RQgiRCMrUR6JMPUKUqStTf+1PZChTF0IIYUvqjbqzl8wi1wPSyb+b0Owizk4qsxJy\nVkJGuMh0UKYuhBCiNsrUR6JMPUKUqStTf+1PZChTF0IIYUvqjbqzl8wi1wPSyb+b0Owizk4qsxJy\nVkJGuMh0UKYuRHosB9YA9wP3ARdM1o4Q7TIqbNL91NO+j3SbxLQPDmD24SJLgIeAwwvLm/LaQl3W\n/dQr0NS2lNpenakLUZ2n8LecBngJeAA4aHJ2hLB5nN0kcZgHVxm2caG1HtDIdiejGSvTwNHA2hKv\n2R04H1g4ptwh+CeKGZBhVB8dcR1bh40fKx0aeu+PJfVGXYgYWAJcD1yIP2Mvsoqhz+XlPbDHZXBB\nfsX8s7zb2ht2zJ3uzfvZPPjxPP9YYCj/uNcsX31o+dcmqqyozHTJl83xZu2nznN1mduQrx9SbMDL\n5i7ojU/3F4gNZerK1C2IbR/sBtwMXDRg2Tivb4KDXyhXB66skY9HmV3H6qsKTW1Lqe1Vpi5EdeYB\nVwMb8Y9yFGLipN6oO3vJLHI9IJ0+5U1oxsQJwHuBk4F1+XCq/WqyGLWclZARLjIdYn5G6TXAvwGe\nBo4cUuYK4DRgG/A+fOUWYlfnTtI/MRId5CT8t/r3Dlm+ErgpHz8WuHtIuVG5kDL1tLPENklpH4zz\nqkw9Xl9VaGpbSm1vyFnGHcBzI5afDlybj68F9gWWljEhhBDCBotLx4OBxwvTm4FlBrohOHvJLHI9\nIJ38uwnNDpLFqOWshIxwkekQc6YeQv9tIYddLqxiaJ/d7Qvm9vHMZw+d3v46/H03ShCiz5hp69eX\n0Z8PzJS4FJu/DWYWBwjD7M/dQ4zk0/P/DmYWNeSnN90bnw5fjxBiHNMMz9S/DJxZmH6QwfHLqMao\nQqbedH4da6YeVfY4CT+x5qmDGOdVmXq8vqrQ1LaU2l6L+GU1cHY+fhz+DHyLga4QQoiShDTq1wF3\nAW/EZ+fnAuflA/ieL48CjwBXAR+0tzmMLAFNa72mNFP4fqKrZDFqOSshI1xkOsScqZ8VUOb8ukaE\nEEKkxahcSJl6Y+WbPqat+4k1Tx3EOK/K1OP1VYWmtqXU9urXcEIIsQuReKOeJaBprdeUpjL1eMli\n1HJWQka4yHTQM0qFEEIkxahcSJl6Y+WbPqat+4k1Tx3EOK/K1OP1VYWmtqXU9upMXQghdiESb9Sz\nBDSt9ZrSVKYeL1mMWs5KyAgXmQ7K1IUQQiTFqFxImXpj5Zs+pq37iTVPHcQ4r8rU4/VVhaa2pdT2\n6kxdCCF2IRJv1LMENK31mtJUph4vWYxazkrICBeZDsrUhRBCJMWoXEiZemPlmz6mrfuJNU8dxDiv\nytTj9VWFpral1PbqTF0IIXYhEm/UswQ0rfWa0lSmHi9ZjFrOSsgIF5kOytSFSI9r8E/5GvaoRyF2\naUblQsrUGyvf9DFt3U9MeepJwNEMb9THeVWmHq+vKjS1LaW2V2fqQlTnDuC5SZsQokhoo34q8CDw\nMPCxAcsdsBVYlw+XWJgbT5aAprVeU5rK1OMli1HLWQkZ4SLTIeZnlC4ArgROAZ4AfgCsBh7oK3c7\ncLqpOyHSZxWwKR9/HljP7Lv9GNi+YLZob7brmy4u/3Hf9KDyo6bXlyj/2kSVFZWZLvmyOd4CXzj1\nAryyV7/AcKZehFd67VkJY1lh9vrwl81d0BufHmuzIv8a+E5h+uP50G/mxjE6o3IhZeqNlW+USfiJ\nLU+dRpl6HdryFeE6gtdTyktI/HIw8HhhenM+r3+lxwMbgJuAI8qYEEIIYUNIox7yKXEPsBw4CvgC\ncEMdU+FkCWha6zWlqUy9AtcBdwGH4U983t/MarIYtZyVkBHORiazkTHXCickU38C32D3WI4/Wy/y\nYmH828CXgP2AZ/vKrWJovrh9wdw8Kp89MoYrmw+G6DNm2vr1dfVDy48tuKKkcG96RLFBxyfYT1Hf\n0WC+WIOzJm1AiCpMAT/Bv6kW4t+ph/eVWQrMy8ePYbbhLjLqjF+ZemPlG2USfpreJkvGeVWm3p6v\nCNcRvJ5SXkLO1F8BzgduxveEuRrf8+W8fPlVwLuB38/LbgPOLGNCCCFEeoz6tKl4pr6mgTOVNRVe\nM6r8KI9V9UO3u9QnvKt2TEP3pdmZn4VGW4zzGnimXtyPdc/Uy75nhuLs9lGVbSntJ1B7Tci2B64j\nZF/rF6VCCCEiYdSnjTL1xso3fUxb99P0Nlkyzqsy9fZ8RbiO4PWU8qIzdSGE2IVIvFHPEtC01mtK\nU/3U4yWLUctZCRnhbGQyGxlzrXASb9SFEEJMilG5kDL1xso3fUxb99P0Nlkyzqsy9fZ8RbiO4PWU\n8qIzdSGE2IVIvFHPEtC01mtKU5l6vGQxajkrISOcjUxmI2OuFU7ijboQQohJMSoXUqbeWPmmj2nr\nfpreJkvGeVWm3p6vCNcRvJ5SXnSmLoQQuxCJN+pZAprWek1pKlOPlyxGLWclZISzkclsZMy1wkm8\nURdCCDEpRuVCytQbK9/0MW3dT9PbZMk4r8rU2/MV4TqC11PKi87UhRBiFyLxRj1LQNNarylNZerx\nksWo5ayEjHA2MpmNjLlWOIk36kIIISbFqFxImXpj5Zs+pq37aXqbLBnnVZl6e74iXEfwekp50Zm6\nEELsQoQ06qcCDwIPAx8bUuaKfPkG4GgbayFkCWha6zWlqUy9AiHvDQOyGLWclZARzkYms5Ex1wpn\nXKO+ALgSX3mPAM4CDu8rsxL4FeBXgd8D/tzY4wjWJ6CZgkcAVthLNuIzFkLeG0ZY7kczrQbqSy2M\n/ES5r0sxrlE/BngE2AT8EvgGcEZfmdOBa/PxtcC+wFI7i6N4PgHNFDwC/rgZ04jPWAh5bxhhuR/N\ntBqoL7Uw8hPlvi7FuEb9YODxwvTmfN64MsvqWxMiakLeG0K0ztSY5aHfus6r+LoeM/DUnvD2rWHF\nt88H9vInSdZYa1rrNaXJtL3kJnvJeLDoGTIDz+wxvt7fvwi+v82PP747sEf1VW6q/tK5TFsJGTFt\nI7PJRsZcy47jgO8Upj/Bzl8IfRk4szD9IIPjl0fwbwQNGuoMjxAHIe8N1XkNFoNpnZ8CfoL/FFyI\nT/4HfVF6Uz5+HHC3pQEhIiXkvSFElJwGPIT/tPhEPu+8fOhxZb58A/CWVt0JMTkGvTeEEEIIIdLC\n+kcam4AfAeuAv6+ocQ2wBbi3MG8/4Bbgx8B3Kd9NapDmpfieEevy4dSSmsuBNcD9wH3ABTW9DtOr\n43MPfHfW9cBG4LKaHkdp1vE5Cf4UeAB/FftNYJ+Sr7d47ww75lVZgN/3N9bU2Re4Hr9/NuLj2yp8\nAr9t9wJ/Bewe+DqrNmCQTpXjPkinx4eBmdzfxFmAvzydBnbDJnt8jPobdxL+16/FHfgnwEfz8Y8B\nnzXQ/AxwcUWPAAcw+8OKJfjL/cOp7nWYXl2fi/L/U/jvVU6s4XGUZl2fbfMOZrsOf5Zy+8DqvTPs\nmFflYuDrwOoaGuB/33JuPj5F+Q888PvmUWYb8r8Gzgl8rVUbMEinynEfpAP+Q/k7BLZ7bdz7pakf\nafR3oyzLHcBzffOKP6S6FniXgSbU8/oUsz9Newn/6X8w1b0O06vrM+9yx0J8Y/RcDY+jNKH+sW+T\nW/BnWOCvPMr8hsPqvTPomB9UQQe8/5XAV6h3HPbBN2LX5NOvAIFdmufwAn7fLMJ/MCwCngh8rVUb\nMEinynEf1n78GbMfNGNpo1Fv4kcaO4BbgR8CH6ipVWQp/vKH/L/VL2P/E/4y7Grq/fJtGv9JvhYb\nrz29Xo+lOj7n4xuOLcxe6tf1OEizrs9Jci6zPcVCaOK9M81sHarC54GPMNtgVeUQ4Bngq8A9wF8w\ne2VWhmf6HDIpAAACp0lEQVSB/wb8DPhH/M84b63hq4k2oOxxL3IG/rj/KPQFbTTqOxrQPAFfMU8D\n/iP+E9+aXh/Ruvw5vgKvAJ7EV8AqLAH+BrgQeLFvWRWvS/B55oX4s7e6Pmfy1y4D3gqcbOCxX9MZ\n+GyCW/CXzP3DbxfK/AHwC3zmG4r1e6f/mJflncDT+Dy97tXSFL6n3Jfy/y8DH6+gcyhwEf7D6iD8\nNv5OTW89LNqAKse9xyLgk/jIscfY/d5Go/4EPhPqsRz/yVOHJ/P/zwB/i79MtWALPn8EOBBfgevy\nNLOV4ytU87obvkH/S+CGfF4drz29/1HQs/AJ/hL674Bfq+lxkOavG/q05B3AkQOG3heJ78NHFmUb\nG8v3zqBjXpbj8fHEY8B1wNuBr1XU2pwPP8inr6dad+hfB+4Cfo6PcL6Z+6yKZRvwPqod9x6H4j+s\nNuD3+TLgH4B/NupFbTTqP8TfwXEan42+h3pfsCwC9srHFwO/yeBvi6uwmtkvWc6heuUvcmBh/N9S\n3us8fMywEbi8ML+q12F6dXzuz2wMsie+kVtXw+MozQMKZarsz7Y5FR9XnAFsL/laq/fOsGNelk/i\nP1gOwf+K/Dbg7IpaT+GjpcPy6VOYjdfK8CC+18ye+O08Bb+dVbFqA+oc9x734uOfQ/JhM/6Dz+Jk\nszaWP9I4BJ+zrsd3z6qqdx0+g/sFvnK9H//N8q1U79LYr3ku/kzmR/hP2xson9GdiI8h1jO3G19V\nr4P0Tqvp80h8Lro+1/hIPr/O/hymWXd/ts3DwE+Z3ddfKvl6i/fOsDpUh7dRv/fLUfgz9ardPXt8\nlNkujdfir0pCsGoDBr3vqxz3ns7/K/gp8iiRdGkUQgghhBBCCCGEEEIIIYQQQgghhBBCCCGEEEII\nIYQQQiTI/wdNoBoh4015oQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x105f31fd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "results.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To combine prediction results we can use `merge_results` from `Fred2.Core`. In addition to the result objecte we want to merge, also have to specify the type of these objects (here `EpitopePredictionResult`). The function will return a merged results object of the same type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "      <th>B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">(((F)), ((I)), ((A)), ((S)), ((N)), ((G)), ((V)), ((K)), ((L)))</th>\n",
       "      <th>bimas</th>\n",
       "      <td>13.511718</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>438.641785</td>\n",
       "      <td>4315.190768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>svmhc</th>\n",
       "      <td>-0.358894</td>\n",
       "      <td>-0.836802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>25.000000</td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)), ((F)), ((V)))</th>\n",
       "      <th>bimas</th>\n",
       "      <td>650.311200</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>27.108782</td>\n",
       "      <td>3689.775986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>svmhc</th>\n",
       "      <td>0.969253</td>\n",
       "      <td>-0.548826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>22.000000</td>\n",
       "      <td>10.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)), ((H)), ((I)))</th>\n",
       "      <th>bimas</th>\n",
       "      <td>0.012331</td>\n",
       "      <td>0.014400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>2931.635689</td>\n",
       "      <td>7014.552984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>svmhc</th>\n",
       "      <td>0.096541</td>\n",
       "      <td>-0.524968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>17.000000</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  A*02:01  \\\n",
       "Seq                                                Method                   \n",
       "(((F)), ((I)), ((A)), ((S)), ((N)), ((G)), ((V)... bimas        13.511718   \n",
       "                                                   smm         438.641785   \n",
       "                                                   svmhc        -0.358894   \n",
       "                                                   syfpeithi    25.000000   \n",
       "(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)... bimas       650.311200   \n",
       "                                                   smm          27.108782   \n",
       "                                                   svmhc         0.969253   \n",
       "                                                   syfpeithi    22.000000   \n",
       "(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)... bimas         0.012331   \n",
       "                                                   smm        2931.635689   \n",
       "                                                   svmhc         0.096541   \n",
       "                                                   syfpeithi    17.000000   \n",
       "\n",
       "                                                                  B*15:01  \n",
       "Seq                                                Method                  \n",
       "(((F)), ((I)), ((A)), ((S)), ((N)), ((G)), ((V)... bimas         0.600000  \n",
       "                                                   smm        4315.190768  \n",
       "                                                   svmhc        -0.836802  \n",
       "                                                   syfpeithi     7.000000  \n",
       "(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)... bimas         4.000000  \n",
       "                                                   smm        3689.775986  \n",
       "                                                   svmhc        -0.548826  \n",
       "                                                   syfpeithi    10.000000  \n",
       "(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)... bimas         0.014400  \n",
       "                                                   smm        7014.552984  \n",
       "                                                   svmhc        -0.524968  \n",
       "                                                   syfpeithi     4.000000  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "results = [EpitopePredictorFactory(m).predict(peptides,alleles=alleles) \n",
    "                    for m in [\"Syfpeithi\",\"BIMAS\",\"SMM\",\"SVMHC\"]]\n",
    "df = results[0].merge_results(results[1:])\n",
    "df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We also can filter the predicted epitopes based on ther prediction values with the function `filter_result` from `Fred2.Core`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "      <th>B*15:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)), ((F)), ((V)))</th>\n",
       "      <th>bimas</th>\n",
       "      <td>650.311200</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>27.108782</td>\n",
       "      <td>3689.775986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>svmhc</th>\n",
       "      <td>0.969253</td>\n",
       "      <td>-0.548826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>22.000000</td>\n",
       "      <td>10.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)), ((H)), ((I)))</th>\n",
       "      <th>bimas</th>\n",
       "      <td>0.012331</td>\n",
       "      <td>0.014400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smm</th>\n",
       "      <td>2931.635689</td>\n",
       "      <td>7014.552984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>svmhc</th>\n",
       "      <td>0.096541</td>\n",
       "      <td>-0.524968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>syfpeithi</th>\n",
       "      <td>17.000000</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  A*02:01  \\\n",
       "Seq                                                Method                   \n",
       "(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)... bimas       650.311200   \n",
       "                                                   smm          27.108782   \n",
       "                                                   svmhc         0.969253   \n",
       "                                                   syfpeithi    22.000000   \n",
       "(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)... bimas         0.012331   \n",
       "                                                   smm        2931.635689   \n",
       "                                                   svmhc         0.096541   \n",
       "                                                   syfpeithi    17.000000   \n",
       "\n",
       "                                                                  B*15:01  \n",
       "Seq                                                Method                  \n",
       "(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)... bimas         4.000000  \n",
       "                                                   smm        3689.775986  \n",
       "                                                   svmhc        -0.548826  \n",
       "                                                   syfpeithi    10.000000  \n",
       "(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)... bimas         0.014400  \n",
       "                                                   smm        7014.552984  \n",
       "                                                   svmhc        -0.524968  \n",
       "                                                   syfpeithi     4.000000  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#you can either use pre-defined operators from `operator`\n",
    "from operator import ge\n",
    "#or define you own comparator function like this\n",
    "comparator = lambda a,b: a > b\n",
    "\n",
    "df.filter_result([(\"svmhc\",comparator,0.0)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With that one can combine several prediction tools to form a consensus prediction method.<br/><br/>\n",
    "<h2> Chapter 3: Implementation of a new epitope prediction method </h2>\n",
    "<br/>\n",
    "Fred2 posseses a pontent plugin system allwoing the user to extend its capability quite easyly. To include a new epitope prediction method one simply has to inherite from `Fred2.Core.AEpitopePrediction` and implement its interface. For methods calling an external prediction tool additionally have to inherite from `Fred2.Core.AExternal`. SVM based methods also define a specific interface via `Fred2.Core.ASVM`. Fred2 uses SVMlight and its python binding svmlight 0.4.\n",
    "\n",
    "If you want to be very specific and fully integrate your method in all of Fred2's capabilities please use one of the three major interfaces `APSSMEpitopePrediction`, `ASVMEpitopePrediction`, or `AExternalEpitopePrediction`from `Fred2.EpitopePrediction`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Fred2.EpitopePrediction import APSSMEpitopePrediction\n",
    "from Fred2.Core import EpitopePredictionResult\n",
    "import random\n",
    "import pandas\n",
    "\n",
    "class RandomEpitopePrediction(APSSMEpitopePrediction):\n",
    "    __alleles = [\"A*02:01\"]\n",
    "    __supported_length = [9]\n",
    "    __name = \"random\"\n",
    "    __version= \"1.0\"\n",
    "    \n",
    "    #the interface defines three class properties\n",
    "    @property\n",
    "    def name(self):\n",
    "        #retunrs the name of the predictor\n",
    "        return self.__name\n",
    "    \n",
    "    @property\n",
    "    def supportedAlleles(self):\n",
    "        #returns the supported alleles as strings (without the HLA prefix)\n",
    "        return self.__alleles\n",
    "    \n",
    "    @property\n",
    "    def supportedLength(self):\n",
    "        #returns the supported epitope lengths as iterable\n",
    "        return self.__supported_length\n",
    "    \n",
    "    @property\n",
    "    def version(self):\n",
    "        #returns the version of the predictor\n",
    "        return self.__version\n",
    "    \n",
    "    #the interface defines a function converting Fred2's HLA allele presentation\n",
    "    #into an internal presentation used by different methods.\n",
    "    #for this predictor we won't need it but still have to provide it!\n",
    "    #the function consumes a list of alleles and converts them into the internally used presentation\n",
    "    def convert_alleles(self, alleles):\n",
    "        #we just use the identity function\n",
    "        return alleles\n",
    "    \n",
    "    #additionally the interface defines a function `predict` \n",
    "    #that consumes a list of peptides or a single peptide and optionally a list \n",
    "    #of allele objects\n",
    "    #\n",
    "    #this method implements the complete prediction routine\n",
    "    def predict(self, peptides, alleles=None):\n",
    "        \n",
    "        #test whether one peptide or a list\n",
    "        if isinstance(peptides, basestring):\n",
    "            peptides = list(peptides)\n",
    "        \n",
    "        #if no alleles are specified do predictions for all supported alleles\n",
    "        if alleles is None:\n",
    "            alleles = self.supportedAlleles\n",
    "        else:\n",
    "            #filter for supported alleles\n",
    "            alleles = filter(lambda a: a.name in self.supportedAlleles, alleles) \n",
    "        \n",
    "        result = {}\n",
    "        #now predict binding/non-binding for each peptide at random\n",
    "        for a in alleles:\n",
    "            result[a] = {}\n",
    "            for p in peptides:\n",
    "                if random.random() >= 0.5:\n",
    "                    result[a][p] = 1.0\n",
    "                else:\n",
    "                    result[a][p] = 0.0\n",
    "        \n",
    "        #create EpitopePredictionResult object. This is a multiindexed DataFrame \n",
    "        #with Peptide and Method as multi-index and alleles as columns\n",
    "        df_result = EpitopePredictionResult.from_dict(result)\n",
    "        df_result.index = pandas.MultiIndex.from_tuples([tuple((i,self.name)) for i in df_result.index],\n",
    "                                                        names=['Seq','Method'])\n",
    "        return df_result\n",
    "            \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets use our new predictor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A*02:01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq</th>\n",
       "      <th>Method</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(((F)), ((I)), ((A)), ((S)), ((N)), ((G)), ((V)), ((K)), ((L)))</th>\n",
       "      <th>random</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)), ((F)), ((V)))</th>\n",
       "      <th>random</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)), ((H)), ((I)))</th>\n",
       "      <th>random</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           A*02:01\n",
       "Seq                                                Method         \n",
       "(((F)), ((I)), ((A)), ((S)), ((N)), ((G)), ((V)... random        0\n",
       "(((L)), ((L)), ((G)), ((A)), ((T)), ((C)), ((M)... random        0\n",
       "(((S)), ((Y)), ((F)), ((P)), ((E)), ((I)), ((T)... random        1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EpitopePredictorFactory(\"random\").predict(peptides)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The predictor is now fully integrated and can be used in any context defined by Fred2."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
