<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Fred2.IO.MartsAdapter &mdash; Fred2  documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="top" title="Fred2  documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">Fred2  documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for Fred2.IO.MartsAdapter</h1><div class="highlight"><pre>
<span class="c"># This code is part of the Fred2 distribution and governed by its</span>
<span class="c"># license.  Please see the LICENSE file that should have been included</span>
<span class="c"># as part of this package.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. module:: IO.MartsAdapter</span>
<span class="sd">   :synopsis: BDB-Adapter for BioMart</span>
<span class="sd">.. moduleauthor:: walzer, schubert</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">urllib2</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="kn">from</span> <span class="nn">Fred2.IO.ADBAdapter</span> <span class="kn">import</span> <span class="n">ADBAdapter</span><span class="p">,</span> <span class="n">EAdapterFields</span>
<span class="kn">import</span> <span class="nn">MySQLdb</span>

<div class="viewcode-block" id="MartsAdapter"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter">[docs]</a><span class="k">class</span> <span class="nc">MartsAdapter</span><span class="p">(</span><span class="n">ADBAdapter</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">usr</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">pwd</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        used to fetch sequences from given RefSeq id&#39;s either from BioMart if no credentials given else from a MySQLdb</span>
<span class="sd">        :param usr: db user e.g. = &#39;ucsc_annot_query&#39;</span>
<span class="sd">        :param host: db host e.g. = &quot;pride&quot;</span>
<span class="sd">        :param pwd: pw for user e.g. = &#39;an0q3ry&#39;</span>
<span class="sd">        :param db: db on host e.g. = &quot;hg18_ucsc_annotation&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene_proxy</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">usr</span> <span class="ow">and</span> <span class="n">host</span> <span class="ow">and</span> <span class="n">pwd</span> <span class="ow">and</span> <span class="n">db</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">connection</span> <span class="o">=</span> <span class="n">MySQLdb</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="n">usr</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="n">host</span><span class="p">,</span> <span class="n">passwd</span><span class="o">=</span><span class="n">pwd</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="n">db</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">connection</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span> <span class="o">=</span> <span class="s">&quot;http://biomart.org/biomart/martservice?query=&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">new_biomart_url</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;http://central.biomart.org/biomart/martservice?query=&quot;&quot;&quot;</span>
<span class="c">#        self.biomart_url = &quot;&quot;&quot;http://biomart.org/biomart/martservice?query=&quot;&quot;&quot;</span>
<span class="c">#http://localhost:9000/biomart/martservice?query=%3C!DOCTYPE%20Query%3E%3CQuery%20client=%22biomartclient%22%20processor=%22TSV%22%20limit=%22-1%22%20header=%221%22%3E%3CDataset%20name=%22hsapiens_gene_ensembl%22%20config=%22gene_ensembl_config%22%3E%3CFilter%20name=%22uniprot_genename%22%20value=%22TP53%22%20filter_list=%22%22/%3E%3CFilter%20name=%22germ_line_variation_source%22%20value=%22dbSNP%22%20filter_list=%22%22/%3E%3CAttribute%20name=%22snp_ensembl_gene_id%22/%3E%3CAttribute%20name=%22snp_chromosome_name%22/%3E%3CAttribute%20name=%22snp_ensembl_transcript_id%22/%3E%3CAttribute%20name=%22snp_start_position%22/%3E%3CAttribute%20name=%22snp_ensembl_peptide_id%22/%3E%3CAttribute%20name=%22snp_end_position%22/%3E%3CAttribute%20name=%22snp_external_gene_name%22/%3E%3CAttribute%20name=%22snp_strand%22/%3E%3CAttribute%20name=%22source_description%22/%3E%3CAttribute%20name=%22germ_line_variation_source%22/%3E%3CAttribute%20name=%22allele%22/%3E%3CAttribute%20name=%22variation_name%22/%3E%3C/Dataset%3E%3C/Query%3E</span>
<span class="c">#        self.biomart_url = &quot;&quot;&quot;http://134.2.9.124/biomart/martservice?query=&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">        &lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;</span>
<span class="s">            &lt;!DOCTYPE Query&gt;</span>
<span class="s">            &lt;Query client=&quot;true&quot; processor=&quot;TSV&quot; limit=&quot;-1&quot; header=&quot;1&quot; uniqueRows = &quot;1&quot; &gt;</span>
<span class="s">                &lt;Dataset name=&quot;</span><span class="si">%s</span><span class="s">&quot; config=&quot;</span><span class="si">%s</span><span class="s">&quot;&gt;</span>
<span class="s">        &quot;&quot;&quot;</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;</span>
<span class="s">                &lt;/Dataset&gt;</span>
<span class="s">            &lt;/Query&gt;</span>
<span class="s">        &quot;&quot;&quot;</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;&lt;Filter name=&quot;</span><span class="si">%s</span><span class="s">&quot; value=&quot;</span><span class="si">%s</span><span class="s">&quot; filter_list=&quot;&quot;/&gt;&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span> <span class="o">=</span> <span class="s">&quot;&quot;&quot;&lt;Attribute name=&quot;</span><span class="si">%s</span><span class="s">&quot;/&gt;&quot;&quot;&quot;</span>

    <span class="c">#TODO: refactor ... function based on old code</span>
<div class="viewcode-block" id="MartsAdapter.get_product_sequence"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_product_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">get_product_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">product_refseq</span><span class="p">,</span> <span class="n">_db</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">_dataset</span><span class="o">=</span><span class="s">&#39;gene_ensembl_config&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        fetches product sequence for the given id</span>
<span class="sd">        :param product_refseq: given refseq id</span>
<span class="sd">        :return: list of dictionaries of the requested sequence, the respective strand and the associated gene name</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">product_refseq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">product_refseq</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">connection</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
            <span class="c">#query = &quot;SELECT * FROM sbs_ncbi_mrna WHERE id=&#39;%s&#39;;&quot;%(&#39;5&#39;)</span>
            <span class="n">query</span> <span class="o">=</span> <span class="s">&quot;SELECT refseq,seq FROM hg19_ucsc_annotation.refLink INNER JOIN sbs_ncbi_protein on protAcc=refseq WHERE mrnaAcc=&#39;</span><span class="si">%s</span><span class="s">&#39;;&quot;</span> <span class="o">%</span> <span class="p">(</span>
                <span class="n">product_refseq</span><span class="p">)</span>  <span class="c"># gives NP plus prot seq</span>
            <span class="c">#mysql -h genome-mysql.cse.ucsc.edu -u genome -D hg19 -N -A -e &#39;SELECT refseq,seq FROM hg19_ucsc_annotation.refLink INNER JOIN sbs_ncbi_protein on protAcc=refseq WHERE mrnaAcc=&#39;NM_002486&#39;;&#39;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
                    <span class="n">product_refseq</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">product_sequence</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;An Error occured while executing query: &quot;</span> <span class="o">+</span> <span class="n">query</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="k">return</span> <span class="bp">None</span>
            <span class="k">except</span> <span class="n">MySQLdb</span><span class="o">.</span><span class="n">Error</span><span class="p">,</span> <span class="n">message</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
                    <span class="s">&quot;An Error occured while executing query: &quot;</span> <span class="o">+</span> <span class="n">query</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">+</span> <span class="s">&quot;Error message: &quot;</span> <span class="o">+</span> <span class="n">message</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">product_refseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">product_sequence</span>
            <span class="k">return</span> <span class="p">[{</span><span class="n">product_refseq</span><span class="p">:</span> <span class="n">product_sequence</span><span class="p">}]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="k">if</span> <span class="n">product_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;NP_&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_peptide&quot;</span>
            <span class="k">elif</span> <span class="n">product_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;XP_&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_peptide_predicted&quot;</span>
            <span class="k">elif</span> <span class="n">product_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;ENS&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_peptide_id&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No correct product id: &quot;</span> <span class="o">+</span> <span class="n">product_refseq</span><span class="p">)</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span><span class="n">_dataset</span><span class="p">)</span> \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">product_refseq</span><span class="p">))</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;peptide&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;external_gene_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

            <span class="n">logging</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">rq_n</span><span class="p">)</span>

            <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
            <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">tsvselect</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">product_refseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Protein&quot;</span><span class="p">]</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">product_refseq</span><span class="p">]</span>  <span class="c"># TODO to SeqRecord</span>

    <span class="c">#TODO: refactor ... function based on old code</span></div>
<div class="viewcode-block" id="MartsAdapter.get_transcript_sequence"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_sequence">[docs]</a>    <span class="k">def</span> <span class="nf">get_transcript_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript_refseq</span><span class="p">,</span> <span class="n">_db</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">_dataset</span><span class="o">=</span><span class="s">&#39;gene_ensembl_config&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches transcript sequence for the given id</span>
<span class="sd">        :param transcript_refseq:</span>
<span class="sd">        :return: list of dictionary of the requested sequence, the respective strand and the associated gene name</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># TODO transcript_refseq to transcript_id, sniff which, query according</span>
        <span class="k">if</span> <span class="n">transcript_refseq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">connection</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
            <span class="c">#query = &quot;SELECT * FROM sbs_ncbi_mrna WHERE id=&#39;%s&#39;;&quot;%(&#39;5&#39;)</span>
            <span class="n">query</span> <span class="o">=</span> <span class="s">&quot;SELECT refseq,seq FROM hg19_ucsc_annotation.sbs_ncbi_mrna WHERE refseq LIKE &#39;&quot;</span> <span class="o">+</span> <span class="n">transcript_refseq</span> <span class="o">+</span> <span class="s">&quot;%&#39;&quot;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
                    <span class="n">transcript_refseq</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">transcript_sequence</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Ambiguous transcript refseq query: &quot;</span> <span class="o">+</span> <span class="n">transcript_refseq</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;An Error occured while executing query: &quot;</span> <span class="o">+</span> <span class="n">query</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="k">return</span> <span class="bp">None</span>
            <span class="k">except</span> <span class="n">MySQLdb</span><span class="o">.</span><span class="n">Error</span><span class="p">,</span> <span class="n">message</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;An Error occured while executing query: &quot;</span> <span class="o">+</span> <span class="n">query</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">+</span> <span class="s">&quot;Error message: &quot;</span> <span class="o">+</span> <span class="n">message</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript_sequence</span>
            <span class="k">return</span> <span class="p">[{</span><span class="n">transcript_refseq</span><span class="p">:</span> <span class="n">transcript_sequence</span><span class="p">}]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="k">if</span> <span class="n">transcript_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;NM_&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_mrna&quot;</span>
            <span class="k">elif</span> <span class="n">transcript_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;XM_&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_mrna_predicted&quot;</span>
            <span class="k">elif</span> <span class="n">transcript_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;ENS&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_transcript_id&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No correct transcript id: &quot;</span> <span class="o">+</span> <span class="n">transcript_refseq</span><span class="p">)</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">transcript_refseq</span><span class="p">))</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;coding&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;external_gene_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

            <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
            <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Coding sequence&#39;</span><span class="p">]</span>  <span class="c"># TODO to SeqRecord</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span>
</div>
<div class="viewcode-block" id="MartsAdapter.get_transcript_information"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_information">[docs]</a>    <span class="k">def</span> <span class="nf">get_transcript_information</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript_refseq</span><span class="p">,</span> <span class="n">_db</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">_dataset</span><span class="o">=</span><span class="s">&#39;gene_ensembl_config&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        It also already uses the Field-Enum for DBAdapters</span>

<span class="sd">        Fetches transcript sequence for the given id</span>
<span class="sd">        :param transcript_refseq:</span>
<span class="sd">        :return: list of dictionary of the requested sequence, the respective strand and the associated gene name</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># TODO transcript_refseq to transcript_id, sniff which, query according</span>
        <span class="k">if</span> <span class="n">transcript_refseq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">connection</span><span class="p">:</span>
            <span class="n">cur</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">connection</span><span class="o">.</span><span class="n">cursor</span><span class="p">()</span>
            <span class="c">#query = &quot;SELECT * FROM sbs_ncbi_mrna WHERE id=&#39;%s&#39;;&quot;%(&#39;5&#39;)</span>
            <span class="n">query</span> <span class="o">=</span> <span class="s">&quot;SELECT refseq,seq FROM hg19_ucsc_annotation.sbs_ncbi_mrna WHERE refseq LIKE &#39;&quot;</span> <span class="o">+</span> <span class="n">transcript_refseq</span> <span class="o">+</span> <span class="s">&quot;%&#39;&quot;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">cur</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">cur</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
                    <span class="n">transcript_refseq</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">transcript_sequence</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">result</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Ambiguous transcript refseq query: &quot;</span> <span class="o">+</span> <span class="n">transcript_refseq</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;An Error occured while executing query: &quot;</span> <span class="o">+</span> <span class="n">query</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="k">return</span> <span class="bp">None</span>
            <span class="k">except</span> <span class="n">MySQLdb</span><span class="o">.</span><span class="n">Error</span><span class="p">,</span> <span class="n">message</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;An Error occured while executing query: &quot;</span> <span class="o">+</span> <span class="n">query</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">+</span> <span class="s">&quot;Error message: &quot;</span> <span class="o">+</span> <span class="n">message</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript_sequence</span>
            <span class="k">return</span> <span class="p">[{</span><span class="n">transcript_refseq</span><span class="p">:</span> <span class="n">transcript_sequence</span><span class="p">}]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="k">if</span> <span class="n">transcript_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;NM_&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_mrna&quot;</span>
            <span class="k">elif</span> <span class="n">transcript_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;XM_&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_mrna_predicted&quot;</span>
            <span class="k">elif</span> <span class="n">transcript_refseq</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;ENS&#39;</span><span class="p">):</span>
                <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_transcript_id&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No correct transcript id: &quot;</span> <span class="o">+</span> <span class="n">transcript_refseq</span><span class="p">)</span>
                <span class="k">return</span> <span class="bp">None</span>
            <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">transcript_refseq</span><span class="p">))</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;coding&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

            <span class="k">print</span> <span class="s">&quot;Transcript information &quot;</span><span class="p">,</span><span class="n">rq_n</span>
            <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
            <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">tsvselect</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No entry found for ID </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transcript_refseq</span><span class="p">)</span>
                <span class="k">return</span> <span class="bp">None</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">SEQ</span><span class="p">:</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Coding sequence&#39;</span><span class="p">],</span>
                                                      <span class="n">EAdapterFields</span><span class="o">.</span><span class="n">GENE</span><span class="p">:</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;Associated Gene Name&#39;</span><span class="p">,</span> <span class="s">&quot;&quot;</span><span class="p">),</span>
                                                      <span class="n">EAdapterFields</span><span class="o">.</span><span class="n">STRAND</span><span class="p">:</span> <span class="s">&quot;-&quot;</span> <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Strand&#39;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">0</span>
                                                      <span class="k">else</span> <span class="s">&quot;+&quot;</span><span class="p">}</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span><span class="p">[</span><span class="n">transcript_refseq</span><span class="p">]</span>

    <span class="c">#TODO: refactor ... function based on old code</span></div>
<div class="viewcode-block" id="MartsAdapter.get_variant_gene"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_gene">[docs]</a>    <span class="k">def</span> <span class="nf">get_variant_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">_db</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">_dataset</span><span class="o">=</span><span class="s">&#39;gene_ensembl_config&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches the important db ids and names for given chromosomal location</span>
<span class="sd">        :param chrom: integer value of the chromosome in question</span>
<span class="sd">        :param start: integer value of the variation start position on given chromosome</span>
<span class="sd">        :param stop: integer value of the variation stop position on given chromosome</span>
<span class="sd">        :return: The respective gene name, i.e. the first one reported</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">chrom</span> <span class="o">+</span> <span class="n">start</span> <span class="o">+</span> <span class="n">stop</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_proxy</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_proxy</span><span class="p">[</span><span class="n">chrom</span> <span class="o">+</span> <span class="n">start</span> <span class="o">+</span> <span class="n">stop</span><span class="p">]</span>

        <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">))</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;start&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">start</span><span class="p">))</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;end&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop</span><span class="p">))</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">)</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

        <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">((</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">())</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
        <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">tsvselect</span> <span class="ow">and</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_proxy</span><span class="p">[</span><span class="n">chrom</span> <span class="o">+</span> <span class="n">start</span> <span class="o">+</span> <span class="n">stop</span><span class="p">]</span> <span class="o">=</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;UniProt Gene Name&#39;</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;UniProt Gene Name&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">chrom</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop</span><span class="p">)])</span> <span class="o">+</span> <span class="s">&#39; does not denote a known gene location&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="s">&#39;&#39;</span>
</div>
<div class="viewcode-block" id="MartsAdapter.get_transcript_information_from_protein_id"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_transcript_information_from_protein_id">[docs]</a>    <span class="k">def</span> <span class="nf">get_transcript_information_from_protein_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        It also already uses the Field-Enum for DBAdapters</span>

<span class="sd">        Fetches transcript sequence for the given id</span>
<span class="sd">        :param transcript_refseq:</span>
<span class="sd">        :return: list of dictionary of the requested sequence, the respective strand and the associated gene name</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_db</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_db&quot;</span><span class="p">,</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
        <span class="n">_dataset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_dataset&quot;</span><span class="p">,</span> <span class="s">&quot;gene_ensembl_config&quot;</span><span class="p">)</span>
        <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">db_id</span> <span class="o">=</span> <span class="s">&quot;&quot;</span>
        <span class="k">if</span> <span class="s">&quot;refseq&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_peptide&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;refseq&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;ensemble&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_peptide_id&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;ensemble&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_accid&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;uniprot_swissprot_accession&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_accid&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_gene&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span><span class="o">=</span> <span class="s">&quot;uniprot_swissprot&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_gene&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No correct transcript id&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">None</span>
        <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">db_id</span><span class="p">))</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;coding&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;external_gene_id&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

        <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
        <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">tsvselect</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No entry found for ID </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">db_id</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span><span class="p">[</span><span class="n">db_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">SEQ</span><span class="p">:</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Coding sequence&#39;</span><span class="p">],</span>
                                             <span class="n">EAdapterFields</span><span class="o">.</span><span class="n">GENE</span><span class="p">:</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Associated Gene Name&#39;</span><span class="p">],</span>
                                             <span class="n">EAdapterFields</span><span class="o">.</span><span class="n">STRAND</span><span class="p">:</span> <span class="s">&quot;-&quot;</span> <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Strand&#39;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">0</span>
                                             <span class="k">else</span> <span class="s">&quot;+&quot;</span><span class="p">}</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span><span class="p">[</span><span class="n">db_id</span><span class="p">]</span>
</div>
<div class="viewcode-block" id="MartsAdapter.get_variant_id_from_protein_id"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_id_from_protein_id">[docs]</a>    <span class="k">def</span> <span class="nf">get_variant_id_from_protein_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>  <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        returns all information needed to instantiate a variation</span>

<span class="sd">        :param trans_id: A transcript ID (either ENSAMBLE (ENS) or RefSeq (NM, XN)</span>
<span class="sd">        :return: list of dicts -- containing all information needed for a variant initialization</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_db</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_db&quot;</span><span class="p">,</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
        <span class="n">_dataset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_dataset&quot;</span><span class="p">,</span> <span class="s">&quot;gene_ensembl_config&quot;</span><span class="p">)</span>
        <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">db_id</span> <span class="o">=</span> <span class="s">&quot;&quot;</span>
        <span class="k">if</span> <span class="s">&quot;refseq&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_peptide&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;refseq&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;ensemble&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_peptide_id&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;ensemble&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_accid&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;uniprot_swissprot_accession&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_accid&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_gene&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span><span class="o">=</span> <span class="s">&quot;uniprot_swissprot&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_gene&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">war</span>
        <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">db_id</span><span class="p">))</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;germ_line_variation_source&quot;</span><span class="p">,</span> <span class="s">&quot;dbSNP&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;snp_ensembl_gene_id&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;variation_name&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;snp_chromosome_name&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;chromosome_location&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;allele&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;snp_strand&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;peptide_location&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>
        <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">new_biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
        <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">tsvselect</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No entry found for ID </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">db_id</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">None</span>

        <span class="k">return</span> <span class="n">tsvselect</span>
</div>
<div class="viewcode-block" id="MartsAdapter.get_variant_id_from_gene_id"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_id_from_gene_id">[docs]</a>    <span class="k">def</span> <span class="nf">get_variant_id_from_gene_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>  <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        returns all information needed to instantiate a variation</span>

<span class="sd">        :param trans_id: A transcript ID (either ENSAMBLE (ENS) or RefSeq (NM, XN)</span>
<span class="sd">        :return: list of dicts -- containing all information needed for a variant initialization</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_db</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_db&quot;</span><span class="p">,</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
        <span class="n">_dataset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_dataset&quot;</span><span class="p">,</span> <span class="s">&quot;gene_ensembl_config&quot;</span><span class="p">)</span>
        <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">db_id</span> <span class="o">=</span> <span class="s">&quot;&quot;</span>
        <span class="k">if</span> <span class="s">&quot;hgnc_symbol&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;hgnc_symbol&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;hgnc_symbol&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;ensemble&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_gene_id&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;ensemble&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;hgnc_id&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;hgnc_id&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;hgnc_id&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_gene&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;uniprot_swissprot&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_gene&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;external_gene_name&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;external_gene_name&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;external_gene_name&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">war</span>
        <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">db_id</span><span class="p">))</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;germ_line_variation_source&quot;</span><span class="p">,</span> <span class="s">&quot;dbSNP&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;snp_ensembl_gene_id&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;variation_name&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;snp_chromosome_name&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;chromosome_location&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;allele&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;snp_strand&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;peptide_location&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>
        <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">new_biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
        <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">tsvselect</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No entry found for ID </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">db_id</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">None</span>

        <span class="k">return</span> <span class="n">tsvselect</span>
</div>
<div class="viewcode-block" id="MartsAdapter.get_protein_sequence_from_protein_id"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_protein_sequence_from_protein_id">[docs]</a>    <span class="k">def</span> <span class="nf">get_protein_sequence_from_protein_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the protein sequence for a given protein ID that can either be refeseq, uniprot or ensamble id</span>

<span class="sd">        :param kwargs:</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">filter</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">db_id</span> <span class="o">=</span> <span class="s">&quot;&quot;</span>

        <span class="n">_db</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_db&quot;</span><span class="p">,</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
        <span class="n">_dataset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_dataset&quot;</span><span class="p">,</span> <span class="s">&quot;gene_ensembl_config&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s">&quot;refseq&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;refseq_peptide&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;refseq&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;ensemble&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;ensembl_peptide_id&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;ensemble&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_accid&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span> <span class="o">=</span> <span class="s">&quot;uniprot_swissprot_accession&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_accid&quot;</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&quot;swiss_gene&quot;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="nb">filter</span><span class="o">=</span> <span class="s">&quot;uniprot_swissprot&quot;</span>
            <span class="n">db_id</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;swiss_gene&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No correct transcript id&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">None</span>
        <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">db_id</span><span class="p">))</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="nb">filter</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;peptide&quot;</span><span class="p">)</span> \
               <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>
        <span class="k">print</span> <span class="n">rq_n</span>
        <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
        <span class="n">tsvselect</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">]</span>
        <span class="k">print</span> <span class="n">tsvselect</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">tsvselect</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;No entry found for ID </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">db_id</span><span class="p">)</span>
            <span class="k">return</span> <span class="bp">None</span>

        <span class="k">return</span> <span class="p">{</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">SEQ</span><span class="p">:</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Coding sequence&#39;</span><span class="p">],</span>
                                             <span class="n">EAdapterFields</span><span class="o">.</span><span class="n">GENE</span><span class="p">:</span> <span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Associated Gene Name&#39;</span><span class="p">],</span>
                                             <span class="n">EAdapterFields</span><span class="o">.</span><span class="n">STRAND</span><span class="p">:</span> <span class="s">&quot;-&quot;</span> <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">tsvselect</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&#39;Strand&#39;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">0</span>
                                             <span class="k">else</span> <span class="s">&quot;+&quot;</span><span class="p">}</span>

    <span class="c">#TODO: refactor ... function based on old code</span></div>
<div class="viewcode-block" id="MartsAdapter.get_all_variant_gene"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_all_variant_gene">[docs]</a>    <span class="k">def</span> <span class="nf">get_all_variant_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span><span class="n">_db</span><span class="o">=</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span> <span class="n">_dataset</span><span class="o">=</span><span class="s">&#39;gene_ensembl_config&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches the important db ids and names for given chromosomal location</span>
<span class="sd">        :param chrom: integer value of the chromosome in question</span>
<span class="sd">        :param start: integer value of the variation start position on given chromosome</span>
<span class="sd">        :param stop: integer value of the variation stop position on given chromosome</span>
<span class="sd">        :return: The respective gene name, i.e. the first one reported</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c">#TODO assert types</span>
        <span class="c">#&lt;!DOCTYPE Query&gt;&lt;Query client=&quot;true&quot; processor=&quot;TSV&quot; limit=&quot;-1&quot; header=&quot;1&quot;&gt;&lt;Dataset name=&quot;hsapiens_gene_ensembl&quot; config=&quot;gene_ensembl_config&quot;&gt;&lt;Filter name=&quot;chromosomal_region&quot; value=&quot;1:40367114:40367114,1:40702744:40702744,1:40705023:40705023,1:40771399:40771399,1:40777210:40777210,1:40881015:40881015,1:41235036:41235036,1:42048927:42048927,1:43002232:43002232,1:43308758:43308758,1:43393391:43630154,1:43772617:43772617,1:43772834:43772834&quot; filter_list=&quot;&quot;/&gt;&lt;Attribute name=&quot;uniprot_genename&quot;/&gt;&lt;/Dataset&gt;&lt;/Query&gt;</span>
        <span class="n">queries</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;genes&#39;</span><span class="p">][</span><span class="n">x</span><span class="p">:</span><span class="n">x</span><span class="o">+</span><span class="mi">300</span><span class="p">]))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;genes&#39;</span><span class="p">]),</span> <span class="mi">300</span><span class="p">)]</span>
        <span class="c"># if chrom + start + stop in self.gene_proxy:</span>
        <span class="c">#     return self.gene_proxy[chrom + start + stop]</span>
        <span class="c">#</span>
        <span class="c"># rq_n = self.biomart_head \</span>
        <span class="c">#     + self.biomart_filter%(&quot;chromosome_name&quot;, str(chrom))  \</span>
        <span class="c">#     + self.biomart_filter%(&quot;start&quot;, str(start))  \</span>
        <span class="c">#     + self.biomart_filter%(&quot;end&quot;, str(stop))  \</span>
        <span class="c">#     + self.biomart_attribute%(&quot;uniprot_genename&quot;)  \</span>
        <span class="c">#     + self.biomart_tail</span>
        <span class="c">#</span>
        <span class="c"># tsvreader = csv.DictReader((urllib2.urlopen(self.biomart_url+urllib2.quote(rq_n)).read()).splitlines(), dialect=&#39;excel-tab&#39;)</span>
        <span class="c"># tsvselect = [x for x in tsvreader]</span>
        <span class="c"># if tsvselect and tsvselect[0]:</span>
        <span class="c">#     self.gene_proxy[chrom + start + stop] = tsvselect[0][&#39;UniProt Gene Name&#39;]</span>
        <span class="c">#     return tsvselect[0][&#39;UniProt Gene Name&#39;]</span>
        <span class="c"># else:</span>
        <span class="c">#     logging.warn(&#39;,&#39;.join([str(chrom), str(start), str(stop)]) + &#39; does not denote a known gene location&#39;)</span>
        <span class="c">#     return &#39;&#39;</span>

    <span class="c">#TODO: refactor ... function based on old code</span></div>
<div class="viewcode-block" id="MartsAdapter.get_variant_ids"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_variant_ids">[docs]</a>    <span class="k">def</span> <span class="nf">get_variant_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches the important db ids and names for given gene _or_ chromosomal location. The former is recommended.</span>
<span class="sd">        AResult is a list of dicts with either of the tree combinations:</span>
<span class="sd">            - &#39;Ensembl Gene ID&#39;, &#39;Ensembl Transcript ID&#39;, &#39;Ensembl Protein ID&#39;</span>
<span class="sd">            - &#39;RefSeq Protein ID [e.g. NP_001005353]&#39;, &#39;RefSeq mRNA [e.g. NM_001195597]&#39;, first triplet</span>
<span class="sd">            - &#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;, &#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;, first triplet</span>
<span class="sd">        :keyword &#39;chrom&#39;: integer value of the chromosome in question</span>
<span class="sd">        :keyword &#39;start&#39;: integer value of the variation start position on given chromosome</span>
<span class="sd">        :keyword &#39;stop&#39;: integer value of the variation stop position on given chromosome</span>
<span class="sd">        :keyword &#39;gene&#39;: string value of the gene of variation</span>
<span class="sd">        :keyword &#39;transcript_id&#39;: string value of the gene of variation</span>
<span class="sd">        :return: The list of dicts of entries with transcript and protein ids (either NM+NP or XM+XP)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># TODO type assessment</span>
        <span class="n">_db</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_db&quot;</span><span class="p">,</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
        <span class="n">_dataset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_dataset&quot;</span><span class="p">,</span> <span class="s">&quot;gene_ensembl_config&quot;</span><span class="p">)</span>
        <span class="n">ensemble_only</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="n">query</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">4</span> <span class="ow">and</span> <span class="s">&#39;chrom&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;start&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;stop&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;ensemble_only&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">ensemble_only</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;ensemble_only&#39;</span><span class="p">]</span>
            <span class="n">query</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;chrom&#39;</span><span class="p">])</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;start&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;start&#39;</span><span class="p">])</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;end&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;stop&#39;</span><span class="p">])</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span> <span class="ow">and</span> <span class="s">&#39;chrom&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;start&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;stop&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">query</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;chrom&#39;</span><span class="p">])</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;start&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;start&#39;</span><span class="p">])</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;end&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;stop&#39;</span><span class="p">])</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="s">&#39;gene&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;ensemble_only&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">ensemble_only</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;ensemble_only&#39;</span><span class="p">]</span>
            <span class="n">query</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;gene&#39;</span><span class="p">])</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="s">&#39;transcript_id&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="s">&#39;ensemble_only&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">ensemble_only</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;ensemble_only&#39;</span><span class="p">]</span>
            <span class="n">query</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_transcript_id&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;transcript_id&#39;</span><span class="p">])</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="s">&#39;gene&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">query</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">,</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;gene&#39;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;wrong arguments to get_variant_ids&quot;</span><span class="p">)</span>

        <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
            <span class="o">+</span> <span class="n">query</span> \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">)</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">)</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_peptide_id&quot;</span><span class="p">)</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_transcript_id&quot;</span><span class="p">)</span>  \
            <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ensemble_only</span><span class="p">:</span>
            <span class="n">rq_n</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_mrna&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_peptide&quot;</span><span class="p">)</span>
        <span class="n">rq_n</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_swissprot&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

        <span class="c"># logging.warning(rq_n)</span>

        <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">((</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">())</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ensemble_only</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Transcript ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Protein ID&#39;</span><span class="p">]:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Transcript ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Protein ID&#39;</span><span class="p">]:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span>
                  <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq Protein ID [e.g. NP_001005353]&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq mRNA [e.g. NM_001195597]&#39;</span><span class="p">]}</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">ensemble_only</span><span class="p">:</span>
            <span class="n">rq_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span> \
                <span class="o">+</span> <span class="n">query</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_peptide_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_transcript_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_peptide_predicted&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_mrna_predicted&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

            <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">((</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_x</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">())</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>

            <span class="n">result2</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Transcript ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Protein ID&#39;</span><span class="p">]:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span>
                       <span class="k">if</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;</span><span class="p">])</span>
                       <span class="ow">or</span> <span class="p">(</span><span class="ow">not</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;</span><span class="p">])}</span>
            <span class="n">result</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result2</span><span class="p">)</span>

        <span class="n">g</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
            <span class="k">if</span> <span class="s">&#39;uniprot_swissprot&#39;</span> <span class="ow">in</span> <span class="n">v</span><span class="p">:</span>
                <span class="n">g</span> <span class="o">=</span> <span class="n">v</span><span class="p">[</span><span class="s">&#39;uniprot_swissprot&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids_proxy</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">result</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>

    <span class="c">#TODO: refactor ... function based on old code</span></div>
<div class="viewcode-block" id="MartsAdapter.get_all_variant_ids"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.MartsAdapter.MartsAdapter.get_all_variant_ids">[docs]</a>    <span class="k">def</span> <span class="nf">get_all_variant_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>  <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Fetches the important db ids and names for given gene _or_ chromosomal location. The former is recommended.</span>
<span class="sd">        AResult is a list of dicts with either of the tree combinations:</span>
<span class="sd">            - &#39;Ensembl Gene ID&#39;, &#39;Ensembl Transcript ID&#39;, &#39;Ensembl Protein ID&#39;</span>
<span class="sd">            - &#39;RefSeq Protein ID [e.g. NP_001005353]&#39;, &#39;RefSeq mRNA [e.g. NM_001195597]&#39;, first triplet</span>
<span class="sd">            - &#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;, &#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;, first triplet</span>
<span class="sd">        :keyword &#39;locations&#39;: list of locations as triplets of integer values representing (chrom, start, stop)</span>
<span class="sd">        :keyword &#39;genes&#39;: list of genes as string value of the genes of variation</span>
<span class="sd">        :return: The list of dicts of entries with transcript and protein ids (either NM+NP or XM+XP)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_db</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_db&quot;</span><span class="p">,</span><span class="s">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">)</span>
        <span class="n">_dataset</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;_dataset&quot;</span><span class="p">,</span> <span class="s">&quot;gene_ensembl_config&quot;</span><span class="p">)</span>
        <span class="n">end_result</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
        <span class="c"># TODO type assessment</span>
        <span class="n">ensemble_only</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="n">query</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="s">&#39;ensemble_only&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">ensemble_only</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;ensemble_only&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="s">&#39;locations&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="k">pass</span>
            <span class="c">#TODO</span>
            <span class="c"># query = self.biomart_filter%(&quot;chromosome_name&quot;, kwargs[&#39;chrom&#39;])  \</span>
            <span class="c">#         + self.biomart_filter%(&quot;start&quot;, kwargs[&#39;start&#39;])  \</span>
            <span class="c">#         + self.biomart_filter%(&quot;end&quot;, kwargs[&#39;stop&#39;])</span>
        <span class="k">elif</span> <span class="s">&#39;genes&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">:</span>
            <span class="n">queries</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;genes&#39;</span><span class="p">][</span><span class="n">x</span><span class="p">:</span><span class="n">x</span><span class="o">+</span><span class="mi">250</span><span class="p">]))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;genes&#39;</span><span class="p">]),</span> <span class="mi">250</span><span class="p">)]</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&#39;***&#39;</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_filter</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;genes&#39;</span><span class="p">][</span><span class="n">x</span><span class="p">:</span><span class="n">x</span><span class="o">+</span><span class="mi">250</span><span class="p">]))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;genes&#39;</span><span class="p">]),</span> <span class="mi">250</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&quot;wrong arguments to get_variant_ids&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">query</span> <span class="ow">in</span> <span class="n">queries</span><span class="p">:</span>
            <span class="n">rq_n</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span><span class="n">_dataset</span><span class="p">)</span> \
                <span class="o">+</span> <span class="n">query</span> \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_peptide_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_transcript_id&quot;</span><span class="p">)</span>  \
                <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ensemble_only</span><span class="p">:</span>
                <span class="n">rq_n</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_mrna&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_peptide&quot;</span><span class="p">)</span>
            <span class="n">rq_n</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>
            <span class="c"># rq_n += self.biomart_attribute%(&quot;uniprot_swissprot&quot;) + self.biomart_tail</span>

            <span class="c"># logging.warning(rq_n)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">((</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_n</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">())</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">ensemble_only</span><span class="p">:</span>
                    <span class="n">result</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Transcript ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Protein ID&#39;</span><span class="p">]:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">}</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">result</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Transcript ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Protein ID&#39;</span><span class="p">]:</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span>
                          <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq Protein ID [e.g. NP_001005353]&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq mRNA [e.g. NM_001195597]&#39;</span><span class="p">]}</span>
                <span class="n">end_result</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s">&#39;Bad Mart Query: &#39;</span><span class="o">+</span><span class="n">rq_n</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">ensemble_only</span><span class="p">:</span>
                <span class="n">rq_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_head</span><span class="o">%</span><span class="p">(</span><span class="n">_db</span><span class="p">,</span> <span class="n">_dataset</span><span class="p">)</span> \
                    <span class="o">+</span> <span class="n">query</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;uniprot_genename&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_gene_id&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_peptide_id&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;ensembl_transcript_id&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_peptide_predicted&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;refseq_mrna_predicted&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_attribute</span><span class="o">%</span><span class="p">(</span><span class="s">&quot;strand&quot;</span><span class="p">)</span>  \
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">biomart_tail</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="n">tsvreader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">((</span><span class="n">urllib2</span><span class="o">.</span><span class="n">urlopen</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">biomart_url</span><span class="o">+</span><span class="n">urllib2</span><span class="o">.</span><span class="n">quote</span><span class="p">(</span><span class="n">rq_x</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">())</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">dialect</span><span class="o">=</span><span class="s">&#39;excel-tab&#39;</span><span class="p">)</span>

                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tsvreader</span><span class="p">:</span>
                        <span class="k">if</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s">&#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;</span><span class="p">]):</span>
                            <span class="n">end_result</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Gene ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Transcript ID&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">&#39;Ensembl Protein ID&#39;</span><span class="p">],</span> <span class="n">x</span><span class="p">)</span>
                    <span class="c"># result2 = {x[&#39;Ensembl Gene ID&#39;]+x[&#39;Ensembl Transcript ID&#39;]+x[&#39;Ensembl Protein ID&#39;]: x for x in tsvreader</span>
                    <span class="c">#            if (x[&#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;] and x[&#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;])</span>
                    <span class="c">#     this line is a troublemaker and does not help       or (not x[&#39;RefSeq Predicted Protein ID [e.g. XP_001720922]&#39;] and not x[&#39;RefSeq mRNA predicted [e.g. XM_001125684]&#39;])}</span>
                    <span class="c"># end_result.setdefault(result2)</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s">&#39;Bad Mart Query: &#39;</span><span class="o">+</span><span class="n">rq_n</span><span class="p">)</span>

        <span class="c"># g = None</span>
        <span class="c"># for k, v in result.iteritems():</span>
        <span class="c">#     if &#39;uniprot_swissprot&#39; in v:</span>
        <span class="c">#         g = v[&#39;uniprot_swissprot&#39;]</span>
        <span class="c"># self.ids_proxy[g] = result.values()</span>
        <span class="k">return</span> <span class="n">end_result</span><span class="o">.</span><span class="n">values</span><span class="p">()</span></div></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Author.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.2</a>
      
    </div>

    

    
  </body>
</html>