<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Fred2.IO.FileReader &mdash; FRED 2.01 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '2.01',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="FRED 2.01 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">FRED 2.01 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for Fred2.IO.FileReader</h1><div class="highlight"><pre>
<span class="c"># This code is part of the Fred2 distribution and governed by its</span>
<span class="c"># license.  Please see the LICENSE file that should have been included</span>
<span class="c"># as part of this package.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. module:: Reader</span>
<span class="sd">   :synopsis: Module handles reading of files. line reading, FASTA reading, annovar reading</span>
<span class="sd">.. moduleauthor:: brachvogel, schubert</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">from</span> <span class="nn">Bio.SeqIO.FastaIO</span> <span class="kn">import</span> <span class="n">SimpleFastaParser</span>

<span class="kn">from</span> <span class="nn">Fred2.Core.Allele</span> <span class="kn">import</span> <span class="n">Allele</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Peptide</span> <span class="kn">import</span> <span class="n">Peptide</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Protein</span> <span class="kn">import</span> <span class="n">Protein</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Transcript</span> <span class="kn">import</span> <span class="n">Transcript</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Variant</span> <span class="kn">import</span> <span class="n">Variant</span><span class="p">,</span> <span class="n">VariationType</span><span class="p">,</span> <span class="n">MutationSyntax</span>


<span class="k">def</span> <span class="nf">__check_type</span><span class="p">(</span><span class="nb">type</span><span class="p">,</span> <span class="n">allowed</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;Peptide&#39;</span><span class="p">,</span> <span class="s">&#39;Protein&#39;</span><span class="p">,</span><span class="s">&#39;Transcript&#39;</span><span class="p">]):</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">type</span> <span class="ow">in</span> <span class="n">allowed</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;An invalid sequence object type was specified for </span><span class="se">\</span>
<span class="s">parsing a FASTA file. Type was </span><span class="si">%s</span><span class="s"> but allowed types are: </span><span class="si">%s</span><span class="s">.&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">type</span><span class="p">,</span> <span class="n">allowed</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">__create_single_sequ</span><span class="p">(</span><span class="n">sequ</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="nb">type</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s">&quot;Peptide&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Peptide</span><span class="p">(</span><span class="n">sequ</span><span class="p">)</span>

    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s">&quot;Protein&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Protein</span><span class="p">(</span><span class="n">sequ</span><span class="p">,</span> <span class="s">&quot;unknown&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="p">)</span>

    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s">&quot;Transcript&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Transcript</span><span class="p">(</span><span class="s">&quot;unkown&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">sequ</span><span class="p">)</span>

    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s">&quot;Allele&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">Allele</span><span class="p">(</span><span class="n">sequ</span><span class="p">)</span>


<span class="c">####################################</span>
<span class="c">#       F A S T A  -  R E A D E R</span>
<span class="c">####################################</span>
<div class="viewcode-block" id="read_fasta"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.FileReader.read_fasta">[docs]</a><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="o">*</span><span class="n">argv</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read a (couple of) peptide, protein or rna sequence from a FASTA file.</span>
<span class="sd">    User needs to specify the correct type of the underlying sequences. It can</span>
<span class="sd">    either be: Peptide, Protein or Transcript (for RNA).</span>

<span class="sd">    :param *argv: a string list of absolute file names of the FASTA files to be</span>
<span class="sd">                  read. Give as: get_sequence(*[&quot;path/name1&quot;, &quot;path/name2&quot;]).</span>
<span class="sd">                  This field is required!</span>
<span class="sd">    :param **kwargs: optional. Use get_sequence(*[&quot;path/name1&quot;, &quot;path/name2&quot;,</span>
<span class="sd">                     type=&quot;Protein&quot;). Possible types are Peptide&#39;, &#39;Protein&#39;</span>
<span class="sd">                     and &#39;Transcript&#39;.</span>
<span class="sd">    :returns: (list(SequenceType)) -- a list of the specified sequence type</span>
<span class="sd">              derived from the FASTA file sequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_type</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;type&quot;</span><span class="p">,</span> <span class="s">&quot;Peptide&quot;</span><span class="p">)</span>

    <span class="n">__check_type</span><span class="p">(</span><span class="n">_type</span><span class="p">)</span>
    <span class="k">print</span> <span class="n">kwargs</span>
    <span class="n">collect</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c"># open all specified files:</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">argv</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="c"># iterate over all FASTA entries:</span>
            <span class="k">for</span> <span class="n">_id</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SimpleFastaParser</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
                <span class="c"># generate element:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">collect</span><span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span> <span class="o">=</span> <span class="n">_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;|&quot;</span><span class="p">)[</span><span class="n">kwargs</span><span class="p">[</span><span class="s">&quot;id_position&quot;</span><span class="p">]]</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">collect</span><span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span> <span class="o">=</span> <span class="n">_id</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">__create_single_sequ</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">_id</span><span class="p">,</span> <span class="n">_type</span><span class="p">)</span> \
            <span class="k">for</span> <span class="n">seq</span><span class="p">,</span> <span class="n">_id</span> <span class="ow">in</span> <span class="n">collect</span><span class="o">.</span><span class="n">items</span><span class="p">()]</span>


<span class="c">####################################</span>
<span class="c">#       L I N E  -  R E A D E R</span>
<span class="c">####################################</span></div>
<div class="viewcode-block" id="read_lines"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.FileReader.read_lines">[docs]</a><span class="k">def</span> <span class="nf">read_lines</span><span class="p">(</span><span class="o">*</span><span class="n">argv</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read a sequence directly from a line. User needs to manually specify the </span>
<span class="sd">    correct type of the underlying sequences. It can either be: </span>
<span class="sd">    Peptide, Protein or Transcript (for RNA).</span>

<span class="sd">    :param *argv: a list of strings of absolute file names of the FASTA files </span>
<span class="sd">                  that are to be read. Give as: </span>
<span class="sd">                  get_sequence(*[&quot;path/name1&quot;, &quot;path/name2&quot;]).</span>
<span class="sd">                  This field is required!</span>
<span class="sd">    :param **kwargs: optional. Use get_sequence(*[&quot;path/name1&quot;, &quot;path/name2&quot;,</span>
<span class="sd">                     type=&quot;Protein&quot;). Possible types are Peptide&#39;, &#39;Protein&#39;</span>
<span class="sd">                     and &#39;Transcript&#39;.</span>
<span class="sd">    :returns: (list(SequenceType)) -- a list of the specified sequence type</span>
<span class="sd">              derived from the FASTA file sequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_type</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;type&quot;</span><span class="p">,</span> <span class="s">&quot;Peptide&quot;</span><span class="p">)</span>

    <span class="n">__check_type</span><span class="p">(</span><span class="n">_type</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;Peptide&#39;</span><span class="p">,</span> <span class="s">&#39;Protein&#39;</span><span class="p">,</span> <span class="s">&#39;Transcript&#39;</span><span class="p">,</span> <span class="s">&#39;Allele&#39;</span><span class="p">])</span>

    <span class="n">collect</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">argv</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
            <span class="c"># iterate over all lines:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">handle</span><span class="p">:</span>
                <span class="c"># generate element:</span>
                <span class="n">collect</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">upper</span><span class="p">())</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">__create_single_sequ</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s">&quot;generic_id_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">_id</span><span class="p">),</span> <span class="n">_type</span><span class="p">)</span> \
            <span class="k">for</span> <span class="n">_id</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">collect</span><span class="p">)]</span>


<span class="c">#####################################</span>
<span class="c">#       A N N O V A R  -  R E A D E R</span>
<span class="c">#####################################</span></div>
<div class="viewcode-block" id="read_annovar_exonic"><a class="viewcode-back" href="../../../Fred2.IO.html#Fred2.IO.FileReader.read_annovar_exonic">[docs]</a><span class="k">def</span> <span class="nf">read_annovar_exonic</span><span class="p">(</span><span class="n">annovar_file</span><span class="p">,</span> <span class="n">gene_filter</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">experimentalDesig</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads an gene-based ANNOVAR output file and generates Variant objects containing</span>
<span class="sd">    all annotated transcript ids an outputs a list variants.</span>

<span class="sd">    :param str annovar_file: The path ot the ANNOVAR file</span>
<span class="sd">    :param list(str) gene_filter: A list of gene names of interest (only variants associated</span>
<span class="sd">                                  with these genes are generated)</span>
<span class="sd">    :return: list(AVariant) -- List of variants fully annotated</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">re</span>

    <span class="nb">vars</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_filter</span> <span class="o">=</span> <span class="n">gene_filter</span> <span class="k">if</span> <span class="n">gene_filter</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="k">else</span> <span class="p">[]</span>
    <span class="n">nm_var_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">nm_gene_name</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c">#fgd3:nm_001083536:exon6:c.g823a:p.v275i,fgd3:nm_001286993:exon6:c.g823a:p.v275i,fgd3:nm_033086:exon6:c.g823a:p.v275i</span>
    <span class="c">#RE = re.compile(&quot;\w+:(\w+):exon\d+:c.(\D*)(\d+)_*(\d*)(\D\w*):p.\w+:\D*-&gt;\D*:(\D).*?,&quot;)</span>
    <span class="c">#RE = re.compile(&quot;\w+:(\w+):exon\d+:c.(\D*)(\d+)_*(\d*)(\D\w*):p.(\D*)(\d+)_*(\d*)(\D\w*):(\D).*?,&quot;)</span>
    <span class="n">RE</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">&quot;((\w+):exon\d+:c.\D*(\d+)\D\w*:p.\D*(\d+)\D\w*)&quot;</span><span class="p">)</span>
    <span class="n">type_mapper</span> <span class="o">=</span> <span class="p">{(</span><span class="s">&#39;synonymous&#39;</span><span class="p">,</span> <span class="s">&#39;snv&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;nonsynonymous&#39;</span><span class="p">,</span> <span class="s">&#39;snv&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;stoploss&#39;</span><span class="p">,</span> <span class="s">&#39;snv&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;stopgain&#39;</span><span class="p">,</span> <span class="s">&#39;snv&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">SNP</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;nonframeshift&#39;</span><span class="p">,</span> <span class="s">&#39;deletion&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">DEL</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;frameshift&#39;</span><span class="p">,</span> <span class="s">&#39;deletion&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">FSDEL</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;nonframeshift&#39;</span><span class="p">,</span> <span class="s">&#39;insertion&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">INS</span><span class="p">,</span>
                   <span class="p">(</span><span class="s">&#39;frameshift&#39;</span><span class="p">,</span> <span class="s">&#39;insertion&#39;</span><span class="p">):</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">FSINS</span><span class="p">}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">annovar_file</span><span class="p">,</span> <span class="s">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
            <span class="nb">id</span><span class="p">,</span> <span class="nb">type</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">genome_start</span><span class="p">,</span> <span class="n">genome_stop</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">,</span> <span class="n">zygos</span><span class="o">=</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span>
                                                                              <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)[:</span><span class="mi">9</span><span class="p">])</span>
            <span class="c">#print ref, alt</span>

            <span class="c">#test if its a intersting snp</span>

            <span class="n">gene</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gene_filter</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_filter</span><span class="p">):</span>
                <span class="k">continue</span>

            <span class="k">if</span> <span class="n">gene</span> <span class="o">==</span> <span class="s">&quot;UNKNOWN&quot;</span><span class="p">:</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Skipping UNKWON gene&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

           <span class="c"># print &quot;Debug &quot;, gene, type.split(),id</span>
            <span class="c">#print &quot;Debug &quot;, line, RE.findall(line), type, zygos</span>
            <span class="n">coding</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">nm_id_pos</span> <span class="ow">in</span> <span class="n">RE</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">line</span><span class="p">):</span>
                <span class="n">mutation_string</span><span class="p">,</span> <span class="n">nm_id</span><span class="p">,</span> <span class="n">trans_pos</span><span class="p">,</span> <span class="n">prot_start</span> <span class="o">=</span> <span class="n">nm_id_pos</span>
                <span class="c">#print &quot;Debug &quot;,nm_id_pos</span>

                <span class="n">nm_id</span> <span class="o">=</span> <span class="n">nm_id</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">_</span><span class="p">,</span><span class="n">_</span><span class="p">,</span><span class="n">trans_coding</span><span class="p">,</span><span class="n">prot_coding</span> <span class="o">=</span> <span class="n">mutation_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">)</span>
                <span class="n">coding</span><span class="p">[</span><span class="n">nm_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">MutationSyntax</span><span class="p">(</span><span class="n">nm_id</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="n">trans_pos</span><span class="p">),</span><span class="nb">int</span><span class="p">(</span><span class="n">prot_start</span><span class="p">),</span><span class="n">trans_coding</span><span class="p">,</span><span class="n">prot_coding</span><span class="p">)</span>

            <span class="n">ty</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">type</span><span class="o">.</span><span class="n">split</span><span class="p">())</span>

            <span class="nb">vars</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">Variant</span><span class="p">(</span><span class="nb">id</span><span class="p">,</span> <span class="n">type_mapper</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ty</span><span class="p">,</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">UNKNOWN</span><span class="p">),</span> <span class="n">chrom</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">genome_start</span><span class="p">),</span> <span class="n">ref</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span>
                        <span class="n">alt</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="n">coding</span><span class="p">,</span> <span class="n">zygos</span> <span class="o">==</span> <span class="s">&quot;hom&quot;</span><span class="p">,</span> <span class="n">ty</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s">&quot;synonymous&quot;</span><span class="p">,</span>
                        <span class="n">experimentalDesign</span><span class="o">=</span><span class="n">experimentalDesig</span><span class="p">))</span>
    <span class="k">return</span> <span class="nb">vars</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">FRED 2.01 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Schbert Walzer Brachvogel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.
    </div>
  </body>
</html>