<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Fred2.Core.Generator &mdash; Fred2  documentation</title>
    
    <link rel="stylesheet" href="../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="top" title="Fred2  documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9">

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">Fred2  documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for Fred2.Core.Generator</h1><div class="highlight"><pre>
<span class="c"># This code is part of the Fred2 distribution and governed by its</span>
<span class="c"># license.  Please see the LICENSE file that should have been included</span>
<span class="c"># as part of this package.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. module:: Core.Generator</span>
<span class="sd">   :synopsis: Contains functions to transform variants to transcripts and </span>
<span class="sd">              proteins to peptides. Transition of transcripts to proteins</span>
<span class="sd">              is done via :meth:`~Fred2.Core.Transcript.translate`</span>
<span class="sd">.. moduleauthor:: schubert</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">collections</span>

<span class="kn">from</span> <span class="nn">Fred2.Core.Base</span> <span class="kn">import</span> <span class="n">COMPLEMENT</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Protein</span> <span class="kn">import</span> <span class="n">Protein</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Peptide</span> <span class="kn">import</span> <span class="n">Peptide</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Transcript</span> <span class="kn">import</span> <span class="n">Transcript</span>
<span class="kn">from</span> <span class="nn">Fred2.Core.Variant</span> <span class="kn">import</span> <span class="n">VariationType</span>
<span class="kn">from</span> <span class="nn">Fred2.IO.ADBAdapter</span> <span class="kn">import</span> <span class="n">ADBAdapter</span><span class="p">,</span> <span class="n">EAdapterFields</span>

<span class="c">################################################################################</span>
<span class="c">#Private module functions. It should not be possible to import these!</span>


<span class="k">def</span> <span class="nf">_update_var_offset</span><span class="p">(</span><span class="nb">vars</span><span class="p">,</span> <span class="n">transId_old</span><span class="p">,</span> <span class="n">transId_new</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    :param var:</span>
<span class="sd">    :param transId_old:</span>
<span class="sd">    :param transId_new:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c">#TODO as this is an manipulator doing an assignment a clearer signature nomenclature would be left- and righthand_side</span>
    <span class="c">#TODO doc for intended use case</span>
    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">:</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">offsets</span><span class="p">[</span><span class="n">transId_old</span><span class="p">]</span>
        <span class="n">v</span><span class="o">.</span><span class="n">offsets</span><span class="p">[</span><span class="n">transId_new</span><span class="p">]</span> <span class="o">=</span> <span class="n">offset</span>


<span class="c">#TODO for _incorp... : shouldnt the offset be determined by the variant and _NOT_ be given as argument - otherwise the argument name should be renamed to something like offset_from_previous or transcript_current_offset</span>

<span class="k">def</span> <span class="nf">_incorp_snp</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">var</span><span class="p">,</span> <span class="n">transId</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    incorporates a snp into the given transcript sequence</span>

<span class="sd">    :param list(char) seq: transcript sequence as a list</span>
<span class="sd">    :param Variant var: the snp variant to incorporate</span>
<span class="sd">    :param str transId: the transcript ID of seq</span>
<span class="sd">    :param int offset: the offset which has to be added onto the transcript</span>
<span class="sd">                       position of the variant</span>
<span class="sd">    :return: (list,int) the modified seq, the modified offset</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">SNP</span> <span class="o">!=</span> <span class="n">var</span><span class="o">.</span><span class="n">type</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> is not a SNP&quot;</span><span class="o">%</span><span class="nb">str</span><span class="p">(</span><span class="n">var</span><span class="p">))</span>
    <span class="n">var</span><span class="o">.</span><span class="n">offsets</span><span class="p">[</span><span class="n">transId</span><span class="p">]</span> <span class="o">=</span> <span class="n">offset</span>

    <span class="c">#print transId, len(seq), var.get_transcript_position(transId)-1</span>
    <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">var</span><span class="o">.</span><span class="n">get_transcript_position</span><span class="p">(</span><span class="n">transId</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="n">var</span><span class="o">.</span><span class="n">ref</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;For </span><span class="si">%s</span><span class="s"> bp dos not mmatch ref of assigned variant </span><span class="si">%s</span><span class="s">. Pos </span><span class="si">%i</span><span class="s">, var ref </span><span class="si">%s</span><span class="s">, seq ref </span><span class="si">%s</span><span class="s"> &quot;</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">transId</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">var</span><span class="p">),</span> <span class="n">var</span><span class="o">.</span><span class="n">get_transcript_position</span><span class="p">(</span><span class="n">transId</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">var</span><span class="o">.</span><span class="n">ref</span><span class="p">,</span>
        <span class="n">seq</span><span class="p">[</span><span class="n">var</span><span class="o">.</span><span class="n">get_transcript_position</span><span class="p">(</span><span class="n">transId</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]))</span>

    <span class="n">seq</span><span class="p">[</span><span class="n">var</span><span class="o">.</span><span class="n">get_transcript_position</span><span class="p">(</span><span class="n">transId</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">var</span><span class="o">.</span><span class="n">obs</span>

    <span class="k">return</span> <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span>


<span class="k">def</span> <span class="nf">_incorp_insertion</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">var</span><span class="p">,</span> <span class="n">transId</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    incorporates an insertion into the given transcript sequence</span>

<span class="sd">    :param list(char) seq: (list) transcript sequence as a list</span>
<span class="sd">    :param Variant var: the snp variant to incorporate</span>
<span class="sd">    :param str transId: the transcript ID of seq</span>
<span class="sd">    :param int offset: the offset which has to be added onto the transcript</span>
<span class="sd">                       position of the variant</span>
<span class="sd">    :return: (list,int) modified sequence, the modified offset</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">var</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">VariationType</span><span class="o">.</span><span class="n">INS</span><span class="p">,</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">FSINS</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> is not a insertion&quot;</span><span class="o">%</span><span class="nb">str</span><span class="p">(</span><span class="n">var</span><span class="p">))</span>

    <span class="n">var</span><span class="o">.</span><span class="n">offsets</span><span class="p">[</span><span class="n">transId</span><span class="p">]</span> <span class="o">=</span> <span class="n">offset</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">var</span><span class="o">.</span><span class="n">get_transcript_position</span><span class="p">(</span><span class="n">transId</span><span class="p">)</span>

    <span class="n">seq</span><span class="p">[</span><span class="n">pos</span><span class="p">:</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">var</span><span class="o">.</span><span class="n">obs</span>
    <span class="k">return</span> <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">var</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_incorp_deletion</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">var</span><span class="p">,</span> <span class="n">transId</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    incorporates a deletion into the given transcript sequence</span>

<span class="sd">    :param list(char) seq: transcript sequence as a list</span>
<span class="sd">    :param Variant var: the snp variant to incorporate</span>
<span class="sd">    :param str transId: the transcript ID of seq</span>
<span class="sd">    :param int offset: the offset which has to be added onto the transcript </span>
<span class="sd">                       position of the variant</span>
<span class="sd">    :return: (list,int) -- modified sequence, the modified offset</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">var</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">VariationType</span><span class="o">.</span><span class="n">DEL</span><span class="p">,</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">FSDEL</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s"> is not a deletion&quot;</span><span class="o">%</span><span class="nb">str</span><span class="p">(</span><span class="n">var</span><span class="p">))</span>

    <span class="n">var</span><span class="o">.</span><span class="n">offsets</span><span class="p">[</span><span class="n">transId</span><span class="p">]</span> <span class="o">=</span> <span class="n">offset</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">var</span><span class="o">.</span><span class="n">get_transcript_position</span><span class="p">(</span><span class="n">transId</span><span class="p">)</span>
    <span class="n">s</span> <span class="o">=</span> <span class="nb">slice</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">pos</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">var</span><span class="o">.</span><span class="n">ref</span><span class="p">))</span>
    <span class="k">del</span> <span class="n">seq</span><span class="p">[</span><span class="n">s</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">var</span><span class="o">.</span><span class="n">ref</span><span class="p">)</span>


<span class="n">_incorp</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">VariationType</span><span class="o">.</span><span class="n">DEL</span><span class="p">:</span> <span class="n">_incorp_deletion</span><span class="p">,</span>
            <span class="n">VariationType</span><span class="o">.</span><span class="n">FSDEL</span><span class="p">:</span> <span class="n">_incorp_deletion</span><span class="p">,</span>
            <span class="n">VariationType</span><span class="o">.</span><span class="n">INS</span><span class="p">:</span> <span class="n">_incorp_insertion</span><span class="p">,</span>
            <span class="n">VariationType</span><span class="o">.</span><span class="n">FSINS</span><span class="p">:</span> <span class="n">_incorp_insertion</span><span class="p">,</span>
            <span class="n">VariationType</span><span class="o">.</span><span class="n">SNP</span><span class="p">:</span> <span class="n">_incorp_snp</span>
         <span class="p">}</span>


<span class="k">def</span> <span class="nf">_check_for_problematic_variants</span><span class="p">(</span><span class="nb">vars</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters problematic variants, e.g. variants that coincide.</span>

<span class="sd">    :param list(Variant) vars: initial list of variants</span>
<span class="sd">    :return: bool -- true if no intersecting variants were found</span>
<span class="sd">    :invariant: list(Variant) vars: List is sorted based on genome position</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">v_tmp</span> <span class="o">=</span> <span class="nb">vars</span><span class="p">[:]</span>
    <span class="n">v</span> <span class="o">=</span> <span class="n">v_tmp</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
    <span class="n">current_range</span> <span class="o">=</span> <span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">,</span> <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span>
                                      <span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">ref</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="n">VariationType</span><span class="o">.</span><span class="n">FSDEL</span><span class="p">,</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">DEL</span><span class="p">]</span> <span class="k">else</span>
                                      <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">reversed</span><span class="p">(</span><span class="n">v_tmp</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span> <span class="o">&lt;=</span> <span class="n">current_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
            <span class="c">#print &quot;crash&quot;,current_range, v</span>
            <span class="k">return</span> <span class="bp">False</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">current_range</span> <span class="o">=</span> <span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">,</span> <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span>
                                      <span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">ref</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span> <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="n">VariationType</span><span class="o">.</span><span class="n">FSDEL</span><span class="p">,</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">DEL</span><span class="p">]</span> <span class="k">else</span>
                                      <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">)</span>
            <span class="c">#print &quot;new block&quot;,v, current_range</span>
    <span class="k">return</span> <span class="bp">True</span>


<span class="c">#################################################################</span>
<span class="c"># Public transcript generator functions</span>

<span class="c">################################################################################</span>
<span class="c">#        V A R I A N T S     = = &gt;    P E P T I D E S</span>
<span class="c">################################################################################</span>
<div class="viewcode-block" id="generate_peptides_from_variants"><a class="viewcode-back" href="../../../Fred2.Core.html#Fred2.Core.Generator.generate_peptides_from_variants">[docs]</a><span class="k">def</span> <span class="nf">generate_peptides_from_variants</span><span class="p">(</span><span class="nb">vars</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">dbadapter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    generates polymorphic peptides based on given variants</span>

<span class="sd">    :param list(Variant) vars: list of Variants</span>
<span class="sd">    :param int length: length of peptides</span>
<span class="sd">    :param dbadapter: DBAdapter to fetch the transcript sequences</span>
<span class="sd">    :return: List(Peptide) -- A list of polymorphic peptides</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        recursive variant combination generator</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span>
        <span class="c">#print &quot;TransOffset &quot;, transOff, tId,usedVs</span>
        <span class="k">if</span> <span class="n">vs</span><span class="p">:</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">vs</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">isHomozygous</span><span class="p">:</span>
                <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>
                <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">vs_tmp</span> <span class="o">=</span> <span class="n">vs</span><span class="p">[:]</span>
                <span class="n">tmp_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[:]</span>
                <span class="n">tmp_usedVs</span> <span class="o">=</span> <span class="n">usedVs</span><span class="p">[:]</span>

                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs_tmp</span><span class="p">,</span> <span class="n">tmp_seq</span><span class="p">,</span> <span class="n">tmp_usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>

                <span class="c"># update the transcript variant id</span>
                <span class="n">old_trans</span> <span class="o">=</span> <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span>
                <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span>
                <span class="n">_update_var_offset</span><span class="p">(</span><span class="n">usedVs</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">old_trans</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">transOff</span><span class="p">))</span>

                <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>

                <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span>

    <span class="k">def</span> <span class="nf">_generate_heterozygous</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            incorporates heterozygous variants into polymorphic sequences</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">vs</span><span class="p">:</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">vs</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="c">#find offset</span>
            <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">offset</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">uv</span> <span class="ow">in</span> <span class="n">usedVs</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">uv</span><span class="o">.</span><span class="n">genomePos</span> <span class="o">&lt;</span> <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">:</span>
                        <span class="n">offset</span> <span class="o">=</span> <span class="n">uv</span><span class="o">.</span><span class="n">offsets</span><span class="p">[</span><span class="n">tId</span><span class="p">]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">break</span>

            <span class="c">#generate combinatorial branches:</span>
            <span class="n">vs_tmp</span> <span class="o">=</span> <span class="n">vs</span><span class="p">[:]</span>
            <span class="n">tmp_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[:]</span>
            <span class="n">tmp_usedVs</span> <span class="o">=</span> <span class="n">usedVs</span><span class="p">[:]</span>

            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs_tmp</span><span class="p">,</span> <span class="n">tmp_seq</span><span class="p">,</span> <span class="n">tmp_usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>

            <span class="n">old_trans</span> <span class="o">=</span> <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span>
            <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span>
            <span class="n">_update_var_offset</span><span class="p">(</span><span class="n">usedVs</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">old_trans</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">transOff</span><span class="p">))</span>

            <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>

            <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="n">offset</span><span class="p">):</span>
                <span class="k">yield</span> <span class="n">s</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dbadapter</span><span class="p">,</span> <span class="n">ADBAdapter</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&quot;The given dbadapter is not of type ADBAdapter&quot;</span><span class="p">)</span>

    <span class="n">transToVar</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">trans_id</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">coding</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
            <span class="n">transToVar</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">trans_id</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">tId</span><span class="p">,</span> <span class="n">vs</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="k">print</span> <span class="n">tId</span>
        <span class="n">query</span> <span class="o">=</span> <span class="n">dbadapter</span><span class="o">.</span><span class="n">get_transcript_information</span><span class="p">(</span><span class="n">tId</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">query</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Transcript with ID </span><span class="si">%s</span><span class="s"> not found in DB&quot;</span><span class="o">%</span><span class="n">tId</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">tSeq</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">SEQ</span><span class="p">]</span>
        <span class="n">geneid</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">GENE</span><span class="p">]</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">STRAND</span><span class="p">]</span>


        <span class="c">#if its a reverse transcript form the complement of the variants</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s">&quot;-&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="p">[</span><span class="n">tId</span><span class="p">]:</span>
                <span class="n">v</span><span class="o">.</span><span class="n">ref</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">ref</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMPLEMENT</span><span class="p">)</span>
                <span class="n">v</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">obs</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMPLEMENT</span><span class="p">)</span>

        <span class="n">vs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">v</span><span class="p">:</span> <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">_check_for_problematic_variants</span><span class="p">(</span><span class="n">vs</span><span class="p">):</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Intersecting variants found for Transcript </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">tId</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="n">generate_peptides_from_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">vs_homo_and_fs</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="n">VariationType</span><span class="o">.</span><span class="n">FSINS</span><span class="p">,</span> <span class="n">VariationType</span><span class="o">.</span><span class="n">FSDEL</span><span class="p">]</span> <span class="ow">or</span> <span class="n">x</span><span class="o">.</span><span class="n">isHomozygous</span><span class="p">,</span> <span class="n">vs</span><span class="p">)</span>
        <span class="n">vs_hetero</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">isHomozygous</span><span class="p">,</span> <span class="n">vs</span><span class="p">)</span>

        <span class="n">prots</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">tId</span><span class="p">,</span> <span class="n">varSeq</span><span class="p">,</span> <span class="n">varComb</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs_homo_and_fs</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">tSeq</span><span class="p">),</span> <span class="p">[],</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">vs_hetero</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">varSeq</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="o">-</span><span class="mi">3</span><span class="o">*</span><span class="n">length</span><span class="p">):</span>
                    <span class="n">end</span> <span class="o">=</span> <span class="n">i</span><span class="o">+</span><span class="mi">3</span><span class="o">*</span><span class="n">length</span>
                    <span class="n">frac_seq</span> <span class="o">=</span> <span class="n">varSeq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
                    <span class="n">frac_var</span> <span class="o">=</span> <span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="n">x</span><span class="o">.</span><span class="n">get_transcript_position</span> <span class="o">&lt;</span> <span class="n">end</span><span class="p">,</span> <span class="n">vs_hetero</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">ttId</span><span class="p">,</span> <span class="n">vvarSeq</span><span class="p">,</span> <span class="n">vvarComb</span> <span class="ow">in</span> <span class="n">_generate_heterozygous</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">frac_var</span><span class="p">,</span> <span class="n">frac_seq</span><span class="p">,</span> <span class="n">varComb</span><span class="p">):</span>
                        <span class="n">prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="n">generate_proteins_from_transcripts</span><span class="p">(</span>
                                <span class="n">Transcript</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vvarSeq</span><span class="p">),</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">ttId</span><span class="p">,</span> <span class="n">_vars</span><span class="o">=</span><span class="n">vvarComb</span><span class="p">))</span><span class="o">.</span><span class="n">next</span><span class="p">())</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">prots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">generate_proteins_from_transcripts</span><span class="p">(</span>
                        <span class="n">Transcript</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">varSeq</span><span class="p">),</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">tId</span><span class="p">,</span> <span class="n">_vars</span><span class="o">=</span><span class="n">varComb</span><span class="p">))</span><span class="o">.</span><span class="n">next</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">generate_peptides_from_protein</span><span class="p">(</span><span class="n">prots</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span>

<span class="c">################################################################################</span>
<span class="c">#        V A R I A N T S     = = &gt;    T R A N S C R I P T S</span>
<span class="c">################################################################################</span>

</div>
<div class="viewcode-block" id="generate_transcripts_from_variants"><a class="viewcode-back" href="../../../Fred2.Core.html#Fred2.Core.Generator.generate_transcripts_from_variants">[docs]</a><span class="k">def</span> <span class="nf">generate_transcripts_from_variants</span><span class="p">(</span><span class="nb">vars</span><span class="p">,</span> <span class="n">dbadapter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    generates all possible transcript variations of the given variants</span>

<span class="sd">    :param list(Variant) vars: A list of variants for which transcripts should </span>
<span class="sd">                               be build</span>
<span class="sd">    :param ADBAdapter dbadapter: a DBAdapter to fetch the transcript sequences</span>
<span class="sd">    :return: (Generator(Transcripts)) -- a generator of transcripts with all </span>
<span class="sd">             possible variations determined by the given</span>
<span class="sd">             variant list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        recursive variant combination generator</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_transcripts_from_variants</span><span class="o">.</span><span class="n">transOff</span>
        <span class="c">#print &quot;TransOffset &quot;, transOff, tId,usedVs</span>
        <span class="k">if</span> <span class="n">vs</span><span class="p">:</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">vs</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">isHomozygous</span><span class="p">:</span>
                <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>
                <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">vs_tmp</span> <span class="o">=</span> <span class="n">vs</span><span class="p">[:]</span>
                <span class="n">tmp_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[:]</span>
                <span class="n">tmp_usedVs</span> <span class="o">=</span> <span class="n">usedVs</span><span class="p">[:]</span>

                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs_tmp</span><span class="p">,</span> <span class="n">tmp_seq</span><span class="p">,</span> <span class="n">tmp_usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>

                <span class="c"># update the transcript variant id</span>
                <span class="n">old_trans</span> <span class="o">=</span> <span class="n">generate_transcripts_from_variants</span><span class="o">.</span><span class="n">transOff</span>
                <span class="n">generate_transcripts_from_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_transcripts_from_variants</span><span class="o">.</span><span class="n">transOff</span>
                <span class="n">_update_var_offset</span><span class="p">(</span><span class="n">usedVs</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">old_trans</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">transOff</span><span class="p">))</span>

                <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>

                <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span>

    <span class="c">#1) get all transcripts and sort the variants to transcripts</span>

    <span class="c">#For a transcript do:</span>
        <span class="c">#A) get transcript sequences</span>
        <span class="c">#B) generate all possible combinations of variants</span>
        <span class="c">#C) apply variants to transcript and generate transcript object</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dbadapter</span><span class="p">,</span> <span class="n">ADBAdapter</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&quot;The given dbadapter is not of type ADBAdapter&quot;</span><span class="p">)</span>

    <span class="n">transToVar</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">vars</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">trans_id</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">coding</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
            <span class="n">transToVar</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">trans_id</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">tId</span><span class="p">,</span> <span class="n">vs</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="k">print</span> <span class="n">tId</span>
        <span class="n">query</span> <span class="o">=</span> <span class="n">dbadapter</span><span class="o">.</span><span class="n">get_transcript_information</span><span class="p">(</span><span class="n">tId</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">query</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Transcript with ID </span><span class="si">%s</span><span class="s"> not found in DB&quot;</span><span class="o">%</span><span class="n">tId</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">tSeq</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">SEQ</span><span class="p">]</span>
        <span class="n">geneid</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">GENE</span><span class="p">]</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">STRAND</span><span class="p">]</span>


        <span class="c">#if its a reverse transcript form the complement of the variants</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s">&quot;-&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="p">[</span><span class="n">tId</span><span class="p">]:</span>
                <span class="n">v</span><span class="o">.</span><span class="n">ref</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">ref</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMPLEMENT</span><span class="p">)</span>
                <span class="n">v</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">obs</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMPLEMENT</span><span class="p">)</span>

        <span class="n">vs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">v</span><span class="p">:</span> <span class="n">v</span><span class="o">.</span><span class="n">genomePos</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">_check_for_problematic_variants</span><span class="p">(</span><span class="n">vs</span><span class="p">):</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Intersecting variants found for Transcript </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">tId</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="n">generate_transcripts_from_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">tId</span><span class="p">,</span> <span class="n">varSeq</span><span class="p">,</span> <span class="n">varComb</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">tSeq</span><span class="p">),</span> <span class="p">[],</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">Transcript</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">varSeq</span><span class="p">),</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">tId</span><span class="p">,</span> <span class="n">_vars</span><span class="o">=</span><span class="n">varComb</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="generate_transcripts_from_tumor_variants"><a class="viewcode-back" href="../../../Fred2.Core.html#Fred2.Core.Generator.generate_transcripts_from_tumor_variants">[docs]</a><span class="k">def</span> <span class="nf">generate_transcripts_from_tumor_variants</span><span class="p">(</span><span class="n">normal</span><span class="p">,</span> <span class="n">tumor</span><span class="p">,</span> <span class="n">dbadapter</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    generates all possible transcript variations of the given variants</span>

<span class="sd">    :param list(Variant) normal: A list of variants of the normal tissue</span>
<span class="sd">    :param list(Variant) tumor: A list of variant of the cancer tissue for which transcript should be generated</span>
<span class="sd">    :param ADBAdapter dbadapter: a DBAdapter to fetch the transcript sequences</span>
<span class="sd">    :return: (Generator(Transcripts)) -- a generator of transcripts with all</span>
<span class="sd">             possible variations determined by the given</span>
<span class="sd">             variant list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        recursive variant combination generator</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_transcripts_from_tumor_variants</span><span class="o">.</span><span class="n">transOff</span>
        <span class="c">#print &quot;TransOffset &quot;, transOff, tId,usedVs</span>
        <span class="k">if</span> <span class="n">vs</span><span class="p">:</span>
            <span class="n">flag</span><span class="p">,</span> <span class="n">v</span> <span class="o">=</span> <span class="n">vs</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">v</span><span class="o">.</span><span class="n">isHomozygous</span><span class="p">:</span>
                <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>
                <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">vs_tmp</span> <span class="o">=</span> <span class="n">vs</span><span class="p">[:]</span>
                <span class="n">tmp_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[:]</span>
                <span class="n">tmp_usedVs</span> <span class="o">=</span> <span class="n">usedVs</span><span class="p">[:]</span>

                <span class="k">if</span> <span class="n">flag</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs_tmp</span><span class="p">,</span> <span class="n">tmp_seq</span><span class="p">,</span> <span class="n">tmp_usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                        <span class="k">yield</span> <span class="n">s</span>

                <span class="c"># update the transcript variant id</span>
                <span class="n">old_trans</span> <span class="o">=</span> <span class="n">generate_transcripts_from_tumor_variants</span><span class="o">.</span><span class="n">transOff</span>
                <span class="n">generate_transcripts_from_tumor_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">transOff</span> <span class="o">=</span> <span class="n">generate_transcripts_from_tumor_variants</span><span class="o">.</span><span class="n">transOff</span>
                <span class="n">_update_var_offset</span><span class="p">(</span><span class="n">usedVs</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">old_trans</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="p">(</span><span class="n">transOff</span><span class="p">))</span>

                <span class="n">seq</span><span class="p">,</span> <span class="n">offset</span> <span class="o">=</span> <span class="n">_incorp</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">c</span><span class="p">,</span> <span class="n">d</span><span class="p">:</span> <span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">d</span><span class="p">))(</span><span class="n">seq</span><span class="p">,</span> <span class="n">v</span><span class="p">,</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>

                <span class="n">usedVs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span><span class="p">,</span> <span class="n">offset</span><span class="p">):</span>
                    <span class="k">yield</span> <span class="n">s</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">yield</span> <span class="n">tId</span><span class="o">+</span><span class="s">&quot;:FRED2_</span><span class="si">%i</span><span class="s">&quot;</span><span class="o">%</span><span class="n">transOff</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">usedVs</span>

    <span class="c">#1) get all transcripts and sort the variants to transcripts</span>

    <span class="c">#For a transcript do:</span>
        <span class="c">#A) get transcript sequences</span>
        <span class="c">#B) generate all possible combinations of variants</span>
        <span class="c">#C) apply variants to transcript and generate transcript object</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">dbadapter</span><span class="p">,</span> <span class="n">ADBAdapter</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&quot;The given dbadapter is not of type ADBAdapter&quot;</span><span class="p">)</span>

    <span class="n">transToVar</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">tumor</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">trans_id</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">coding</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
            <span class="n">transToVar</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">trans_id</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="bp">False</span><span class="p">,</span> <span class="n">v</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">normal</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">trans_id</span> <span class="ow">in</span> <span class="n">v</span><span class="o">.</span><span class="n">coding</span><span class="o">.</span><span class="n">iterkeys</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">trans_id</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="p">:</span>
                <span class="n">transToVar</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">trans_id</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="bp">True</span><span class="p">,</span> <span class="n">v</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">tId</span><span class="p">,</span> <span class="n">vs</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="o">.</span><span class="n">iteritems</span><span class="p">():</span>
        <span class="n">query</span> <span class="o">=</span> <span class="n">dbadapter</span><span class="o">.</span><span class="n">get_transcript_information</span><span class="p">(</span><span class="n">tId</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">query</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Transcript with ID </span><span class="si">%s</span><span class="s"> not found in DB&quot;</span><span class="o">%</span><span class="n">tId</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">tSeq</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">SEQ</span><span class="p">]</span>
        <span class="n">geneid</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">GENE</span><span class="p">]</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">query</span><span class="p">[</span><span class="n">EAdapterFields</span><span class="o">.</span><span class="n">STRAND</span><span class="p">]</span>


        <span class="c">#if its a reverse transcript form the complement of the variants</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s">&quot;-&quot;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">flag</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">transToVar</span><span class="p">[</span><span class="n">tId</span><span class="p">]:</span>
                <span class="n">v</span><span class="o">.</span><span class="n">ref</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">ref</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMPLEMENT</span><span class="p">)</span>
                <span class="n">v</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">obs</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">COMPLEMENT</span><span class="p">)</span>

        <span class="n">vs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">v</span><span class="p">:</span> <span class="n">v</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">genomePos</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">_check_for_problematic_variants</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">vs</span><span class="p">)):</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&quot;Intersecting variants found for Transcript </span><span class="si">%s</span><span class="s">&quot;</span><span class="o">%</span><span class="n">tId</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="n">generate_transcripts_from_tumor_variants</span><span class="o">.</span><span class="n">transOff</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">tId</span><span class="p">,</span> <span class="n">varSeq</span><span class="p">,</span> <span class="n">varComb</span> <span class="ow">in</span> <span class="n">_generate_combinations</span><span class="p">(</span><span class="n">tId</span><span class="p">,</span> <span class="n">vs</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">tSeq</span><span class="p">),</span> <span class="p">[],</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">Transcript</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">varSeq</span><span class="p">),</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">tId</span><span class="p">,</span> <span class="n">_vars</span><span class="o">=</span><span class="n">varComb</span><span class="p">)</span>


<span class="c">################################################################################</span>
<span class="c">#        T R A N S C R I P T    = = &gt;    P R O T E I N</span>
<span class="c">################################################################################</span></div>
<div class="viewcode-block" id="generate_proteins_from_transcripts"><a class="viewcode-back" href="../../../Fred2.Core.html#Fred2.Core.Generator.generate_proteins_from_transcripts">[docs]</a><span class="k">def</span> <span class="nf">generate_proteins_from_transcripts</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span> <span class="n">table</span><span class="o">=</span><span class="s">&#39;Standard&#39;</span><span class="p">,</span> <span class="n">stop_symbol</span><span class="o">=</span><span class="s">&#39;*&#39;</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Enables the translation from a transcript to a protein instance</span>

<span class="sd">        :param: list(Transcript)/Transcript - a list of transcripts to translate to proteins</span>
<span class="sd">        :returns: (Protein) -- the protein that corresponds to the transcript</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span> <span class="n">Transcript</span><span class="p">):</span>
            <span class="n">transcripts</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcripts</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">Transcript</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">transcripts</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Specified input is not of type Transcript&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">transcripts</span><span class="p">:</span>
            <span class="c"># translate to a protein sequence</span>
            <span class="c">#if len(str(self)) % 3 != 0:</span>
            <span class="c">#    raise ValueError(&#39;ERROR while translating: lenght of transcript %s is no multiple of 3, the transcript is:\n %s&#39; % (self.transcript_id, self))</span>

            <span class="c">#TODO warn if intrasequence stops - biopython warns if  % 3 != 0</span>
            <span class="n">prot_seq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">table</span><span class="p">,</span> <span class="n">stop_symbol</span><span class="o">=</span><span class="n">stop_symbol</span><span class="p">,</span> <span class="n">to_stop</span><span class="o">=</span><span class="n">to_stop</span><span class="p">,</span> <span class="n">cds</span><span class="o">=</span><span class="n">cds</span><span class="p">))</span>

            <span class="c"># only transfer the non-synonymous variants to the protein as an</span>
            <span class="c"># dict, also translate into protein positions</span>
            <span class="n">new_vars</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">var</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">vars</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">var</span><span class="o">.</span><span class="n">isSynonymous</span><span class="p">:</span>
                    <span class="n">pos</span> <span class="o">=</span> <span class="n">var</span><span class="o">.</span><span class="n">get_protein_position</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">transcript_id</span><span class="p">)</span>
                    <span class="n">new_vars</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="p">[])</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">var</span><span class="p">)</span>

            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">gene_id</span>
            <span class="k">yield</span> <span class="n">Protein</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">transcript_id</span><span class="p">,</span> <span class="n">t</span><span class="p">,</span> <span class="n">new_vars</span><span class="p">)</span>


<span class="c">################################################################################</span>
<span class="c">#        P R O T E I N    = = &gt;    P E P T I D E</span>
<span class="c">################################################################################</span>
</div>
<div class="viewcode-block" id="generate_peptides_from_protein"><a class="viewcode-back" href="../../../Fred2.Core.html#Fred2.Core.Generator.generate_peptides_from_protein">[docs]</a><span class="k">def</span> <span class="nf">generate_peptides_from_protein</span><span class="p">(</span><span class="n">proteins</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">peptides</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates all peptides for a given window size, from a given protein. The</span>
<span class="sd">    result is a generator.</span>

<span class="sd">    :param Protein proteins: (list of) protein(s) from which a list of unique</span>
<span class="sd">                            peptides should be generated</span>
<span class="sd">    :param int window_size: size of peptide fragments</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">gen_peptide_info</span><span class="p">(</span><span class="n">protein</span><span class="p">):</span>
        <span class="c"># Generate peptide sequences and returns the sequence</span>
        <span class="c"># #and start position within the protein</span>
        <span class="n">res</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">seq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">xrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="o">-</span><span class="n">window_size</span><span class="p">):</span>
            <span class="c"># generate peptide fragment</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">i</span><span class="o">+</span><span class="n">window_size</span>
            <span class="n">pep_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
            <span class="n">res</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">pep_seq</span><span class="p">,</span> <span class="n">i</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">res</span>

    <span class="c"># sequence : peptide-instance</span>
    <span class="n">final_peptides</span> <span class="o">=</span> <span class="p">{}</span> <span class="k">if</span> <span class="n">peptides</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="p">{</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">):</span><span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">peptides</span><span class="p">}</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">proteins</span><span class="p">,</span> <span class="n">Protein</span><span class="p">):</span>
        <span class="n">proteins</span> <span class="o">=</span> <span class="p">[</span><span class="n">proteins</span><span class="p">]</span>

    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">Protein</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">proteins</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Input does contain non protein objects.&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">prot</span> <span class="ow">in</span> <span class="n">proteins</span><span class="p">:</span>
        <span class="c"># generate all peptide sequences per protein:</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">_pos</span><span class="p">)</span> <span class="ow">in</span> <span class="n">gen_peptide_info</span><span class="p">(</span><span class="n">prot</span><span class="p">):</span>

            <span class="n">t_id</span> <span class="o">=</span> <span class="n">prot</span><span class="o">.</span><span class="n">transcript_id</span>
            <span class="k">if</span> <span class="n">seq</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">final_peptides</span><span class="p">:</span>
                <span class="n">final_peptides</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span> <span class="o">=</span> <span class="n">Peptide</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">t_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">final_peptides</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">:</span>
                <span class="n">final_peptides</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span><span class="o">.</span><span class="n">proteins</span><span class="p">[</span><span class="n">t_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">prot</span>
                <span class="n">final_peptides</span><span class="p">[</span><span class="n">seq</span><span class="p">]</span><span class="o">.</span><span class="n">proteinPos</span><span class="p">[</span><span class="n">t_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_pos</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">final_peptides</span><span class="o">.</span><span class="n">values</span><span class="p">()</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015, Author.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.2</a>
      
    </div>

    

    
  </body>
</html>